// @trimble-oss/moduswebcomponents/collection/components/modus-wc-typography/modus-wc-typography.js
import { h, Host } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-typography.tailwind";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable typography component used to render text with different sizes, hierarchy, and weights.
 *
 * Note:
 * - When using heading elements (h1-h6), the default heading CSS styling can be accessed without modifying
 * the default size (size="md") and weight (weight="normal") properties. Default styling can be overridden by
 * providing your own custom values for the size or weight properties from the available options.
 *

 * - If both slot content and `label` are provided, only the slot content will be rendered
 * - Use the `label` prop when you need to dynamically update the text.
 */
export class ModusWCTypography {
    constructor() {
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the typography element. */
        this.customClass = '';
        /** The hierarchy of the typography component. */
        this.hierarchy = 'p';
        /** The size of the font. */
        this.size = 'md';
        /** The weight of the text. */
        this.weight = 'normal';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-typography'];
        // Check if we're dealing with a heading and have size/weight overrides
        const isHeading = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].includes(this.hierarchy);
        const hasOverrides = this.size !== 'md' || this.weight !== 'normal';
        if (isHeading && hasOverrides) {
            // Add a class to indicate overrides for headings
            classList.push('modus-wc-typography-override');
        }
        const propClasses = convertPropsToClasses({
            size: this.size,
            weight: this.weight,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        const Element = this.hierarchy;
        return (h(Host, { key: 'b9f340f7348ccf18ae168020cccabe14a1a8c5de' }, h(Element, Object.assign({ key: '6bf43d72c246ac1d38b241f8596d953dbed044a0', class: this.getClasses() }, this.inheritedAttributes), h("slot", { key: '7f59a5319bbf237101acaa67065bf5bec3565c4e' }, this.label))));
    }
    static get is() { return "modus-wc-typography"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-typography.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-typography.css"]
        };
    }
    static get properties() {
        return {
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the typography element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "hierarchy": {
                "type": "string",
                "attribute": "hierarchy",
                "mutable": false,
                "complexType": {
                    "original": "TypographyHierarchy",
                    "resolved": "\"h1\" | \"h2\" | \"h3\" | \"h4\" | \"h5\" | \"h6\" | \"p\"",
                    "references": {
                        "TypographyHierarchy": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::TypographyHierarchy"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The hierarchy of the typography component."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'p'"
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The text label to display."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "TypographySize",
                    "resolved": "\"2xl\" | \"3xl\" | \"4xl\" | \"5xl\" | \"6xl\" | \"7xl\" | \"8xl\" | \"9xl\" | \"lg\" | \"md\" | \"sm\" | \"xl\" | \"xs\" | undefined",
                    "references": {
                        "TypographySize": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::TypographySize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The size of the font."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'md'"
            },
            "weight": {
                "type": "string",
                "attribute": "weight",
                "mutable": false,
                "complexType": {
                    "original": "TypographyWeight",
                    "resolved": "\"bold\" | \"light\" | \"normal\" | \"semibold\" | undefined",
                    "references": {
                        "TypographyWeight": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::TypographyWeight"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The weight of the text."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'normal'"
            }
        };
    }
    static get elementRef() { return "el"; }
}
