// @trimble-oss/moduswebcomponents/collection/components/modus-wc-navbar/modus-wc-navbar.js
// TODO - add coverage
// istanbul ignore file
import { Fragment, h, Host, } from "@stencil/core";
import { AiDarkIcon } from "../../icons/ai-dark.icon";
import { AiLightIcon } from "../../icons/ai-light.icon";
import { AppsSolidIcon } from "../../icons/apps-solid.icon";
import { HelpSolidIcon } from "../../icons/help-solid.icon";
import { MenuSolidIcon } from "../../icons/menu-solid.icon";
import { MoreVerticalSolidIcon } from "../../icons/more-vertical-solid.icon";
import { NotificationsSolidIcon } from "../../icons/notifications-solid.icon";
import { SearchSolidIcon } from "../../icons/search-solid.icon";
import { handleShadowDOMStyles } from "../base-component";
import { inheritAriaAttributes, isLightMode } from "../utils";
/**
 * A customizable navbar component used for top level navigation of all Trimble applications.
 */
export class ModusWcNavbar {
    constructor() {
        this.inheritedAttributes = {};
        this.searchDebounceTimer = null;
        /** The name of the logo to display. Supports any valid 'logo-name' from the 'modus-wc-logo' component. Defaults to 'trimble'. */
        this.logoName = 'trimble';
        /** The open state of the apps menu. */
        this.appsMenuOpen = false;
        /** Applies condensed layout and styling. */
        this.condensed = false;
        /** The open state of the condensed menu. */
        this.condensedMenuOpen = false;
        /** Custom CSS class to apply to the host element. */
        this.customClass = '';
        /** The open state of the main menu. */
        this.mainMenuOpen = false;
        /** The open state of the notifications menu. */
        this.notificationsMenuOpen = false;
        /** Debounce time in milliseconds for search input changes. Default is 300ms. */
        this.searchDebounceMs = 300;
        /** The open state of the search input. */
        this.searchInputOpen = false;
        /** The open state of the user menu. */
        this.userMenuOpen = false;
        /** The visibility of individual navbar buttons. Default is user profile visible, others hidden. */
        this.visibility = {
            ai: false,
            apps: false,
            help: false,
            logo: true,
            mainMenu: false,
            notifications: false,
            search: false,
            searchInput: false,
            user: true,
        };
        this.isLight = true;
        this.searchValue = '';
        this.themeObserver = null;
        this.handleAiClick = (event) => {
            this.aiClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleAppsClick = (event) => {
            this.toggleApps();
            this.appsClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleHelpClick = (event) => {
            if (this.condensed) {
                this.toggleCondensedMenu();
            }
            this.helpClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleMyTrimbleClick = (event) => {
            this.myTrimbleClick.emit(event.detail);
        };
        this.handleNotificationsClick = (event) => {
            this.toggleNotifications();
            this.notificationsClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleSearchChange = (event) => {
            // Update the search value immediately for UI responsiveness
            this.searchValue = event.target.value;
            // Clear any existing timer
            if (this.searchDebounceTimer !== null) {
                window.clearTimeout(this.searchDebounceTimer);
            }
            // Set a new timer
            this.searchDebounceTimer = window.setTimeout(() => {
                this.searchChange.emit({ value: this.searchValue });
                this.searchDebounceTimer = null;
            }, this.searchDebounceMs);
        };
        this.handleSearchClick = (event) => {
            this.toggleSearch();
            this.searchClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleSignOutClick = (event) => {
            this.signOutClick.emit(event.detail);
        };
        this.handleTrimbleLogoClick = (event) => {
            this.trimbleLogoClick.emit(event.detail);
        };
        this.toggleApps = () => {
            if (this.condensed) {
                this.toggleCondensedMenu();
            }
            else {
                this.appsMenuOpen = !this.appsMenuOpen;
                this.appsMenuOpenChange.emit(this.appsMenuOpen);
            }
        };
        this.toggleCondensedMenu = () => {
            this.condensedMenuOpen = !this.condensedMenuOpen;
            this.condensedMenuOpenChange.emit(this.condensedMenuOpen);
        };
        this.toggleMainMenu = () => {
            this.mainMenuOpen = !this.mainMenuOpen;
            this.mainMenuOpenChange.emit(this.mainMenuOpen);
        };
        this.toggleNotifications = () => {
            if (this.condensed) {
                this.toggleCondensedMenu();
            }
            else {
                this.notificationsMenuOpen = !this.notificationsMenuOpen;
                this.notificationsMenuOpenChange.emit(this.notificationsMenuOpen);
            }
        };
        this.toggleSearch = () => {
            if (this.condensed) {
                this.toggleCondensedMenu();
            }
            else {
                this.searchInputOpen = !this.searchInputOpen;
                this.searchInputOpenChange.emit(this.searchInputOpen);
            }
        };
        this.toggleUser = () => {
            this.userMenuOpen = !this.userMenuOpen;
            this.userMenuOpenChange.emit(this.userMenuOpen);
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        this.inheritedAttributes = inheritAriaAttributes(this.el);
        this.isLight = isLightMode();
        // Watch for theme attribute changes
        this.themeObserver = new MutationObserver(() => {
            this.isLight = isLightMode();
        });
        // Observe the html element for data-theme attribute changes
        this.themeObserver.observe(document.documentElement, {
            attributes: true,
            attributeFilter: ['data-theme'],
        });
    }
    disconnectedCallback() {
        if (this.themeObserver) {
            this.themeObserver.disconnect();
        }
    }
    handleClickOutside(event) {
        const path = event.composedPath ? event.composedPath() : [event.target];
        if (this.appsMenuOpen) {
            const appsButton = this.el.querySelector('modus-wc-button:has(svg[class*="apps"])');
            if (this.appsRef &&
                !path.includes(this.appsRef) &&
                !path.includes(appsButton)) {
                this.appsMenuOpen = false;
                this.appsMenuOpenChange.emit(false);
            }
        }
        if (this.condensedMenuOpen) {
            const condenseMenuButton = this.el.querySelector('modus-wc-button:has(svg[class*="more-vertical"])');
            if (this.condensedMenuRef &&
                !path.includes(this.condensedMenuRef) &&
                !path.includes(condenseMenuButton)) {
                this.condensedMenuOpen = false;
                this.condensedMenuOpenChange.emit(false);
            }
        }
        if (this.mainMenuOpen) {
            const menuButton = this.el.querySelector('modus-wc-button:has(svg[class*="menu"])');
            if (this.menuRef &&
                !path.includes(this.menuRef) &&
                !path.includes(menuButton)) {
                this.mainMenuOpen = false;
            }
        }
        if (this.notificationsMenuOpen) {
            const notificationsButton = this.el.querySelector('modus-wc-button:has(svg[class*="notifications"])');
            if (this.notificationsRef &&
                !path.includes(this.notificationsRef) &&
                !path.includes(notificationsButton)) {
                this.notificationsMenuOpen = false;
                this.notificationsMenuOpenChange.emit(false);
            }
        }
        if (this.userMenuOpen) {
            const userButton = this.el.querySelector('modus-wc-button:has([class*="user-button"])');
            if (this.userRef &&
                !path.includes(this.userRef) &&
                !path.includes(userButton)) {
                this.userMenuOpen = false;
                this.userMenuOpenChange.emit(false);
            }
        }
    }
    getClasses() {
        const classList = [''];
        // The order CSS classes are added matters to CSS specificity
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7;
        const condensedHasItems = ((_a = this.visibility) === null || _a === void 0 ? void 0 : _a.apps) ||
            ((_b = this.visibility) === null || _b === void 0 ? void 0 : _b.help) ||
            ((_c = this.visibility) === null || _c === void 0 ? void 0 : _c.notifications) ||
            ((_d = this.visibility) === null || _d === void 0 ? void 0 : _d.search);
        const accessibleName = (this.logoName || 'trimble')
            .replace(/_/g, ' ')
            .replace(/\b\w/g, (char) => char.toUpperCase());
        return (h(Host, Object.assign({ key: 'a96c0a2319d2f2ef4a88a21526f3ae25c68a9ae7', class: this.getClasses() }, this.inheritedAttributes), h("modus-wc-toolbar", { key: 'd9c07e6f478ad4860b18df7c599602e13d780806' }, h("div", { key: 'ca4a0291a3e4bbfd039ee14309743645776b722e', slot: "start" }, ((_e = this.visibility) === null || _e === void 0 ? void 0 : _e.mainMenu) && (h(Fragment, { key: 'b6331e0655045949275e66468e67fa842ace3ea8' }, h("modus-wc-button", { key: 'b6be968a376e2b14d636302704d3a3f89541253e', "aria-label": "Main menu", onButtonClick: this.toggleMainMenu, shape: "square", size: "sm", variant: "borderless" }, h(MenuSolidIcon, { key: 'df192f45e0b87aced69ca6fe97beeca559f938cf' })), h("div", { key: 'a41e65cd3cf4db1779bbbf3996818479ff0ef251', class: `main-menu ${this.mainMenuOpen ? 'visible' : 'hidden'}`, ref: (el) => (this.menuRef = el) }, h("slot", { key: 'e24743884b7d9069d71daf361755f51aba9996d9', name: "main-menu" })))), ((_f = this.visibility) === null || _f === void 0 ? void 0 : _f.logo) !== false && (h("modus-wc-button", { key: '7069e5c3a245ade840093907959860fcfa0fd84c', "aria-label": `${accessibleName} logo`, customClass: "logo", onButtonClick: this.handleTrimbleLogoClick, size: "sm", variant: "borderless" }, h("modus-wc-logo", { key: '687aa275911464af35b570dc4a6810f9a78ba1df', name: this.logoName || 'trimble', emblem: this.condensed }))), h("slot", { key: 'e49084b82343f0d161dba0975227ff452346c541', name: "start" })), h("div", { key: '966773d1b7330f25726cff5606fa2e498b301a1f', slot: "center" }, h("slot", { key: '40e4719991c6d35fbd8ae2d6595238bd733edc61', name: "center" })), h("div", { key: 'b59fa42ff2daa381c606177e96b246eb824b99ba', slot: "end" }, h("slot", { key: 'f38a0b1fcee9a8693ef0b938ff1a5d278733d9c8', name: "end" }), ((_g = this.visibility) === null || _g === void 0 ? void 0 : _g.ai) && (h(Fragment, { key: '672c2710f8870ced86378538d69630017faee7ff' }, h("modus-wc-button", { key: 'c000cc90884e74e259d5635434096efc27dc6a41', "aria-label": "AI assistant", customClass: "ai", onButtonClick: this.handleAiClick, shape: "square", size: "sm", variant: "borderless" }, this.isLight ? h(AiLightIcon, null) : h(AiDarkIcon, null)))), this.condensed && condensedHasItems && (h(Fragment, { key: 'b23dfa323a818cbd20d3c7e98e37b724dbc00226' }, h("modus-wc-button", { key: '60df9dde5da5e66dcfaf55863b603910c859fb6e', "aria-label": "More options", onButtonClick: this.toggleCondensedMenu, shape: "square", size: "sm", variant: "borderless" }, h(MoreVerticalSolidIcon, { key: '7e4a6cf4dabf5d9ea91d18915ce6746fb2433c12' })), this.condensedMenuOpen && (h("modus-wc-menu", { key: '686ee34b1283158eaeb65dcd47664fb7db3b7d60', bordered: true, customClass: "condense-menu", ref: (el) => (this.condensedMenuRef = el) }, ((_h = this.visibility) === null || _h === void 0 ? void 0 : _h.search) && (h("modus-wc-menu-item", { key: 'aaf51a3ef070f7a2cee76e5346e745190fdf75f1', label: ((_j = this.textOverrides) === null || _j === void 0 ? void 0 : _j.search) || 'Search', onItemSelect: () => this.handleSearchClick(), value: "search" })), ((_k = this.visibility) === null || _k === void 0 ? void 0 : _k.notifications) && (h("modus-wc-menu-item", { key: '1bbfb98c1aaea55c28824afb32d3530a8fba38e8', label: ((_l = this.textOverrides) === null || _l === void 0 ? void 0 : _l.notifications) || 'Notifications', onItemSelect: () => this.handleNotificationsClick(), value: "notifications" })), ((_m = this.visibility) === null || _m === void 0 ? void 0 : _m.help) && (h("modus-wc-menu-item", { key: '3c729efc88d91bb40c46688e14c0c2ae84d49cb2', label: ((_o = this.textOverrides) === null || _o === void 0 ? void 0 : _o.help) || 'Help', onItemSelect: () => this.handleHelpClick(), value: "help" })), ((_p = this.visibility) === null || _p === void 0 ? void 0 : _p.apps) && (h("modus-wc-menu-item", { key: 'e622c9e085472d47f0a914bde52645f8d2a7f00e', label: ((_q = this.textOverrides) === null || _q === void 0 ? void 0 : _q.apps) || 'Apps', onItemSelect: () => this.handleAppsClick(), value: "apps" })))))), ((_r = this.visibility) === null || _r === void 0 ? void 0 : _r.search) && !this.condensed && (h(Fragment, { key: '89dfb9e22e604cbe60dba111b818b442cbc291e9' }, ((_s = this.visibility) === null || _s === void 0 ? void 0 : _s.searchInput) && this.searchInputOpen && (h("modus-wc-text-input", { key: '004b55ac6610f7aaaa3bb5cc9e71a1e78e67cd93', includeClear: true, includeSearch: true, onInputChange: this.handleSearchChange, placeholder: ((_t = this.textOverrides) === null || _t === void 0 ? void 0 : _t.search) || 'Search', size: "sm", value: this.searchValue })), h("modus-wc-button", { key: 'c25e08a4ca0d65437aa47eb3885db97b40a5975f', "aria-label": "Search", onButtonClick: this.handleSearchClick, shape: "square", size: "sm", variant: "borderless" }, h(SearchSolidIcon, { key: '60d6121651269f4b8d58d4cabffa694aef808684' })))), ((_u = this.visibility) === null || _u === void 0 ? void 0 : _u.notifications) && !this.condensed && (h(Fragment, { key: 'e23fbf1ecf390eb115945e9c5d81f8ccc3f08b16' }, h("modus-wc-button", { key: '81a7d2ff8e8ce919d4baaeaa9c855b8ccb29a496', "aria-label": "Notifications", onButtonClick: this.handleNotificationsClick, shape: "square", size: "sm", variant: "borderless" }, h(NotificationsSolidIcon, { key: '0bb481ee4264decd8554afa0f411e289c084456c' })), h("div", { key: '180f55b4c89a9ece7c3c9464d2fe017aa9ac2525', class: `notifications ${this.notificationsMenuOpen ? 'visible' : 'hidden'}`, ref: (el) => (this.notificationsRef = el) }, h("slot", { key: 'dc3c02f0a431740b77add477214d83edd1ead726', name: "notifications" })))), ((_v = this.visibility) === null || _v === void 0 ? void 0 : _v.help) && !this.condensed && (h("modus-wc-button", { key: 'c846db586e982dec0364b79cdf3b13e4019b1414', "aria-label": "Help", onButtonClick: this.handleHelpClick, shape: "square", size: "sm", variant: "borderless" }, h(HelpSolidIcon, { key: '8a86189c80fa17d9ad420eb0ed9ae98e368ac935' }))), ((_w = this.visibility) === null || _w === void 0 ? void 0 : _w.apps) && !this.condensed && (h(Fragment, { key: 'aa2bc840fab19c8aa2c2c77b668493e04671973f' }, h("modus-wc-button", { key: 'c680be3d30780e247c0330abe26273fa8ba25c63', "aria-label": "Apps", onButtonClick: this.handleAppsClick, shape: "square", size: "sm", variant: "borderless" }, h(AppsSolidIcon, { key: '42047b4d08693b4a16d0e3b56cadac31b2528d82' })), h("div", { key: 'c7ff74df644692b970acf2a4eec6dea83db05147', class: `apps ${this.appsMenuOpen ? 'visible' : 'hidden'}`, ref: (el) => (this.appsRef = el) }, h("slot", { key: '592395ca8508d30800561d42e60541eb409f1e2b', name: "apps" })))), ((_x = this.visibility) === null || _x === void 0 ? void 0 : _x.user) && (h(Fragment, { key: '0a94e30ed9398bf8e98badd32ffb6218637fdcad' }, h("modus-wc-button", { key: '849bf9b2e3bcb3256c800f3aba3a60f35651153a', "aria-label": "User profile", customClass: "user-button", onButtonClick: this.toggleUser, shape: "circle", size: "sm", variant: "borderless" }, h("modus-wc-avatar", { key: '4013ddeea80555a4458f852ab1c6882588fff5bf', alt: ((_y = this.userCard) === null || _y === void 0 ? void 0 : _y.avatarAlt) || '', imgSrc: (_z = this.userCard) === null || _z === void 0 ? void 0 : _z.avatarSrc, initials: (_0 = this.userCard) === null || _0 === void 0 ? void 0 : _0.name, size: "xs" })), h("div", { key: '476e159c4fdbe1c5ad737978b71b7db03d91d408', class: `user ${this.userMenuOpen ? 'visible' : 'hidden'}`, ref: (el) => (this.userRef = el) }, h("modus-wc-card", { key: 'c3d4d4303f987e1df8f351176812ebbb1c177e0d' }, h("div", { key: '2316ec90fb1caa232dd584db490d1be837940654', slot: "header" }, h("modus-wc-avatar", { key: '52811d868a45c3d5902819cabd6a06a426dbf944', alt: ((_1 = this.userCard) === null || _1 === void 0 ? void 0 : _1.avatarAlt) || '', imgSrc: (_2 = this.userCard) === null || _2 === void 0 ? void 0 : _2.avatarSrc, initials: (_3 = this.userCard) === null || _3 === void 0 ? void 0 : _3.name })), h("div", { key: 'db820f5379b1de56c8f6a7e16de8d9d80247a71a', slot: "title" }, (_4 = this.userCard) === null || _4 === void 0 ? void 0 : _4.name), h("div", { key: '79190afffffa0cc62aebf099acbe8637a15c4ffa' }, (_5 = this.userCard) === null || _5 === void 0 ? void 0 : _5.email), h("div", { key: '2b87dcebe0948a6a19ca6f63b75d640a37591c20', slot: "actions" }, h("modus-wc-button", { key: 'a2dbe18354bc72eede0eaece239df961686dcfbc', customClass: "my-trimble", onButtonClick: this.handleMyTrimbleClick }, ((_6 = this.userCard) === null || _6 === void 0 ? void 0 : _6.myTrimbleButton) || 'Access MyTrimble')), h("div", { key: '30f50417ac0d1ad4767d05172085162fdfd63f0a', slot: "footer" }, h("modus-wc-button", { key: '5019a97641df187bb58a79cdb16e0b179fe5a069', customClass: "sign-out", onButtonClick: this.handleSignOutClick, variant: "borderless" }, ((_7 = this.userCard) === null || _7 === void 0 ? void 0 : _7.signOutButton) || 'Sign out'))))))))));
    }
    static get is() { return "modus-wc-navbar"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-navbar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-navbar.css"]
        };
    }
    static get properties() {
        return {
            "logoName": {
                "type": "string",
                "attribute": "logo-name",
                "mutable": false,
                "complexType": {
                    "original": "LogoName",
                    "resolved": "LogoName | undefined",
                    "references": {
                        "LogoName": {
                            "location": "import",
                            "path": "../modus-wc-logo/logo-constants",
                            "id": "src/components/modus-wc-logo/logo-constants.ts::LogoName"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The name of the logo to display. Supports any valid 'logo-name' from the 'modus-wc-logo' component. Defaults to 'trimble'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'trimble'"
            },
            "appsMenuOpen": {
                "type": "boolean",
                "attribute": "apps-menu-open",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The open state of the apps menu."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "condensed": {
                "type": "boolean",
                "attribute": "condensed",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Applies condensed layout and styling."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "condensedMenuOpen": {
                "type": "boolean",
                "attribute": "condensed-menu-open",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The open state of the condensed menu."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the host element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "mainMenuOpen": {
                "type": "boolean",
                "attribute": "main-menu-open",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The open state of the main menu."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "notificationsMenuOpen": {
                "type": "boolean",
                "attribute": "notifications-menu-open",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The open state of the notifications menu."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "searchDebounceMs": {
                "type": "number",
                "attribute": "search-debounce-ms",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Debounce time in milliseconds for search input changes. Default is 300ms."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "300"
            },
            "searchInputOpen": {
                "type": "boolean",
                "attribute": "search-input-open",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The open state of the search input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "textOverrides": {
                "type": "unknown",
                "attribute": "text-overrides",
                "mutable": false,
                "complexType": {
                    "original": "INavbarTextOverrides",
                    "resolved": "INavbarTextOverrides | undefined",
                    "references": {
                        "INavbarTextOverrides": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-navbar/modus-wc-navbar.tsx",
                            "id": "src/components/modus-wc-navbar/modus-wc-navbar.tsx::INavbarTextOverrides"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Text replacements for the navbar."
                },
                "getter": false,
                "setter": false
            },
            "userCard": {
                "type": "unknown",
                "attribute": "user-card",
                "mutable": false,
                "complexType": {
                    "original": "INavbarUserCard",
                    "resolved": "INavbarUserCard",
                    "references": {
                        "INavbarUserCard": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-navbar/modus-wc-navbar.tsx",
                            "id": "src/components/modus-wc-navbar/modus-wc-navbar.tsx::INavbarUserCard"
                        }
                    }
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "User information used to render the user card."
                },
                "getter": false,
                "setter": false
            },
            "userMenuOpen": {
                "type": "boolean",
                "attribute": "user-menu-open",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The open state of the user menu."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "visibility": {
                "type": "unknown",
                "attribute": "visibility",
                "mutable": false,
                "complexType": {
                    "original": "INavbarVisibility",
                    "resolved": "INavbarVisibility | undefined",
                    "references": {
                        "INavbarVisibility": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-navbar/modus-wc-navbar.tsx",
                            "id": "src/components/modus-wc-navbar/modus-wc-navbar.tsx::INavbarVisibility"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The visibility of individual navbar buttons. Default is user profile visible, others hidden."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "{\n    ai: false,\n    apps: false,\n    help: false,\n    logo: true,\n    mainMenu: false,\n    notifications: false,\n    search: false,\n    searchInput: false,\n    user: true,\n  }"
            }
        };
    }
    static get states() {
        return {
            "isLight": {},
            "searchValue": {}
        };
    }
    static get events() {
        return [{
                "method": "aiClick",
                "name": "aiClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the AI button is clicked or activated via keyboard."
                },
                "complexType": {
                    "original": "MouseEvent | KeyboardEvent",
                    "resolved": "KeyboardEvent | MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        },
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "appsClick",
                "name": "appsClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the apps button is clicked or activated via keyboard."
                },
                "complexType": {
                    "original": "MouseEvent | KeyboardEvent",
                    "resolved": "KeyboardEvent | MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        },
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "appsMenuOpenChange",
                "name": "appsMenuOpenChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the apps menu open state changes."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "condensedMenuOpenChange",
                "name": "condensedMenuOpenChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the condensed menu open state changes."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "helpClick",
                "name": "helpClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the help button is clicked or activated via keyboard."
                },
                "complexType": {
                    "original": "MouseEvent | KeyboardEvent",
                    "resolved": "KeyboardEvent | MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        },
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "mainMenuOpenChange",
                "name": "mainMenuOpenChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the main menu open state changes."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "myTrimbleClick",
                "name": "myTrimbleClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the user profile Access MyTrimble button is clicked or activated via keyboard."
                },
                "complexType": {
                    "original": "MouseEvent | KeyboardEvent",
                    "resolved": "KeyboardEvent | MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        },
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "notificationsClick",
                "name": "notificationsClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the notifications button is clicked or activated via keyboard."
                },
                "complexType": {
                    "original": "MouseEvent | KeyboardEvent",
                    "resolved": "KeyboardEvent | MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        },
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "notificationsMenuOpenChange",
                "name": "notificationsMenuOpenChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the notifications menu open state changes."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "searchChange",
                "name": "searchChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the search input value is changed."
                },
                "complexType": {
                    "original": "{ value: string }",
                    "resolved": "{ value: string; }",
                    "references": {}
                }
            }, {
                "method": "searchClick",
                "name": "searchClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the search button is clicked or activated via keyboard."
                },
                "complexType": {
                    "original": "MouseEvent | KeyboardEvent",
                    "resolved": "KeyboardEvent | MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        },
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "searchInputOpenChange",
                "name": "searchInputOpenChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the search input open state changes."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "signOutClick",
                "name": "signOutClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the user profile sign out button is clicked or activated via keyboard."
                },
                "complexType": {
                    "original": "MouseEvent | KeyboardEvent",
                    "resolved": "KeyboardEvent | MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        },
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "trimbleLogoClick",
                "name": "trimbleLogoClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the logo button is clicked or activated via keyboard,regardless of the `logoName` prop value."
                },
                "complexType": {
                    "original": "MouseEvent | KeyboardEvent",
                    "resolved": "KeyboardEvent | MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        },
                        "KeyboardEvent": {
                            "location": "global",
                            "id": "global::KeyboardEvent"
                        }
                    }
                }
            }, {
                "method": "userMenuOpenChange",
                "name": "userMenuOpenChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the user menu open state changes."
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get listeners() {
        return [{
                "name": "click",
                "method": "handleClickOutside",
                "target": "document",
                "capture": false,
                "passive": false
            }];
    }
}
