// @trimble-oss/moduswebcomponents/collection/components/modus-wc-card/modus-wc-card.js
import { h, Host } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-card.tailwind";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable card component used to group and display content in a way that is easily readable.
 *
 * This component supports multiple `<slot>` elements including 'header' for images or custom content, 'title', 'subtitle', a default slot for main content, 'actions' for buttons or interactive elements, and 'footer'.
 */
export class ModusWcCard {
    constructor() {
        this.inheritedAttributes = {};
        /** Makes any \<figure> in the 'header' slot cover the background */
        this.backgroundFigure = false;
        /** Adds a hard border to the card */
        this.bordered = false;
        /** Custom CSS class to apply */
        this.customClass = '';
        /** Determines how the card is laid out */
        this.layout = 'vertical';
        /** Determines the interior padding size */
        this.padding = 'compact';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-card modus-wc-rounded-card'];
        const propClasses = convertPropsToClasses({
            bordered: this.bordered,
            fullImage: this.backgroundFigure,
            layout: this.layout,
            padding: this.padding,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: 'd737fb3a6d8f49a8686c1a900eab31bddd73144d' }, h("article", Object.assign({ key: '95a5971be0b10ad51c61a2e835bc1846ca00df0c', class: this.getClasses(), tabindex: -1 }, this.inheritedAttributes), h("slot", { key: '9b8d14ff59c833299adf6a14778b0532c984d1ce', name: "header" }), h("div", { key: '73bee21273f06e92ee0a1a8fe5d3288c511e604c', class: "modus-wc-card-body" }, h("div", { key: '4d32753ac4e2728853681af151c412d8dd6da4ab', class: "modus-wc-card-title" }, h("slot", { key: '8ca40ad1df84dc30c26f3ddb9450867707e5066c', name: "title" })), h("div", { key: '8f603221c313bc127a28c434573dd7fbe23f416d', class: "modus-wc-card-subtitle" }, h("slot", { key: 'dc8e78370e0f15959e17b2fe7b6cdf5b5565348e', name: "subtitle" })), h("slot", { key: '5c578250ec7c01c90ea1c01f1ac000dc17269dc3' }), " ", h("div", { key: '6b9fc18a4d81036afe58ad2789464abc400e054c', class: "modus-wc-card-actions" }, h("slot", { key: 'ad0b7b0eed980658c20580355f7973f6fcf9b2fc', name: "actions" }))), h("slot", { key: '13aba4456bb3293471e5d25223e8a1f069b98053', name: "footer" }))));
    }
    static get is() { return "modus-wc-card"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-card.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-card.css"]
        };
    }
    static get properties() {
        return {
            "backgroundFigure": {
                "type": "boolean",
                "attribute": "background-figure",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Makes any \\<figure> in the 'header' slot cover the background"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "bordered": {
                "type": "boolean",
                "attribute": "bordered",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Adds a hard border to the card"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "layout": {
                "type": "string",
                "attribute": "layout",
                "mutable": false,
                "complexType": {
                    "original": "'vertical' | 'horizontal'",
                    "resolved": "\"horizontal\" | \"vertical\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Determines how the card is laid out"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'vertical'"
            },
            "padding": {
                "type": "string",
                "attribute": "padding",
                "mutable": false,
                "complexType": {
                    "original": "'compact' | 'comfortable'",
                    "resolved": "\"comfortable\" | \"compact\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Determines the interior padding size"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'compact'"
            }
        };
    }
    static get elementRef() { return "el"; }
}
