// @trimble-oss/moduswebcomponents/collection/components/modus-wc-badge/modus-wc-badge.js
import { h, Host } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-badge.tailwind";
import { inheritAriaAttributes } from "../utils";
const ALERT_COLORS = ['success', 'warning', 'danger'];
/**
 * A customizable badge component used to create badges with different sizes, types, and colors.
 *
 * The component supports a `<slot>` for injecting content within the badge.
 */
export class ModusWcBadge {
    constructor() {
        this.inheritedAttributes = {};
        /** The color variant of the badge. */
        this.color = 'primary';
        /** Custom CSS class to apply to the span element. */
        this.customClass = '';
        /** The size of the badge. */
        this.size = 'md';
        /** The variant of the badge. */
        this.variant = 'filled';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-badge'];
        const propClasses = convertPropsToClasses({
            color: this.color,
            size: this.size,
            variant: this.variant,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        const isAlert = ALERT_COLORS.includes(this.color);
        // Use inherited role if provided, otherwise use default based on color
        const defaultRole = isAlert ? 'alert' : 'status';
        const role = 'role' in this.inheritedAttributes
            ? this.inheritedAttributes.role
            : defaultRole;
        return (h(Host, { key: '019ea0b40cd01fdcc170edd122ab9a76cf35456a' }, h("span", Object.assign({ key: '068dee65942ca7561310457a40b498e647bd3b4b', class: this.getClasses() }, this.inheritedAttributes, { role: role }), h("slot", { key: 'e6e51e1cfa0181069314bba63c9d8913a804deaf' }))));
    }
    static get is() { return "modus-wc-badge"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-badge.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-badge.css"]
        };
    }
    static get properties() {
        return {
            "color": {
                "type": "string",
                "attribute": "color",
                "mutable": false,
                "complexType": {
                    "original": "| 'primary'\n    | 'secondary'\n    | 'tertiary'\n    | 'high-contrast'\n    | 'success'\n    | 'warning'\n    | 'danger'",
                    "resolved": "\"danger\" | \"high-contrast\" | \"primary\" | \"secondary\" | \"success\" | \"tertiary\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The color variant of the badge."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'primary'"
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the span element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "ModusSize",
                    "resolved": "\"lg\" | \"md\" | \"sm\"",
                    "references": {
                        "ModusSize": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::ModusSize"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The size of the badge."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'md'"
            },
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "'counter' | 'filled' | 'outlined' | 'text'",
                    "resolved": "\"counter\" | \"filled\" | \"outlined\" | \"text\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The variant of the badge."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'filled'"
            }
        };
    }
    static get elementRef() { return "el"; }
}
