// @trimble-oss/moduswebcomponents/collection/components/modus-wc-logo/modus-wc-logo.js
import { h, Host, } from "@stencil/core";
import { inheritAriaAttributes } from "../utils";
import { LOGO_VARIANTS } from "./logo-constants";
import { LOGO_SVGS } from "./logo-svg-data";
const LogoSvg = ({ svgText }) => (h("span", { innerHTML: svgText }));
/**
 * A component for displaying Trimble product logos with support for both fixed and scalable sizing.
 * Provides consistent branding across applications with various product logo options.
 * Logo colors automatically adapt to the active Modus theme via CSS variables.
 */
export class ModusWcLogo {
    constructor() {
        this.inheritedAttributes = {};
        /** Show emblem version (icon only) instead of full logo */
        this.emblem = false;
        /** Custom CSS class to apply to the logo container. */
        this.customClass = '';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getSvgContent() {
        const logoKey = this.name.toLowerCase().replace(/\s+/g, '_');
        const logoInfo = LOGO_VARIANTS[logoKey];
        if (!logoInfo) {
            console.warn(`Logo "${this.name}" not found. Available logos:`, Object.keys(LOGO_VARIANTS).join(', '));
            return '';
        }
        const filePath = this.emblem ? logoInfo.emblemPath : logoInfo.path;
        if (!filePath) {
            console.warn(`No ${this.emblem ? 'emblem' : 'logo'} path found for "${this.name}"`);
            return '';
        }
        const svgContent = LOGO_SVGS[filePath];
        if (!svgContent) {
            console.warn(`SVG content not found for logo "${this.name}" at path "${filePath}".` +
                ' This usually indicates a mismatch between LOGO_VARIANTS and LOGO_SVGS.');
            return '';
        }
        return svgContent;
    }
    getEmblemSvgContent() {
        const logoKey = this.name.toLowerCase().replace(/\s+/g, '_');
        const logoInfo = LOGO_VARIANTS[logoKey];
        if (!this.emblem &&
            logoInfo &&
            logoInfo.emblemPath &&
            logoInfo.category !== 'trimble_brand' &&
            logoInfo.category !== 'viewpoint') {
            return LOGO_SVGS[logoInfo.emblemPath] || '';
        }
        return '';
    }
    getClasses() {
        const classList = [];
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        const altText = this.alt || this.name.replace(/_/g, ' ');
        const classes = this.getClasses();
        const svgContent = this.getSvgContent();
        const emblemSvgContent = this.getEmblemSvgContent();
        const isCombined = !this.emblem && !!emblemSvgContent;
        return (h(Host, { key: '4ca3a31d80602de0af4b035794564eeb8d94f297' }, h("span", Object.assign({ key: '1229795bdc165dfc8778a69d53cb0ee60a9ad8a5', class: `modus-wc-logo ${classes} ${this.emblem ? 'logo-emblem' : 'logo-full'} ${isCombined ? 'logo-combined' : ''}` }, this.inheritedAttributes, { role: "img", "aria-label": altText }), isCombined && (h("span", { key: '6e49f1524a0c0077028f3288b419939568f4cf07', class: "logo-combined-emblem" }, h(LogoSvg, { key: 'c7819d639fa3d0d8f0f05e18e951b7b31f1c5906', svgText: emblemSvgContent }))), h("span", { key: 'b53dd16ef78cd399149de813786b2977abc13679', class: isCombined ? 'logo-combined-wordmark' : '' }, h(LogoSvg, { key: '7441aaed4aee12689b0133981ec0a5c1337ba829', svgText: svgContent })))));
    }
    static get is() { return "modus-wc-logo"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-logo.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-logo.css"]
        };
    }
    static get properties() {
        return {
            "name": {
                "type": "string",
                "attribute": "name",
                "mutable": false,
                "complexType": {
                    "original": "LogoName",
                    "resolved": "\"accubid_anywhere\" | \"accubid_classic\" | \"advanced_image_manager\" | \"analytics\" | \"app_xchange\" | \"appian\" | \"atlas\" | \"autobid\" | \"b2w\" | \"buildable_content\" | \"business_center\" | \"connect\" | \"connect2fab\" | \"copilot\" | \"demand_planning\" | \"earthworks\" | \"estimation_construct\" | \"estimation_desktop\" | \"fabrication\" | \"fabshop\" | \"field_points\" | \"fieldlink\" | \"financials\" | \"geosuite\" | \"infrastructure\" | \"innovative\" | \"livecount\" | \"materials\" | \"modus\" | \"modus_blueprint\" | \"nova\" | \"novapoint\" | \"pay\" | \"pc_miler\" | \"prodesign\" | \"projectsight\" | \"quadri\" | \"reality_capture\" | \"sitevision\" | \"siteworks\" | \"sketchup\" | \"sketchup_for_schools\" | \"sketchup_go\" | \"smart_workflow\" | \"stabicad\" | \"supplier_xchange\" | \"sysque\" | \"tekla\" | \"tekla_structures\" | \"terraflex\" | \"tms\" | \"tmt\" | \"tmw_suite\" | \"trade_service\" | \"trade_service_live\" | \"traqspera\" | \"trimble\" | \"truckmate\" | \"unity\" | \"unity_adms\" | \"unity_construct\" | \"unity_field\" | \"unity_maintain\" | \"unity_permit\" | \"winest\" | \"worksmanager\" | \"viewpoint\" | \"viewpoint_analytics\" | \"viewpoint_epayments\" | \"viewpoint_estimating\" | \"viewpoint_field_management\" | \"viewpoint_field_time\" | \"viewpoint_field_view\" | \"viewpoint_financial_controls\" | \"viewpoint_for_projects\" | \"viewpoint_hr_management\" | \"viewpoint_jobpac_connect\" | \"viewpoint_procontractor\" | \"viewpoint_spectrum\" | \"viewpoint_spectrum_service_tech\" | \"viewpoint_team\" | \"viewpoint_vista\" | \"viewpoint_vista_field_service\"",
                    "references": {
                        "LogoName": {
                            "location": "import",
                            "path": "./logo-constants",
                            "id": "src/components/modus-wc-logo/logo-constants.ts::LogoName"
                        }
                    }
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The name of the logo to display. Accepts values like 'trimble', 'viewpoint_field_view', etc."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "emblem": {
                "type": "boolean",
                "attribute": "emblem",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show emblem version (icon only) instead of full logo"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the logo container."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "alt": {
                "type": "string",
                "attribute": "alt",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The alt text for accessibility. If not provided, defaults to the logo name."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get elementRef() { return "el"; }
}
