// @trimble-oss/moduswebcomponents/collection/components/modus-wc-text-input/modus-wc-text-input.js
import { h, Host, } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-text-input.tailwind";
import { CloseSolidIcon } from "../../icons/close-solid.icon";
import { SearchSolidIcon } from "../../icons/search-solid.icon";
import { handleShadowDOMStyles } from "../base-component";
import { generateElementId, inheritAriaAttributes, inheritAttributes, } from "../utils";
/**
 * A customizable input component used to create text inputs with types.
 *
 * The component supports a `<slot>` for injecting additional custom content inside the input, such as icons or formatted text.
 */
export class ModusWcTextInput {
    constructor() {
        this.inheritedAttributes = {};
        this.generatedId = generateElementId();
        /** Indicates that the input should have a border. */
        this.bordered = true;
        /** Aria label for the clear icon button. */
        this.clearAriaLabel = 'Clear text';
        /** Custom CSS class to apply to the input. */
        this.customClass = '';
        /** Whether the form control is disabled. */
        this.disabled = false;
        /** Show the clear button within the input field. */
        this.includeClear = false;
        /** Show the search icon within the input field. */
        this.includeSearch = false;
        /** Text that appears in the form control when it has no value set. */
        this.placeholder = '';
        /** Whether the value is editable. */
        this.readOnly = false;
        /** A value is required for the form to be submittable. */
        this.required = false;
        /** The size of the input. */
        this.size = 'md';
        /** Type of form control. */
        this.type = 'text';
        /** The value of the control. */
        this.value = '';
        this.handleBlur = (event) => {
            this.inputBlur.emit(event);
        };
        this.handleClearText = (event) => {
            this.value = '';
            this.inputChange.emit(event);
            this.clearClick.emit();
        };
        this.handleFocus = (event) => {
            this.inputFocus.emit(event);
        };
        this.handleInput = (event) => {
            this.value = event.target.value;
            this.inputChange.emit(event);
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = this.placeholder || 'Text input';
        }
        this.inheritedAttributes = Object.assign(Object.assign({}, inheritAriaAttributes(this.el)), inheritAttributes(this.el, ['spellcheck', 'inputmode']));
        if (!this.el.hasAttribute('inputmode') &&
            !this.inheritedAttributes.inputmode) {
            this.el.setAttribute('inputmode', 'text');
        }
    }
    getClasses() {
        const classList = [
            'modus-wc-text-input',
            'modus-wc-input',
            'modus-wc-w-full',
            'modus-wc-flex',
            'modus-wc-items-center',
            'modus-wc-gap-1',
        ];
        const propClasses = convertPropsToClasses({
            bordered: this.bordered,
            disabled: this.disabled,
            feedback: this.feedback,
            readOnly: this.readOnly,
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    shouldIncludeClear() {
        return (!!this.includeClear && !this.disabled && !this.readOnly && !!this.value);
    }
    render() {
        const showClear = this.shouldIncludeClear();
        const effectiveId = this.inputId || this.generatedId;
        const hasCustomIcon = !!this.el.querySelector('[slot="custom-icon"]');
        return (h(Host, { key: '08efb8d8b1802b5563abf3e93278aadf2a360ad6' }, this.label && (h("modus-wc-input-label", { key: '7d41ceb1e42256287c269fd92183e03871fe5fd0', forId: effectiveId, labelText: this.label, required: this.required, size: this.size })), h("label", { key: '1b53a067cca02831dba043f58f5d15590db5b2a6', class: this.getClasses() }, hasCustomIcon ? (h("div", { class: "modus-wc-text-input-icon modus-wc-text-input-icon-custom" }, h("slot", { name: "custom-icon" }))) : (this.includeSearch && (h(SearchSolidIcon, { className: "modus-wc-text-input-icon modus-wc-text-input-icon-search" }))), h("input", Object.assign({ key: '9a09e45373bacf5f3bffe519e2f5dec4699942d3', "aria-required": this.required, autocapitalize: this.autoCapitalize, autocomplete: this.autoComplete, autocorrect: this.autoCorrect, class: "modus-wc-grow", disabled: this.disabled, enterkeyhint: this.enterkeyhint, id: effectiveId, maxlength: this.maxLength, minlength: this.minLength, name: this.name, onBlur: this.handleBlur, onFocus: this.handleFocus, onInput: this.handleInput, pattern: this.pattern, placeholder: this.placeholder, readonly: this.readOnly, required: this.required, tabIndex: this.inputTabIndex, type: this.type, value: this.value }, this.inheritedAttributes)), this.includeClear && (h("div", { key: 'ec7a56940b19292f0f1ab4f97c9926ef1a664a31', class: `modus-wc-clear-icon-container ${showClear ? 'modus-wc-clear-icon-visible' : 'modus-wc-clear-icon-hidden'}` }, h(CloseSolidIcon, { key: 'd1f196d1a9825445dce45e596183189b3219c0c7', ariaLabel: this.clearAriaLabel, className: "modus-wc-text-input-icon modus-wc-text-input-icon-clear", decorative: false, onClear: this.handleClearText })))), this.feedback && (h("modus-wc-input-feedback", { key: '4c15132399331207e0f3dbcf7b451729a027c403', level: this.feedback.level, message: this.feedback.message, size: this.size }))));
    }
    static get is() { return "modus-wc-text-input"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-text-input.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-text-input.css"]
        };
    }
    static get properties() {
        return {
            "autoCapitalize": {
                "type": "string",
                "attribute": "auto-capitalize",
                "mutable": false,
                "complexType": {
                    "original": "| 'off'\n    | 'none'\n    | 'on'\n    | 'sentences'\n    | 'words'\n    | 'characters'",
                    "resolved": "\"characters\" | \"none\" | \"off\" | \"on\" | \"sentences\" | \"words\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Controls automatic capitalization in inputted text."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "autoComplete": {
                "type": "string",
                "attribute": "auto-complete",
                "mutable": false,
                "complexType": {
                    "original": "AutocompleteTypes",
                    "resolved": "AutocompleteTypes | undefined",
                    "references": {
                        "AutocompleteTypes": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::AutocompleteTypes"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Hint for form autofill feature."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "autoCorrect": {
                "type": "string",
                "attribute": "auto-correct",
                "mutable": false,
                "complexType": {
                    "original": "'on' | 'off'",
                    "resolved": "\"off\" | \"on\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Controls automatic correction in inputted text. Support by browser varies."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "bordered": {
                "type": "boolean",
                "attribute": "bordered",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Indicates that the input should have a border."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "true"
            },
            "clearAriaLabel": {
                "type": "string",
                "attribute": "clear-aria-label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Aria label for the clear icon button."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Clear text'"
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Whether the form control is disabled."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "enterkeyhint": {
                "type": "string",
                "attribute": "enterkeyhint",
                "mutable": false,
                "complexType": {
                    "original": "| 'enter'\n    | 'done'\n    | 'go'\n    | 'next'\n    | 'previous'\n    | 'search'\n    | 'send'",
                    "resolved": "\"done\" | \"enter\" | \"go\" | \"next\" | \"previous\" | \"search\" | \"send\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "A hint to the browser for which enter key to display."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "feedback": {
                "type": "unknown",
                "attribute": "feedback",
                "mutable": false,
                "complexType": {
                    "original": "IInputFeedbackProp",
                    "resolved": "IInputFeedbackProp | undefined",
                    "references": {
                        "IInputFeedbackProp": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::IInputFeedbackProp"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Feedback to render below the input."
                },
                "getter": false,
                "setter": false
            },
            "includeClear": {
                "type": "boolean",
                "attribute": "include-clear",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show the clear button within the input field."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "includeSearch": {
                "type": "boolean",
                "attribute": "include-search",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show the search icon within the input field."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "inputId": {
                "type": "string",
                "attribute": "input-id",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The ID of the input element."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "inputTabIndex": {
                "type": "number",
                "attribute": "input-tab-index",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Determine the control's relative ordering for sequential focus navigation (typically with the Tab key)."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The text to display within the label."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "maxLength": {
                "type": "number",
                "attribute": "max-length",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Maximum length (number of characters) of value."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "minLength": {
                "type": "number",
                "attribute": "min-length",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Minimum length (number of characters) of value."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "name": {
                "type": "string",
                "attribute": "name",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Name of the form control. Submitted with the form as part of a name/value pair."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "pattern": {
                "type": "string",
                "attribute": "pattern",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Pattern the value must match to be valid"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "placeholder": {
                "type": "string",
                "attribute": "placeholder",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Text that appears in the form control when it has no value set."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "readOnly": {
                "type": "boolean",
                "attribute": "read-only",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Whether the value is editable."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "required": {
                "type": "boolean",
                "attribute": "required",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "A value is required for the form to be submittable."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "ModusSize",
                    "resolved": "\"lg\" | \"md\" | \"sm\" | undefined",
                    "references": {
                        "ModusSize": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::ModusSize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The size of the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'md'"
            },
            "type": {
                "type": "string",
                "attribute": "type",
                "mutable": false,
                "complexType": {
                    "original": "TextFieldTypes",
                    "resolved": "\"date\" | \"datetime-local\" | \"email\" | \"month\" | \"number\" | \"password\" | \"search\" | \"tel\" | \"text\" | \"time\" | \"url\" | \"week\" | undefined",
                    "references": {
                        "TextFieldTypes": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::TextFieldTypes"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Type of form control."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'text'"
            },
            "value": {
                "type": "string",
                "attribute": "value",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The value of the control."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "''"
            }
        };
    }
    static get events() {
        return [{
                "method": "clearClick",
                "name": "clearClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the clear button is clicked."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "inputBlur",
                "name": "inputBlur",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the input loses focus."
                },
                "complexType": {
                    "original": "FocusEvent",
                    "resolved": "FocusEvent",
                    "references": {
                        "FocusEvent": {
                            "location": "global",
                            "id": "global::FocusEvent"
                        }
                    }
                }
            }, {
                "method": "inputChange",
                "name": "inputChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the input value changes."
                },
                "complexType": {
                    "original": "InputEvent",
                    "resolved": "InputEvent",
                    "references": {
                        "InputEvent": {
                            "location": "global",
                            "id": "global::InputEvent"
                        }
                    }
                }
            }, {
                "method": "inputFocus",
                "name": "inputFocus",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the input gains focus."
                },
                "complexType": {
                    "original": "FocusEvent",
                    "resolved": "FocusEvent",
                    "references": {
                        "FocusEvent": {
                            "location": "global",
                            "id": "global::FocusEvent"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "el"; }
}
