// @trimble-oss/moduswebcomponents/collection/components/modus-wc-collapse/modus-wc-collapse.js
import { Fragment, h, Host, } from "@stencil/core";
import { handleShadowDOMStyles } from "../base-component";
import { convertPropsToClasses, convertPropsToDescriptionDivClasses, convertPropsToTitleChildDivClasses, convertPropsToTitleDivClasses, } from "./modus-wc-collapse.tailwind";
import { generateRandomId, inheritAriaAttributes } from "../utils";
/**
 * A customizable collapse component used for showing and hiding content.
 *
 * The component supports a 'header' and 'content' `<slot>` for injecting custom HTML.
 */
export class ModusWcCollapse {
    constructor() {
        this.inheritedAttributes = {};
        /** Indicates that the component should have a border. */
        this.bordered = false;
        /** Custom CSS class to apply to the outer div. */
        this.customClass = '';
        /** Controls whether the collapse is expanded or not. */
        this.expanded = false;
        this.handleToggle = (event) => {
            const detailsElement = event.currentTarget;
            const isOpen = detailsElement.open;
            if (this.expanded !== isOpen) {
                this.expanded = isOpen;
                this.expandedChange.emit({ expanded: isOpen });
            }
        };
    }
    expandedChanged(newValue) {
        if (this.detailsRef && this.detailsRef.open !== newValue) {
            this.detailsRef.open = newValue;
        }
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        if (!this.collapseId) {
            this.collapseId = generateRandomId();
        }
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getOuterDivClasses() {
        const classList = ['modus-wc-collapse modus-wc-collapse-arrow'];
        const propClasses = convertPropsToClasses({
            bordered: this.bordered,
            expanded: this.expanded,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    // istanbul ignore next
    getTitleClasses() {
        var _a;
        const classList = [
            'modus-wc-collapse-title modus-wc-inline-flex modus-wc-items-center modus-wc-justify-between modus-wc-min-h-4',
        ];
        const paddingClass = convertPropsToTitleDivClasses({
            size: (_a = this.options) === null || _a === void 0 ? void 0 : _a.size,
        });
        if (paddingClass)
            classList.push(paddingClass);
        return classList.join(' ');
    }
    // istanbul ignore next
    getTitleChildDivClasses() {
        var _a;
        const classList = ['modus-wc-inline-flex modus-wc-items-center'];
        const titleFontSize = convertPropsToTitleChildDivClasses({
            size: (_a = this.options) === null || _a === void 0 ? void 0 : _a.size,
        });
        if (titleFontSize)
            classList.push(titleFontSize);
        return classList.join(' ');
    }
    // istanbul ignore next
    getDescriptionDivClasses() {
        var _a;
        const classList = ['description modus-wc-font-light'];
        const descriptionFontSize = convertPropsToDescriptionDivClasses({
            size: (_a = this.options) === null || _a === void 0 ? void 0 : _a.size,
        });
        if (descriptionFontSize)
            classList.push(descriptionFontSize);
        return classList.join(' ');
    }
    render() {
        const baseId = this.collapseId;
        const titleId = `${baseId}-title`;
        const contentId = `${baseId}-content`;
        return (h(Host, { key: 'f096cd2fe3aef17701cdfbe716f91d585a5ced46' }, h("details", Object.assign({ key: 'f54d0fd96bece4c26995dd752c0f06a5519d39f7', class: this.getOuterDivClasses(), open: this.expanded, onToggle: this.handleToggle, ref: (el) => (this.detailsRef = el) }, this.inheritedAttributes), h("summary", { key: '53b5f593601bb992bbbfbad84f4ad93301e95fb7', class: this.getTitleClasses(), id: titleId }, this.options ? (h(Fragment, null, h("div", { class: this.getTitleChildDivClasses() }, this.options.icon && (h("modus-wc-icon", { "aria-label": this.options.iconAriaLabel, decorative: true, name: this.options.icon, size: this.options.size })), this.options.title), this.options.description && (h("div", { class: this.getDescriptionDivClasses() }, this.options.description)))) : (h("slot", { name: "header" }))), h("div", { key: '174f108ec846bf76157b695a516b5b82b0a2b42f', "aria-labelledby": titleId, class: "modus-wc-collapse-content modus-wc-cursor-default", id: contentId }, h("slot", { key: '004a91b9ef8066978f7e255ca2c1b4e7232a6534', name: "content" })))));
    }
    static get is() { return "modus-wc-collapse"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-collapse.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-collapse.css"]
        };
    }
    static get properties() {
        return {
            "bordered": {
                "type": "boolean",
                "attribute": "bordered",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Indicates that the component should have a border."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the outer div."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "expanded": {
                "type": "boolean",
                "attribute": "expanded",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Controls whether the collapse is expanded or not."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "collapseId": {
                "type": "string",
                "attribute": "collapse-id",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "A unique identifier used to set the id attributes of various elements."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "options": {
                "type": "unknown",
                "attribute": "options",
                "mutable": false,
                "complexType": {
                    "original": "ICollapseOptions",
                    "resolved": "ICollapseOptions | undefined",
                    "references": {
                        "ICollapseOptions": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-collapse/modus-wc-collapse.tsx",
                            "id": "src/components/modus-wc-collapse/modus-wc-collapse.tsx::ICollapseOptions"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Configuration options for rendering the pre-laid out collapse component.\nDo not set this prop if you intend to use the 'header' slot."
                },
                "getter": false,
                "setter": false
            }
        };
    }
    static get events() {
        return [{
                "method": "expandedChange",
                "name": "expandedChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the expanded prop is internally changed."
                },
                "complexType": {
                    "original": "{ expanded: boolean }",
                    "resolved": "{ expanded: boolean; }",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "expanded",
                "methodName": "expandedChanged"
            }];
    }
}
