// @trimble-oss/moduswebcomponents/collection/components/modus-wc-date/modus-wc-date.stories.js
import { withActions } from "@storybook/addon-actions/decorator";
import { html } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
import { createShadowHostClass } from "../../providers/shadow-dom/shadow-host-helper";
const meta = {
    title: 'Components/Forms/Date',
    component: 'modus-wc-date',
    args: {
        bordered: true,
        'custom-class': '',
        disabled: false,
        label: 'Label',
        'read-only': false,
        required: false,
        'show-week-numbers': false,
        size: 'md',
        value: '',
        'week-start-day': 'sunday',
    },
    argTypes: {
        feedback: {
            table: {
                type: {
                    detail: `
            Interface: IInputFeedbackProp
            Properties:
            - level ('error' | 'info' | 'success' | 'warning'): The feedback level
            - message (string, optional): The feedback message
          `,
                },
            },
        },
        size: {
            control: { type: 'select' },
            options: ['sm', 'md', 'lg'],
        },
        format: {
            control: { type: 'select' },
            options: [
                undefined,
                'yyyy-mm-dd',
                'dd-mm-yyyy',
                'mm-dd-yyyy',
                'yyyy/mm/dd',
                'dd/mm/yyyy',
                'mm/dd/yyyy',
                'MMM DD, YYYY',
            ],
        },
        'week-start-day': {
            control: { type: 'select' },
            options: [
                'sunday',
                'monday',
                'tuesday',
                'wednesday',
                'thursday',
                'friday',
                'saturday',
            ],
        },
    },
    decorators: [withActions],
    parameters: {
        actions: {
            handles: [
                'inputBlur',
                'inputChange',
                'inputFocus',
                'calendarMonthChange',
                'calendarYearChange',
            ],
        },
    },
};
export default meta;
const Template = {
    render: (args) => {
        return html `
      <style>
        div[id^='story--components-forms-date--default'] {
          min-height: 400px;
          width: 300px;
        }
      </style>
      <modus-wc-date
        aria-label="Date input"
        ?bordered=${args.bordered}
        custom-class=${ifDefined(args['custom-class'])}
        ?disabled=${args.disabled}
        .feedback=${args.feedback}
        format=${ifDefined(args.format)}
        input-id=${ifDefined(args['input-id'])}
        input-tab-index=${ifDefined(args['input-tab-index'])}
        label=${ifDefined(args.label)}
        max=${ifDefined(args.max)}
        min=${ifDefined(args.min)}
        name=${ifDefined(args.name)}
        ?read-only=${args['read-only']}
        ?required=${args.required}
        ?show-week-numbers=${args['show-week-numbers']}
        size=${ifDefined(args.size)}
        .value=${args.value}
        week-start-day=${ifDefined(args['week-start-day'])}
      ></modus-wc-date>
    `;
    },
};
export const Default = Object.assign({}, Template);
const errorFeedback = {
    level: 'error',
    message: 'Value is required.',
};
export const WithErrorFeedback = Object.assign(Object.assign({}, Template), { args: { feedback: errorFeedback, required: true }, parameters: {
        docs: {
            source: {
                transform: (src) => `${src}
<script>
  const dateInputElement = document.querySelector('modus-wc-date');
  dateInputElement.feedback = {
    level: 'error',
    message: 'Value is required.'
  };
</script>`,
            },
        },
    } });
export const ShadowDomParent = {
    render: (args) => {
        // Create a unique shadow host for date component
        if (!customElements.get('date-shadow-host')) {
            const DateShadowHost = createShadowHostClass({
                componentTag: 'modus-wc-date',
                propsMapper: (v, el) => {
                    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
                    const dateEl = el;
                    dateEl.bordered = Boolean(v.bordered);
                    dateEl.customClass = v['custom-class'] || '';
                    dateEl.disabled = Boolean(v.disabled);
                    dateEl.format = v.format;
                    dateEl.inputId = (_a = v['input-id']) !== null && _a !== void 0 ? _a : '';
                    dateEl.inputTabIndex = (_b = v['input-tab-index']) !== null && _b !== void 0 ? _b : -1;
                    dateEl.label = (_c = v.label) !== null && _c !== void 0 ? _c : '';
                    dateEl.max = (_d = v.max) !== null && _d !== void 0 ? _d : '';
                    dateEl.min = (_e = v.min) !== null && _e !== void 0 ? _e : '';
                    dateEl.name = (_f = v.name) !== null && _f !== void 0 ? _f : '';
                    dateEl.readOnly = Boolean(v['read-only']);
                    dateEl.required = Boolean(v.required);
                    dateEl.showWeekNumbers = Boolean(v['show-week-numbers']);
                    dateEl.size = (_g = v.size) !== null && _g !== void 0 ? _g : '';
                    dateEl.value = (_h = v.value) !== null && _h !== void 0 ? _h : '';
                    dateEl.weekStartDay = (_j = v['week-start-day']) !== null && _j !== void 0 ? _j : '';
                },
            });
            customElements.define('date-shadow-host', DateShadowHost);
        }
        return html `<date-shadow-host .props=${Object.assign({}, args)}></date-shadow-host>`;
    },
};
export const Migration = {
    parameters: {
        docs: {
            description: {
                story: `
#### Breaking Changes

  - In 1.0 input state was maintained by the component. 2.0 components encourage users to follow a controlled
  input model. See the Form Inputs [documentation]([Angular](?path=/docs/documentation-form-inputs--docs) for
  additional info and examples.
  - Size values have changed from verbose names (\`medium\`, \`large\`) to abbreviations (\`sm\`, \`md\`, \`lg\`).
  - The \`value\` prop now always outputs **ISO 8601 format** (\`YYYY-MM-DD\`), regardless of the display format.
  Previously, \`value\` matched the display format (e.g. \`dd-mm-yyyy\`).
  - The \`format\` prop is now automatically derived from the user's locale when not explicitly set.
  Previously, it defaulted to \`dd-mm-yyyy\`. The accepted values remain the same fixed union
  (\`'yyyy-mm-dd'\`, \`'dd-mm-yyyy'\`, \`'mm-dd-yyyy'\`, \`'yyyy/mm/dd'\`, \`'dd/mm/yyyy'\`, \`'mm/dd/yyyy'\`, \`'MMM DD, YYYY'\`).

#### Prop Mapping

| 1.0 Prop           | 2.0 Prop         | Notes                                   |
|--------------------|------------------|-----------------------------------------|
| allow-chars-regex  |                  | Not carried over                        |
| alt-formats        |                  | Not carried over                        |
| aria-label         | aria-label       |                                         |
| auto-focus-input   |                  | Not carried over                        |
| disabled           | disabled         |                                         |
| disable-validation |                  | Not carried over                        |
| error-text         | feedback.message | Use \`feedback\` level                  |
| filler-date        |                  | Not carried over                        |
| format             | format           | Auto-derived from locale when not set; union type unchanged |
| helper-text        |                  | Not carried over                        |
| label              | label            |                                         |
| max                | max              |                                         |
| min                | min              |                                         |
| placeholder        |                  | Not carried over                        |
| read-only          | read-only        |                                         |
| required           | required         |                                         |
| show-calendar-icon |                  | Not carried over                        |
| size               | size             | \`medium\` → \`md\`, \`large\` → \`lg\` |
| type               |                  | Not carried over                        |
| valid-text         | feedback.message | Use \`feedback\` level                  |
| value              | value            | Now outputs ISO 8601 (\`YYYY-MM-DD\`)   |

#### Event Mapping

| 1.0 Event           | 2.0 Event   | Notes            |
|---------------------|-------------|------------------|
| calendarIconClicked |             | Not carried over |
| dateInputBlur       | inputBlur   |                  |
| valueChange         | inputChange |                  |
| valueError          |             | Not carried over |
        `,
            },
        },
        controls: { disable: true },
        canvas: { disable: true },
    },
    render: () => html `<div></div>`,
};
