// @trimble-oss/moduswebcomponents/cjs/modus-wc-panel_2.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var utils = require('./utils-DiibENGv.js');

const convertPropsToClasses$1 = ({ floating, }) => {
    let classes = '';
    if (floating) {
        classes = `${classes} modus-wc-panel-floating`;
    }
    return classes.trim();
};

const modusWcPanelCss = ".modus-wc-panel{background-color:var(--modus-wc-color-base-page);border:1px solid var(--modus-wc-color-base-200);border-radius:var(--rounded-box);display:flex;flex-direction:column;overflow:hidden}.modus-wc-panel-floating{box-shadow:rgba(36, 35, 45, 0.3) 1px 0 4px}.modus-wc-panel-header,.modus-wc-panel-body,.modus-wc-panel-footer{display:flex;flex-direction:column;width:100%}.modus-wc-panel-header{flex-shrink:0}.modus-wc-panel-header:empty{display:none}.modus-wc-panel-body{flex:1 1 auto;min-height:0;overflow-y:auto}.modus-wc-panel-footer{flex-shrink:0}.modus-wc-panel-footer:empty{display:none}";

const ModusWcPanel = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the outer div. */
        this.customClass = '';
        /** Width of the panel in pixels. */
        this.width = '350px';
        /** Height of the panel in pixels. */
        this.height = '700px';
        /** Enable floating mode with elevated shadow. */
        this.floating = false;
    }
    componentWillLoad() {
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-panel'];
        const propClasses = convertPropsToClasses$1({
            floating: this.floating,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (index.h(index.Host, { key: '8eca46effdba5d4c7c027a78d98e8914247d0187' }, index.h("div", Object.assign({ key: '106cf95c53ce95e4afa6a58845b979d575b3b318', class: this.getClasses(), style: { width: this.width, height: this.height } }, this.inheritedAttributes), index.h("div", { key: '200d9a028744c3027fea736461d50c27eca1ad31', class: "modus-wc-panel-header" }, index.h("slot", { key: '6ef2682a8d4eeb0f253ac6602da9a980b026be84', name: "header" })), index.h("div", { key: '61b44ce89e87e1364a18fb366f464ee25881a5c5', class: "modus-wc-panel-body" }, index.h("slot", { key: '4c6c82a0ad09e81b74dbe9c478f5793eaac702ee', name: "body" })), index.h("div", { key: '2806ffb3225c9c361b55b12e9088cb4326cbf463', class: "modus-wc-panel-footer" }, index.h("slot", { key: '577fadfaeb76f7c2f475685756b7da6dbb5c05fb', name: "footer" })))));
    }
    get el() { return index.getElement(this); }
};
ModusWcPanel.style = modusWcPanelCss;

const convertPropsToClasses = ({ size, weight, }) => {
    let classes = '';
    if (size) {
        classes = `${classes} modus-wc-text-${size}`;
    }
    if (weight) {
        classes = `${classes} modus-wc-typography-weight-${weight}`;
    }
    return classes.trim();
};

const modusWcTypographyCss = "modus-wc-typography h1.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-3xl);font-weight:var(--modus-wc-font-weight-normal);letter-spacing:0.5px;line-height:36px}modus-wc-typography h2.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-2xl);font-weight:var(--modus-wc-font-weight-normal);letter-spacing:0.15px;line-height:30px}modus-wc-typography h3.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-xl);font-weight:var(--modus-wc-font-weight-semibold);letter-spacing:0.15px;line-height:27px}modus-wc-typography h4.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-lg);font-weight:var(--modus-wc-font-weight-semibold);letter-spacing:0.15px;line-height:24px}modus-wc-typography h5.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-md);font-weight:var(--modus-wc-font-weight-bold);letter-spacing:0.45px;line-height:24px}modus-wc-typography h6.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-sm);font-weight:var(--modus-wc-font-weight-bold);letter-spacing:0.45px;line-height:18px}modus-wc-typography .modus-wc-typography-weight-light{font-weight:var(--modus-wc-font-weight-light)}modus-wc-typography .modus-wc-typography-weight-normal{font-weight:var(--modus-wc-font-weight-normal)}modus-wc-typography .modus-wc-typography-weight-semibold{font-weight:var(--modus-wc-font-weight-semibold)}modus-wc-typography .modus-wc-typography-weight-bold{font-weight:var(--modus-wc-font-weight-bold)}modus-wc-typography .modus-wc-text-xs{font-size:var(--modus-wc-font-size-xs)}modus-wc-typography .modus-wc-text-sm{font-size:var(--modus-wc-font-size-sm)}modus-wc-typography .modus-wc-text-md{font-size:var(--modus-wc-font-size-md)}modus-wc-typography .modus-wc-text-lg{font-size:var(--modus-wc-font-size-lg)}modus-wc-typography .modus-wc-text-xl{font-size:var(--modus-wc-font-size-xl)}modus-wc-typography .modus-wc-text-2xl{font-size:var(--modus-wc-font-size-2xl)}modus-wc-typography .modus-wc-text-3xl{font-size:var(--modus-wc-font-size-3xl)}[data-theme=connect-light] modus-wc-typography :not(a),[data-theme=connect-dark] modus-wc-typography :not(a){color:var(--modus-wc-color-base-content-low-contrast)}[data-theme=connect-light] modus-wc-typography h1.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h1.modus-wc-typography{font-size:var(--modus-wc-font-size-4xl);font-weight:var(--modus-wc-font-weight-normal);letter-spacing:0.5px;line-height:30px}[data-theme=connect-light] modus-wc-typography h2.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h2.modus-wc-typography{font-size:var(--modus-wc-font-size-3xl);font-weight:var(--modus-wc-font-weight-semibold);letter-spacing:0.15px;line-height:24px}[data-theme=connect-light] modus-wc-typography h3.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h3.modus-wc-typography{font-size:var(--modus-wc-font-size-xl);font-weight:var(--modus-wc-font-weight-semibold);letter-spacing:0.15px;line-height:18px}[data-theme=connect-light] modus-wc-typography h4.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h4.modus-wc-typography{font-size:var(--modus-wc-font-size-lg);letter-spacing:0.15px;line-height:24px}[data-theme=connect-light] modus-wc-typography h5.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h5.modus-wc-typography{font-size:var(--modus-wc-font-size-md);font-weight:var(--modus-wc-font-weight-bold);letter-spacing:0.45px;line-height:24px}[data-theme=connect-light] modus-wc-typography h6.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h6.modus-wc-typography{font-size:var(--modus-wc-font-size-xs);font-weight:var(--modus-wc-font-weight-bold);letter-spacing:0.45px;line-height:18px;text-transform:uppercase}";

const ModusWCTypography = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the typography element. */
        this.customClass = '';
        /** The hierarchy of the typography component. */
        this.hierarchy = 'p';
        /** The size of the font. */
        this.size = 'md';
        /** The weight of the text. */
        this.weight = 'normal';
    }
    componentWillLoad() {
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-typography'];
        // Check if we're dealing with a heading and have size/weight overrides
        const isHeading = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].includes(this.hierarchy);
        const hasOverrides = this.size !== 'md' || this.weight !== 'normal';
        if (isHeading && hasOverrides) {
            // Add a class to indicate overrides for headings
            classList.push('modus-wc-typography-override');
        }
        const propClasses = convertPropsToClasses({
            size: this.size,
            weight: this.weight,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        const Element = this.hierarchy;
        return (index.h(index.Host, { key: 'b9f340f7348ccf18ae168020cccabe14a1a8c5de' }, index.h(Element, Object.assign({ key: '6bf43d72c246ac1d38b241f8596d953dbed044a0', class: this.getClasses() }, this.inheritedAttributes), index.h("slot", { key: '7f59a5319bbf237101acaa67065bf5bec3565c4e' }, this.label))));
    }
    get el() { return index.getElement(this); }
};
ModusWCTypography.style = modusWcTypographyCss;

exports.modus_wc_panel = ModusWcPanel;
exports.modus_wc_typography = ModusWCTypography;
