// @trimble-oss/moduswebcomponents/cjs/modus-wc-slider.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var baseComponent = require('./base-component-ZxJWB8zi.js');
var utils = require('./utils-DiibENGv.js');

const convertPropsToClasses = ({ size, }) => {
    let classes = '';
    if (size) {
        classes = `${classes} modus-wc-range-${size}`;
    }
    return classes.trim();
};

const modusWcSliderCss = "modus-wc-slider .modus-wc-range{--rounded-box:200px;--range-shadow:var(--modus-wc-color-trimble-blue)}modus-wc-slider .modus-wc-range .modus-wc-input-label{padding-bottom:var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-slider .modus-wc-range,[data-theme=modus-classic-dark] modus-wc-slider .modus-wc-range{--rounded-box:200px;--range-shadow:var(--modus-wc-color-trimble-blue)}";

const ModusWcSlider = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inputBlur = index.createEvent(this, "inputBlur");
        this.inputChange = index.createEvent(this, "inputChange");
        this.inputFocus = index.createEvent(this, "inputFocus");
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The disabled state of the slider. */
        this.disabled = false;
        /** Name of the form control. Submitted with the form as part of a name/value pair. */
        this.name = '';
        /** A value is required for the form to be submittable. */
        this.required = false;
        /** The size of the input. */
        this.size = 'md';
        /** The value of the slider. */
        this.value = 0;
        this.handleBlur = (event) => {
            this.inputBlur.emit(event);
        };
        this.handleFocus = (event) => {
            this.inputFocus.emit(event);
        };
        this.handleInput = (event) => {
            this.inputChange.emit(event);
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        baseComponent.handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Slider';
        }
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-range'];
        const propClasses = convertPropsToClasses({ size: this.size });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (index.h(index.Host, { key: 'fcf03fb6f845e045fc118b4c4b9b7fd9678d1dda' }, this.label && (index.h("modus-wc-input-label", { key: '4da0e14edc26d12e65b52964615e320b6059bee5', forId: this.inputId, labelText: this.label, required: this.required, size: this.size })), index.h("input", Object.assign({ key: '7ef8bd9690f4ed2352015c8038ff48bfd47a7673', "aria-disabled": this.disabled, class: this.getClasses(), disabled: this.disabled, id: this.inputId, max: this.max, min: this.min, name: this.name, onBlur: this.handleBlur, onFocus: this.handleFocus, onInput: this.handleInput, required: this.required, step: this.step, tabIndex: this.inputTabIndex, type: "range", value: this.value }, this.inheritedAttributes))));
    }
    get el() { return index.getElement(this); }
};
ModusWcSlider.style = modusWcSliderCss;

exports.modus_wc_slider = ModusWcSlider;
