// @trimble-oss/moduswebcomponents/cjs/modus-wc-avatar_2.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var utils = require('./utils-DiibENGv.js');
var baseComponent = require('./base-component-ZxJWB8zi.js');

const convertPropsToClasses$1 = ({ shape, size, }) => {
    let classes = '';
    if (shape) {
        switch (shape) {
            case 'circle':
                classes = `${classes} modus-wc-rounded-full`;
                break;
            case 'square':
                classes = `${classes} modus-wc-rounded-lg`;
                break;
        }
    }
    if (size) {
        switch (size) {
            case 'xs':
                classes = `${classes} modus-wc-w-8`;
                break;
            case 'sm':
                classes = `${classes} modus-wc-w-12`;
                break;
            case 'md':
                classes = `${classes} modus-wc-w-16`;
                break;
            case 'lg':
                classes = `${classes} modus-wc-w-20`;
                break;
        }
    }
    return classes.trim();
};

const modusWcAvatarCss = "modus-wc-avatar{display:inline-flex}modus-wc-avatar .modus-wc-avatar .initials{align-items:center;background-color:var(--modus-wc-color-neutral);color:var(--modus-wc-color-neutral-content);display:flex;font-weight:600;height:100%;justify-content:center;text-transform:uppercase;width:100%}modus-wc-avatar .modus-wc-avatar .no-image{align-items:center;background-color:var(--modus-wc-color-base-200);color:var(--modus-wc-color-base-content);display:flex;justify-content:center}modus-wc-avatar .modus-wc-avatar .modus-wc-w-8 .initials{font-size:var(--modus-wc-font-size-xs, 0.75rem)}modus-wc-avatar .modus-wc-avatar .modus-wc-w-12 .initials{font-size:var(--modus-wc-font-size-sm, 0.875rem)}modus-wc-avatar .modus-wc-avatar .modus-wc-w-16 .initials{font-size:var(--modus-wc-font-size-md, 1rem)}modus-wc-avatar .modus-wc-avatar .modus-wc-w-20 .initials{font-size:var(--modus-wc-font-size-lg, 1.25rem)}";

const ModusWcAvatar = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The location of the image. */
        this.imgSrc = '';
        /** The initials to display when no image is provided. */
        this.initials = '';
        // TODO - add placeholder support (need UX logic)
        /** The shape of the avatar. */
        this.shape = 'circle';
        /** The size of the avatar. */
        this.size = 'md';
    }
    componentWillLoad() {
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = [];
        const propClasses = convertPropsToClasses$1({
            shape: this.shape,
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        if (!this.imgSrc && !this.initials)
            classList.push('no-image');
        return classList.join(' ');
    }
    getUserInitials() {
        if (!this.initials)
            return '';
        return this.initials
            .trim()
            .split(/\s+/)
            .filter(Boolean)
            .map((part) => part.charAt(0))
            .join('')
            .substring(0, 3)
            .toUpperCase();
    }
    render() {
        const altText = this.alt || 'User avatar';
        return (index.h(index.Host, { key: '8ac52db552b5093476ae19f158819a3c9d253ab3' }, index.h("div", Object.assign({ key: 'e3452c9c37a8adad4aaae454112f4020b7bf4f83', class: "modus-wc-avatar" }, this.inheritedAttributes), index.h("div", { key: 'eba4c00724a6f4dff7636d4a43fa46e93a28f75f', class: this.getClasses() }, this.imgSrc ? (index.h("img", { src: this.imgSrc, alt: altText })) : this.initials ? (index.h("span", { class: "initials", "aria-label": this.alt || this.initials }, this.getUserInitials())) : (index.h("modus-wc-icon", { "aria-label": altText, name: "person", size: this.size, variant: "solid" }))))));
    }
    get el() { return index.getElement(this); }
};
ModusWcAvatar.style = modusWcAvatarCss;

const convertPropsToClasses = ({ active, bordered, disabled, focused, size, }) => {
    let classes = '';
    if (bordered) {
        classes = `${classes} modus-wc-menu-item-bordered`;
    }
    if (disabled) {
        classes = `${classes} modus-wc-menu-item-disabled`;
    }
    if (active) {
        classes = `${classes} modus-wc-menu-item-active`;
    }
    if (focused) {
        classes = `${classes} modus-wc-menu-item-focused`;
    }
    if (size) {
        classes = `${classes} modus-wc-menu-item-${size}`;
    }
    return classes.trim();
};

const modusWcMenuItemCss = "modus-wc-menu-item .modus-wc-menu-item button{align-items:center;border-radius:inherit;display:flex;width:100%}modus-wc-menu-item .modus-wc-menu-item button:hover{background-color:transparent}modus-wc-menu-item .modus-wc-menu-item button:focus{background-color:transparent}modus-wc-menu-item .modus-wc-menu-item button.modus-wc-menu-dropdown-toggle::after{color:var(--modus-wc-color-base-content)}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content{align-items:center;display:flex;width:100%}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content [slot=start-icon]{padding-inline-end:var(--modus-wc-spacing-sm)}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content modus-wc-checkbox{padding-inline-end:var(--modus-wc-spacing-sm)}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content .modus-wc-menu-item-labels{padding-inline-end:var(--modus-wc-spacing-xs);padding-inline-start:var(--modus-wc-spacing-md);white-space:nowrap}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content .modus-wc-menu-item-sublabel{font-size:var(--modus-wc-font-size-sm);white-space:nowrap}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content .modus-wc-menu-item-selected-icon{margin-left:auto}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-sm button{font-size:var(--modus-wc-font-size-sm)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-sm button .modus-wc-menu-item-sublabel{font-size:var(--modus-wc-font-size-xs)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-md button{font-size:var(--modus-wc-font-size-md)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-md button .modus-wc-menu-item-sublabel{font-size:var(--modus-wc-font-size-sm)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-lg button{font-size:var(--modus-wc-font-size-lg)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-lg button .modus-wc-menu-item-sublabel{font-size:var(--modus-wc-font-size-md)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-bordered{border-bottom:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-base-200);border-radius:0;border-top:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-base-200)}modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-selected){color:var(--modus-wc-color-base-content)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected{background-color:color-mix(in sRGB, var(--modus-wc-color-primary-pale) 50%, transparent);border-radius:inherit;color:var(--modus-wc-base-content)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected:focus{background-color:color-mix(in sRGB, var(--modus-wc-color-primary-pale) 50%, transparent)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected button{background-color:transparent}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected button:hover,modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected button:focus{background-color:transparent;border-radius:inherit;box-shadow:none;color:inherit}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active{background-color:color-mix(in sRGB, var(--modus-wc-color-primary-pale) 50%, transparent);border-radius:inherit;color:var(--modus-wc-color-base-content)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active:focus{background-color:color-mix(in sRGB, var(--modus-wc-color-primary-pale) 50%, transparent)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button{background-color:transparent}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:hover,modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:focus{background-color:transparent;border-radius:inherit;box-shadow:none;color:inherit}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-disabled{opacity:0.4}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-disabled button{pointer-events:none}modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):hover{background-color:var(--modus-wc-color-base-100);outline:none}modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):focus:not(:focus-visible),modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)).modus-wc-arrow-focused{background-color:var(--modus-wc-color-base-100);color:var(--modus-wc-color-primary);outline:none}modus-wc-menu-item .modus-wc-menu-item:focus-visible{background-color:transparent;outline:2px solid var(--modus-wc-color-primary);outline-offset:-2px}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active),[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active){color:var(--modus-wc-color-gray-light)}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item button.modus-wc-menu-dropdown-toggle::after,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item button.modus-wc-menu-dropdown-toggle::after{color:var(--modus-wc-color-gray-3)}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):hover,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):hover{background-color:var(--modus-wc-color-gray-9);color:inherit;outline:none}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)),[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)){background-color:color-mix(in sRGB, var(--modus-wc-color-primary) 30%, transparent)}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button{background-color:transparent}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:hover,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:focus,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button:hover,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button:focus,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:hover,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:focus,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button:hover,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button:focus{background-color:transparent;box-shadow:none;color:inherit}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):focus:not(:focus-visible),[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)).modus-wc-arrow-focused,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):focus:not(:focus-visible),[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)).modus-wc-arrow-focused{background-color:var(--modus-wc-color-gray-9)}[data-theme=connect-light] modus-wc-menu-item .modus-wc-menu-item>button.modus-wc-menu-dropdown-toggle::after,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item>button.modus-wc-menu-dropdown-toggle::after{color:var(--modus-wc-color-white)}";

const ModusWcMenuItem = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.itemSelect = index.createEvent(this, "itemSelect");
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the li element. */
        this.customClass = '';
        /** The text rendered in the menu item. */
        this.label = '';
        /** The size of the menu item. */
        this.size = 'md';
        /** The position of the tooltip relative to the menu item. */
        this.tooltipPosition = 'auto';
        /** The unique identifying value of the menu item. */
        this.value = '';
        /** Internal state to track if submenu is expanded */
        this.isExpanded = false;
        this.handleItemSelect = () => {
            if (this.hasSubmenu) {
                // Check if side nav is expanded (if this menu is inside a side nav)
                const sideNav = this.el.closest('modus-wc-side-navigation');
                if (sideNav &&
                    !sideNav.expanded) {
                    // Don't allow submenu expansion when side nav is collapsed
                    // Still emit the event for consistency
                    this.itemSelect.emit({ value: this.value });
                    return;
                }
                // The submenu should be inside this menu-item element (slotted content)
                const submenu = this.el.querySelector('.modus-wc-menu-dropdown');
                const liElement = this.el.querySelector('li');
                if (submenu && liElement) {
                    submenu.classList.toggle('modus-wc-menu-dropdown-show');
                    const buttonElement = liElement.querySelector('button');
                    // Update internal expanded state and add/remove class
                    this.isExpanded = submenu.classList.contains('modus-wc-menu-dropdown-show');
                    if (this.isExpanded) {
                        liElement.classList.add('modus-wc-menu-item-expanded');
                        if (buttonElement) {
                            buttonElement.classList.add('modus-wc-menu-dropdown-show');
                        }
                    }
                    else {
                        liElement.classList.remove('modus-wc-menu-item-expanded');
                        if (buttonElement) {
                            buttonElement.classList.remove('modus-wc-menu-dropdown-show');
                        }
                    }
                }
            }
            else if (this.resolveSelectionMode() === 'multiple' || this.checkbox) {
                this.selected = !this.selected;
            }
            else {
                if (this.resolveSelectionMode() === 'single') {
                    this.deselectSiblings();
                }
                this.selected = true;
            }
            this.itemSelect.emit({ value: this.value, selected: this.selected });
        };
    }
    componentWillLoad() {
        baseComponent.handleShadowDOMStyles(this.el);
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
        this._selectionMode = this.resolveSelectionMode();
    }
    componentDidLoad() {
        if (typeof MutationObserver === 'undefined')
            return;
        const parentMenu = this.el.closest('modus-wc-menu');
        if (parentMenu) {
            this.parentMenuObserver = new MutationObserver((mutations) => {
                for (const mutation of mutations) {
                    if (mutation.attributeName === 'selection-mode') {
                        this.handleSelectionModeChange();
                        break;
                    }
                }
            });
            this.parentMenuObserver.observe(parentMenu, { attributes: true });
        }
    }
    disconnectedCallback() {
        var _a;
        (_a = this.parentMenuObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    }
    handleSelectionModeChange() {
        this._selectionMode = this.resolveSelectionMode();
        this.selected = false;
    }
    handleKeyDown(e) {
        if (this.disabled)
            return;
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            this.handleItemSelect();
        }
    }
    /**
     * Public method to collapse the submenu if it's expanded
     */
    async collapseSubmenu() {
        if (this.hasSubmenu && this.isExpanded) {
            const submenu = this.el.querySelector('.modus-wc-menu-dropdown');
            const liElement = this.el.querySelector('li');
            if (submenu && liElement) {
                submenu.classList.remove('modus-wc-menu-dropdown-show');
                liElement.classList.remove('modus-wc-menu-item-expanded');
                this.isExpanded = false;
            }
        }
        return Promise.resolve();
    }
    getRootMenu() {
        var _a, _b;
        let menu = this.el.closest('modus-wc-menu');
        while (menu) {
            const parent = (_b = (_a = menu.parentElement) === null || _a === void 0 ? void 0 : _a.closest('modus-wc-menu')) !== null && _b !== void 0 ? _b : null;
            if (!parent)
                break;
            menu = parent;
        }
        return menu;
    }
    deselectSiblings() {
        const rootMenu = this.getRootMenu();
        if (!rootMenu)
            return;
        const allItems = rootMenu.querySelectorAll('modus-wc-menu-item');
        allItems.forEach((item) => {
            if (item !== this.el) {
                item.selected = false;
            }
        });
    }
    resolveSelectionMode() {
        var _a;
        return (_a = this.el.closest('modus-wc-menu')) === null || _a === void 0 ? void 0 : _a.selectionMode;
    }
    getClasses() {
        const classList = ['modus-wc-menu-item'];
        const isActive = !this.hasSubmenu && !!this.selected;
        const propClasses = convertPropsToClasses({
            active: isActive,
            bordered: this.bordered,
            disabled: this.disabled,
            focused: this.focused,
            size: this.size,
        });
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getButtonClasses() {
        return this.hasSubmenu ? 'modus-wc-menu-dropdown-toggle' : '';
    }
    getRole(mode) {
        if (mode === 'multiple')
            return 'menuitemcheckbox';
        if (mode === 'single')
            return 'menuitemradio';
        return 'menuitem';
    }
    hasCheckbox(mode) {
        return this.checkbox || mode === 'multiple';
    }
    getAriaChecked(mode) {
        if (!this.hasCheckbox(mode))
            return undefined;
        return this.selected ? 'true' : 'false';
    }
    getAriaSelected(mode) {
        if (this.hasCheckbox(mode))
            return undefined;
        if (mode === 'single')
            return this.selected ? 'true' : 'false';
        return undefined;
    }
    render() {
        const mode = this._selectionMode;
        return (index.h(index.Host, { key: 'cf1f2a9d42ca8ef8b74b3b84dbaff8dbb9e3fcbf' }, index.h("li", Object.assign({ key: 'c0304724c03f8508163b3680afae3660aee76ee2', "aria-checked": this.getAriaChecked(mode), "aria-selected": this.getAriaSelected(mode), "aria-disabled": this.disabled, class: this.getClasses(), role: this.getRole(mode), tabIndex: this.disabled ? -1 : 0 }, this.inheritedAttributes), index.h("button", { key: '1e198d92292416964ac46409ba6d088d7b1639a3', class: this.getButtonClasses(), disabled: this.disabled, onClick: this.handleItemSelect, tabIndex: -1, type: "button" }, index.h("div", { key: '28473bf3af4c1de33f00b35a0fcda58c74be90c0', class: "modus-wc-menu-item-content" }, (this.checkbox || mode === 'multiple') && (index.h("modus-wc-checkbox", { key: 'f1f2bed203505f00ed9d055e012e7c73e3bf75d1', "aria-label": "Checkbox", disabled: this.disabled, size: this.size, value: !!this.selected })), index.h("slot", { key: '6f1c7f217c299288f985babe0988e88dec7653c1', name: "start-icon" }), index.h("div", { key: '7d9f55486d9db29bc1f2a2ce5ac21ae281a5a6b8', class: "modus-wc-menu-item-labels" }, this.tooltipContent ? (index.h("modus-wc-tooltip", { content: this.tooltipContent, position: this.tooltipPosition, customClass: "modus-wc-menu-item-tooltip" }, index.h("div", null, this.label))) : (index.h("div", null, this.label)), this.subLabel && (index.h("div", { key: '694c004f591a9fd459ca15ef1a519cd8313a42b8', class: "modus-wc-menu-item-sublabel" }, this.subLabel))))), index.h("slot", { key: '53e15267c8984d5d30ae75d566021b6881c5ea98' }))));
    }
    get el() { return index.getElement(this); }
};
ModusWcMenuItem.style = modusWcMenuItemCss;

exports.modus_wc_avatar = ModusWcAvatar;
exports.modus_wc_menu_item = ModusWcMenuItem;
