// @trimble-oss/moduswebcomponents/cjs/modus-wc-time-input.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var baseComponent = require('./base-component-ZxJWB8zi.js');
var utils = require('./utils-DiibENGv.js');

const convertPropsToClasses = ({ bordered, feedback, readOnly, size, }) => {
    let classes = '';
    if (bordered) {
        classes = `${classes} modus-wc-input-bordered`;
    }
    if (feedback) {
        classes = `${classes} modus-wc-input--${feedback.level}`;
    }
    if (readOnly) {
        classes = `${classes} modus-wc-time-input--readonly`;
    }
    if (size) {
        classes = `${classes} modus-wc-input-${size}`;
    }
    return classes.trim();
};

const modusWcTimeInputCss = ".modus-wc-input--error{border-color:var(--modus-wc-color-error) !important}.modus-wc-input--info{border-color:var(--modus-wc-color-info) !important}.modus-wc-input--success{border-color:var(--modus-wc-color-success) !important}.modus-wc-input--warning{border-color:var(--modus-wc-color-warning) !important}.modus-wc-input-xs{height:var(--modus-wc-size-xs);min-height:var(--modus-wc-size-xs)}.modus-wc-input-sm{height:var(--modus-wc-size-sm);min-height:var(--modus-wc-size-sm)}.modus-wc-input-md{height:var(--modus-wc-size-md);min-height:var(--modus-wc-size-md)}.modus-wc-input-lg{height:var(--modus-wc-size-lg);min-height:var(--modus-wc-size-lg)}.modus-wc-input-xl{height:var(--modus-wc-size-xl);min-height:var(--modus-wc-size-xl)}modus-wc-time-input .modus-wc-input-label{padding-bottom:var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-time-input .modus-wc-time-input.modus-wc-input,[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input{border-radius:var(--modus-wc-border-radius-md)}[data-theme=modus-classic-light] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-input-sm,[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-input-sm{font-size:var(--modus-wc-font-size-sm);height:var(--modus-wc-input-height-sm);padding:var(--modus-wc-spacing-sm) var(--modus-wc-spacing-xs)}[data-theme=modus-classic-light] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-input-md,[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-input-md{font-size:var(--modus-wc-font-size-md);height:var(--modus-wc-input-height-md);padding:var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-input-lg,[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-input-lg{font-size:var(--modus-wc-font-size-lg);height:var(--modus-wc-input-height-lg);padding:var(--modus-wc-spacing-md) var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-time-input .modus-wc-time-input.modus-wc-input:focus,[data-theme=modus-classic-light] modus-wc-time-input .modus-wc-time-input.modus-wc-input:focus-within,[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input:focus,[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input:focus-within{border-color:var(--modus-wc-color-blue-light);border-width:var(--modus-wc-border-width-sm);box-shadow:none;outline:none}[data-theme=modus-classic-light] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-time-input--readonly,[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-time-input--readonly{background-color:var(--modus-wc-color-base-100)}[data-theme=modus-classic-light] modus-wc-time-input .modus-wc-time-input.modus-wc-input.modus-wc-input-bordered:not(:disabled):not(:focus){border-color:var(--modus-wc-color-gray-6)}[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input{color-scheme:dark}[data-theme=modus-classic-dark] modus-wc-time-input .modus-wc-time-input.modus-wc-input:focus{border-color:var(--modus-wc-color-highlight-blue)}[data-theme=connect-light] modus-wc-time-input .modus-wc-time-input,[data-theme=connect-dark] modus-wc-time-input .modus-wc-time-input{border-bottom-width:var(--input-bottom-border-width);outline-width:0 !important}[data-theme=connect-light] modus-wc-time-input .modus-wc-time-input:not(.modus-wc-select,.modus-wc-number-input),[data-theme=connect-dark] modus-wc-time-input .modus-wc-time-input:not(.modus-wc-select,.modus-wc-number-input){padding:0 var(--modus-wc-spacing-sm)}[data-theme=connect-light] modus-wc-time-input .modus-wc-time-input:hover,[data-theme=connect-dark] modus-wc-time-input .modus-wc-time-input:hover{border-bottom-color:var(--modus-wc-color-primary)}[data-theme=connect-light] modus-wc-time-input .modus-wc-time-input:active,[data-theme=connect-dark] modus-wc-time-input .modus-wc-time-input:active{border-bottom-color:var(--modus-wc-color-primary)}[data-theme=connect-light] modus-wc-time-input .modus-wc-time-input:focus,[data-theme=connect-dark] modus-wc-time-input .modus-wc-time-input:focus{border-bottom-color:var(--modus-wc-color-primary);outline:none}[data-theme=connect-light] modus-wc-time-input .modus-wc-time-input:focus-within,[data-theme=connect-dark] modus-wc-time-input .modus-wc-time-input:focus-within{border-bottom-color:var(--modus-wc-color-primary);outline:none}";

const ModusWcTimeInput = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inputBlur = index.createEvent(this, "inputBlur");
        this.inputChange = index.createEvent(this, "inputChange");
        this.inputFocus = index.createEvent(this, "inputFocus");
        this.inheritedAttributes = {};
        /** Indicates that the input should have a border. */
        this.bordered = true;
        /** Custom CSS class to apply to the input. */
        this.customClass = '';
        /** The options to display in the time input dropdown. Options must be in `HH:mm` or `HH:mm:ss` format. */
        this.datalistOptions = [];
        /** Whether the form control is disabled. */
        this.disabled = false;
        /** Whether the value is editable. */
        this.readOnly = false;
        /** A value is required for the form to be submittable. */
        this.required = false;
        /**
         * Displays the time input format as `HH:mm:ss` if `true`.
         * Internally sets the `step` to 1 second.
         * If a `step` value is provided, it will override this attribute.
         */
        this.showSeconds = false;
        /** The size of the input. */
        this.size = 'md';
        /**
         * The value of the time input.
         * Always in 24-hour format that includes leading zeros:
         * `HH:mm` or `HH:mm:ss`, regardless of input format which is likely
         * to be selected based on user's locale (or by the user agent).
         * If time includes seconds the format is always `HH:mm:ss`.
         */
        this.value = '';
        this.handleBlur = (event) => {
            this.inputBlur.emit(event);
        };
        this.handleFocus = (event) => {
            this.inputFocus.emit(event);
        };
        this.handleInput = (event) => {
            this.inputChange.emit(event);
        };
        /*
         * The ID of the internal <datalist> element. Unique to each instance of the time input component.
         * This is used as the `datalistId` id when `datalistOptions` are provided.
         */
        this.internalDatalistId = `modus-wc-datalist-${utils.generateRandomId(10)}`;
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        baseComponent.handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Time input';
        }
        // if no datalistId value provided, use internal datalist id to enable time options
        if (!this.datalistId) {
            this.datalistId = this.internalDatalistId;
        }
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = [
            'modus-wc-time-input',
            'modus-wc-input',
            'modus-wc-w-full',
        ];
        const propClasses = convertPropsToClasses({
            bordered: this.bordered,
            feedback: this.feedback,
            readOnly: this.readOnly,
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    /*
     * Conditionally renders the datalist element with the time options.
     * If no time options are provided or the datalistId prop is not the default,
     * the datalist element will not be rendered (returns `null`).
     */
    renderDatalist() {
        if (this.datalistOptions.length === 0 ||
            this.datalistId !== this.internalDatalistId) {
            return null;
        }
        return (index.h("datalist", { id: this.internalDatalistId }, this.datalistOptions.map((time) => (index.h("option", { value: time })))));
    }
    render() {
        return (index.h(index.Host, { key: '0765a2fd72542f522e4317c488911357c9e07371' }, this.label && (index.h("modus-wc-input-label", { key: '113bab6244d974fe5640ec5786f33e8d1113828a', forId: this.inputId, labelText: this.label, required: this.required, size: this.size })), index.h("input", Object.assign({ key: '360b24527fb7d2b95db882ff90a2d161706a1adc', "aria-required": this.required, autocomplete: this.autoComplete, class: this.getClasses(), disabled: this.disabled, id: this.inputId, list: this.datalistId, max: this.max, min: this.min, name: this.name, onBlur: this.handleBlur, onFocus: this.handleFocus, onInput: this.handleInput, readonly: this.readOnly, required: this.required, step: this.step || this.showSeconds ? 1 : 60, tabIndex: this.inputTabIndex, type: "time", value: this.value }, this.inheritedAttributes)), this.renderDatalist(), this.feedback && (index.h("modus-wc-input-feedback", { key: 'eff36a250933edd6de9cd3d613fc31c98d571dba', level: this.feedback.level, message: this.feedback.message, size: this.size }))));
    }
    get el() { return index.getElement(this); }
};
ModusWcTimeInput.style = modusWcTimeInputCss;

exports.modus_wc_time_input = ModusWcTimeInput;
