// @trimble-oss/moduswebcomponents/cjs/modus-wc-menu_2.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var baseComponent = require('./base-component-ZxJWB8zi.js');
var utils = require('./utils-DiibENGv.js');
var popper = require('./popper-Dmp8OTYh.js');

const convertPropsToClasses = ({ bordered, orientation, size, }) => {
    let classes = '';
    if (bordered) {
        classes = `${classes} modus-wc-menu--bordered`;
    }
    if (orientation === 'horizontal') {
        classes = `${classes} modus-wc-menu-horizontal`;
    }
    if (size) {
        classes = `${classes} modus-wc-menu-${size}`;
    }
    return classes.trim();
};

const modusWcMenuCss = "modus-wc-menu.modus-wc-menu-submenu{display:contents}modus-wc-menu .modus-wc-menu{background-color:var(--modus-wc-color-base-page);border-radius:var(--modus-wc-border-radius-md);list-style:none;margin:0;overflow:hidden;padding:0;padding-inline-start:0}modus-wc-menu .modus-wc-menu.modus-wc-menu--bordered{border:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-base-200)}modus-wc-menu .modus-wc-menu-dropdown{display:none;list-style:none}modus-wc-menu .modus-wc-menu-dropdown-show{display:block}[data-theme=modus-classic-light] modus-wc-menu .modus-wc-menu{background-color:var(--modus-wc-color-white);border-color:var(--modus-wc-color-gray-0)}[data-theme=modus-classic-dark] modus-wc-menu .modus-wc-menu{background-color:var(--modus-wc-color-black);border-color:var(--modus-wc-color-gray-8)}";

const ModusWcMenu = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.menuFocusout = index.createEvent(this, "menuFocusout");
        this.menuSelectionChange = index.createEvent(this, "menuSelectionChange");
        this.inheritedAttributes = {};
        this.selectedItems = [];
        /** Custom CSS class to apply to the ul element. */
        this.customClass = '';
        /** The orientation of the menu. */
        this.orientation = 'vertical';
        /** The selection mode of the menu. */
        this.selectionMode = 'single';
        /** The size of the menu. */
        this.size = 'md';
        this.getMenuRole = () => this.orientation === 'horizontal' ? 'menubar' : 'menu';
    }
    onSelectionModeChange() {
        this.selectedItems = [];
    }
    componentWillLoad() {
        baseComponent.handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Menu';
        }
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        // For submenus, only add the dropdown class
        if (this.isSubMenu) {
            const classList = ['modus-wc-menu-dropdown'];
            if (this.customClass)
                classList.push(this.customClass);
            return classList.join(' ');
        }
        // For regular menus, add all the standard classes
        const classList = ['modus-wc-menu modus-wc-w-full'];
        const propClasses = convertPropsToClasses({
            bordered: this.bordered,
            orientation: this.orientation,
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getMenuItems() {
        return Array.from(this.el.querySelectorAll('modus-wc-menu-item')).filter((item) => item.closest('modus-wc-menu') === this.el);
    }
    handleItemSelect(e) {
        if (this.selectionMode !== 'multiple')
            return;
        const item = e.target;
        const isCurrentlySelected = this.selectedItems.includes(item);
        if (e.detail.selected && !isCurrentlySelected) {
            this.selectedItems = [...this.selectedItems, item];
        }
        else if (!e.detail.selected && isCurrentlySelected) {
            this.selectedItems = this.selectedItems.filter((i) => i !== item);
        }
        this.menuSelectionChange.emit({ selectedItems: this.selectedItems });
    }
    handleKeyDown(e) {
        if (e.key !== 'ArrowDown' && e.key !== 'ArrowUp')
            return;
        e.preventDefault();
        const items = this.getMenuItems();
        const focusableItems = items.filter((item) => !item.disabled);
        if (focusableItems.length === 0)
            return;
        const activeEl = document.activeElement;
        const currentMenuItem = activeEl === null || activeEl === void 0 ? void 0 : activeEl.closest('modus-wc-menu-item');
        const currentIndex = focusableItems.indexOf(currentMenuItem);
        let nextIndex;
        if (e.key === 'ArrowDown') {
            nextIndex =
                currentIndex < focusableItems.length - 1 ? currentIndex + 1 : 0;
        }
        else {
            nextIndex =
                currentIndex > 0 ? currentIndex - 1 : focusableItems.length - 1;
        }
        const nextLi = focusableItems[nextIndex].querySelector('li');
        if (nextLi) {
            nextLi.focus();
        }
    }
    handleFocusout(e) {
        if (!this.el.contains(e.relatedTarget)) {
            this.menuFocusout.emit(e);
            if (this.isSubMenu) {
                e.stopPropagation();
            }
        }
    }
    render() {
        return (index.h(index.Host, { key: '3cb31a2456fba3bbf777adbd1e2d9a076bc2017d', class: this.isSubMenu ? 'modus-wc-menu-submenu' : undefined }, index.h("ul", Object.assign({ key: '3e318db592a937f62b2b0f91a8d38a8dac9d34e7', "aria-orientation": this.orientation, class: this.getClasses(), role: this.getMenuRole() }, this.inheritedAttributes), index.h("slot", { key: 'e62e5f17e2f45b6214cecb605c29787b6ddcdb5e' }))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "selectionMode": ["onSelectionModeChange"]
    }; }
};
ModusWcMenu.style = modusWcMenuCss;

const modusWcTooltipCss = ".modus-wc-tooltip-content[popover]{overflow:visible}.modus-wc-tooltip-content{background-color:var(--modus-wc-color-neutral-content);border-radius:0.25rem;color:var(--modus-wc-color-neutral);font-size:0.875rem;font-weight:var(--modus-wc-font-weight-regular);max-width:20rem;overflow-wrap:break-word;padding:0.25rem 0.5rem;pointer-events:none;position:relative;text-align:center;white-space:normal;width:max-content;word-break:break-word;z-index:1000}.modus-wc-tooltip-arrow,.modus-wc-tooltip-arrow::before{background:inherit;height:8px;position:absolute;width:8px}.modus-wc-tooltip-arrow{text-align:initial;visibility:hidden}.modus-wc-tooltip-arrow::before{background-color:var(--modus-wc-color-neutral-content);content:\"\";transform:rotate(45deg);visibility:visible}.modus-wc-tooltip-content[data-popper-placement^=top] .modus-wc-tooltip-arrow{bottom:-4px}.modus-wc-tooltip-content[data-popper-placement^=bottom] .modus-wc-tooltip-arrow{top:-4px}.modus-wc-tooltip-content[data-popper-placement^=left] .modus-wc-tooltip-arrow{right:-4px}.modus-wc-tooltip-content[data-popper-placement^=right] .modus-wc-tooltip-arrow{left:-4px}[data-theme=modus-classic-light] .modus-wc-tooltip-content{background-color:var(--modus-wc-color-gray-7);color:var(--modus-wc-color-white)}[data-theme=modus-classic-light] .modus-wc-tooltip-arrow::before{background-color:var(--modus-wc-color-gray-7)}[data-theme=modus-classic-dark] .modus-wc-tooltip-content{background-color:var(--modus-wc-color-gray-0);color:var(--modus-wc-color-trimble-gray)}[data-theme=modus-classic-dark] .modus-wc-tooltip-arrow::before{background-color:var(--modus-wc-color-gray-0)}[data-theme=connect-light] .modus-wc-tooltip-content,[data-theme=connect-dark] .modus-wc-tooltip-content{background-color:var(--modus-wc-color-base-content);color:var(--modus-wc-color-base-page)}[data-theme=connect-light] .modus-wc-tooltip-arrow::before,[data-theme=connect-dark] .modus-wc-tooltip-arrow::before{background-color:var(--modus-wc-color-base-content)}";

const ModusWcTooltip = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.dismissEscape = index.createEvent(this, "dismissEscape");
        this.inheritedAttributes = {};
        this.popperInstance = null;
        this.tooltipElement = null;
        this.triggerElement = null;
        /** The text content of the tooltip. */
        this.content = '';
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** Disables displaying the tooltip on hover */
        this.disabled = false;
        /** The position that the tooltip will render in relation to the element. */
        this.position = 'auto';
        /** Track if tooltip was dismissed with Escape key */
        this.escapeDismissed = false;
        /** Track if tooltip is currently visible */
        this.isVisible = false;
        this.handleWindowResize = () => {
            if (this.popperInstance && this.isVisible) {
                void this.popperInstance.update();
            }
        };
        this.handleWindowScroll = () => {
            if (this.popperInstance && this.isVisible) {
                void this.popperInstance.update();
            }
        };
    }
    componentWillLoad() {
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    elementKeyupHandler(event) {
        switch (event.code) {
            case 'Escape': {
                // Allow Escape to dismiss tooltip when it's visible
                // When forceOpen is true, Escape should NOT dismiss it
                if (this.isVisible && !this.forceOpen) {
                    this.escapeDismissed = true;
                    this.dismissEscape.emit();
                    this.hideTooltip();
                }
                break;
            }
        }
    }
    componentDidLoad() {
        this.triggerElement = this.el.querySelector('div > :first-child');
        this.tooltipElement = document.createElement('div');
        this.tooltipElement.className = `modus-wc-tooltip-content ${this.customClass || ''}`;
        this.tooltipElement.textContent = this.content;
        this.tooltipElement.setAttribute('role', 'tooltip');
        if (this.tooltipId) {
            this.tooltipElement.id = this.tooltipId;
        }
        const arrow = document.createElement('div');
        arrow.className = 'modus-wc-tooltip-arrow';
        this.tooltipElement.appendChild(arrow);
        this.tooltipElement.setAttribute('popover', 'manual');
        document.body.appendChild(this.tooltipElement);
        this.tooltipElement.style.display = 'none';
        if (this.triggerElement && this.tooltipElement) {
            this.initializePopper();
        }
        if (this.forceOpen && !this.disabled && !this.escapeDismissed) {
            this.showTooltip();
        }
    }
    disconnectedCallback() {
        if (this.popperInstance) {
            this.popperInstance.destroy();
            this.popperInstance = null;
        }
        if (this.tooltipElement) {
            if (typeof this.tooltipElement.hidePopover === 'function') {
                try {
                    this.tooltipElement.hidePopover();
                }
                catch (_a) {
                    // Already hidden or element not connected
                }
            }
            if (this.tooltipElement.parentElement) {
                this.tooltipElement.parentElement.removeChild(this.tooltipElement);
            }
        }
        window.removeEventListener('resize', this.handleWindowResize);
        window.removeEventListener('scroll', this.handleWindowScroll, true);
    }
    initializePopper() {
        if (!this.triggerElement || !this.tooltipElement)
            return;
        const placement = this.position === 'auto' ? 'top' : this.position;
        const arrowElement = this.tooltipElement.querySelector('.modus-wc-tooltip-arrow');
        this.popperInstance = popper.createPopper(this.triggerElement, this.tooltipElement, {
            placement,
            strategy: 'fixed',
            modifiers: [
                {
                    name: 'offset',
                    options: {
                        offset: [0, 8],
                    },
                },
                {
                    name: 'preventOverflow',
                    options: {
                        padding: 8,
                        boundary: 'viewport',
                    },
                },
                {
                    name: 'flip',
                    options: {
                        fallbackPlacements: ['top', 'right', 'bottom', 'left'],
                        padding: 8,
                        boundary: 'viewport',
                    },
                },
                {
                    name: 'arrow',
                    options: {
                        element: arrowElement,
                        padding: 5,
                    },
                },
                {
                    name: 'computeStyles',
                    options: {
                        adaptive: true,
                        gpuAcceleration: true,
                    },
                },
                {
                    name: 'eventListeners',
                    options: {
                        scroll: true,
                        resize: true,
                    },
                },
            ],
        });
        window.addEventListener('resize', this.handleWindowResize);
        window.addEventListener('scroll', this.handleWindowScroll, true);
    }
    showTooltip() {
        if (this.disabled || this.escapeDismissed || !this.tooltipElement)
            return;
        this.tooltipElement.style.display = 'block';
        if (typeof this.tooltipElement.showPopover === 'function') {
            try {
                this.tooltipElement.showPopover();
            }
            catch (_a) {
                // Already showing or element not connected
            }
        }
        this.isVisible = true;
        if (this.popperInstance) {
            void this.popperInstance.update();
            // Force a second update after a short delay to ensure arrow positioning
            setTimeout(() => {
                if (this.popperInstance) {
                    void this.popperInstance.update();
                }
            }, 10);
        }
    }
    hideTooltip() {
        if (!this.tooltipElement)
            return;
        if (!this.forceOpen || this.escapeDismissed) {
            if (typeof this.tooltipElement.hidePopover === 'function') {
                try {
                    this.tooltipElement.hidePopover();
                }
                catch (_a) {
                    // Already hidden or element not connected
                }
            }
            this.tooltipElement.style.display = 'none';
            this.isVisible = false;
        }
    }
    handlePositionChange() {
        if (this.popperInstance) {
            void this.popperInstance.setOptions({
                placement: this.position === 'auto' ? 'top' : this.position,
            });
            void this.popperInstance.update();
        }
    }
    handleContentChange(newContent) {
        if (this.tooltipElement) {
            const arrow = this.tooltipElement.querySelector('.modus-wc-tooltip-arrow');
            this.tooltipElement.textContent = newContent;
            if (arrow) {
                this.tooltipElement.appendChild(arrow);
            }
        }
    }
    handleForceOpenChange(forceOpen) {
        if (forceOpen && !this.disabled) {
            this.showTooltip();
        }
        else {
            this.hideTooltip();
        }
    }
    handleMouseEnter() {
        this.escapeDismissed = false;
        this.showTooltip();
    }
    handleMouseLeave() {
        if (!this.forceOpen) {
            this.hideTooltip();
        }
    }
    render() {
        return (index.h(index.Host, { key: '622edbe623c751b8cfd6d8ca87be8d62886d0487' }, index.h("div", Object.assign({ key: 'd4d771a5aea600d080704727da8ad02cfbf6d65f', "aria-describedby": this.tooltipId, id: this.tooltipId }, this.inheritedAttributes), index.h("slot", { key: '73d33584d20e1bf545773ca9e613af13cde51106' }))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "position": ["handlePositionChange"],
        "content": ["handleContentChange"],
        "forceOpen": ["handleForceOpenChange"]
    }; }
};
ModusWcTooltip.style = modusWcTooltipCss;

exports.modus_wc_menu = ModusWcMenu;
exports.modus_wc_tooltip = ModusWcTooltip;
