// @trimble-oss/moduswebcomponents/cjs/modus-wc-breadcrumbs.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var utils = require('./utils-DiibENGv.js');

const convertPropsToClasses = ({ size, }) => {
    let classes = '';
    if (size) {
        classes = `${classes} modus-wc-text-${size}`;
    }
    return classes.trim();
};

const modusWcBreadcrumbsCss = "modus-wc-breadcrumbs .modus-wc-breadcrumbs ol{list-style:none}modus-wc-breadcrumbs .modus-wc-breadcrumbs.modus-wc-text-xs{font-size:var(--modus-wc-font-size-xs)}modus-wc-breadcrumbs .modus-wc-breadcrumbs.modus-wc-text-sm{font-size:var(--modus-wc-font-size-sm)}modus-wc-breadcrumbs .modus-wc-breadcrumbs.modus-wc-text-md{font-size:var(--modus-wc-font-size-md)}modus-wc-breadcrumbs .modus-wc-breadcrumbs.modus-wc-text-lg{font-size:var(--modus-wc-font-size-lg)}[data-theme=modus-classic-light] modus-wc-breadcrumbs .modus-wc-breadcrumbs a,[data-theme=modus-classic-dark] modus-wc-breadcrumbs .modus-wc-breadcrumbs a{text-decoration:none}[data-theme=modus-classic-light] modus-wc-breadcrumbs .modus-wc-breadcrumbs a:hover,[data-theme=modus-classic-dark] modus-wc-breadcrumbs .modus-wc-breadcrumbs a:hover{text-decoration:underline}[data-theme=modus-classic-light] modus-wc-breadcrumbs .modus-wc-breadcrumbs a{color:var(--modus-wc-color-blue-light)}[data-theme=modus-classic-light] modus-wc-breadcrumbs .modus-wc-breadcrumbs a:active{color:var(--modus-wc-color-trimble-gray)}[data-theme=modus-classic-light] modus-wc-breadcrumbs .modus-wc-breadcrumbs span{color:var(--modus-wc-color-trimble-gray)}[data-theme=modus-classic-dark] modus-wc-breadcrumbs .modus-wc-breadcrumbs a{color:var(--modus-wc-color-highlight-blue)}[data-theme=modus-classic-dark] modus-wc-breadcrumbs .modus-wc-breadcrumbs a:active{color:var(--modus-wc-color-white)}[data-theme=modus-classic-dark] modus-wc-breadcrumbs .modus-wc-breadcrumbs span{color:var(--modus-wc-color-white)}[data-theme=connect-light] modus-wc-breadcrumbs .modus-wc-breadcrumbs a,[data-theme=connect-dark] modus-wc-breadcrumbs .modus-wc-breadcrumbs a{color:var(--modus-wc-color-base-content);opacity:0.5;text-decoration:none}[data-theme=connect-light] modus-wc-breadcrumbs .modus-wc-breadcrumbs a:active,[data-theme=connect-dark] modus-wc-breadcrumbs .modus-wc-breadcrumbs a:active{color:var(--modus-wc-color-white)}[data-theme=connect-light] modus-wc-breadcrumbs .modus-wc-breadcrumbs a:hover,[data-theme=connect-dark] modus-wc-breadcrumbs .modus-wc-breadcrumbs a:hover{text-decoration:none}";

const ModusWcBreadcrumbs = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.breadcrumbClick = index.createEvent(this, "breadcrumbClick");
        this.inheritedAttributes = {};
        /** The breadcrumbs to render. */
        this.items = [];
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The size of the breadcrumbs. */
        this.size = 'md';
    }
    componentWillLoad() {
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Breadcrumbs';
        }
        if (!this.items || this.items.length === 0) {
            console.error('ModusWcBreadcrumbs: breadcrumb items data is required.');
        }
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-breadcrumbs'];
        const propClasses = convertPropsToClasses({
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    handleClick(event, crumb) {
        if (!crumb.url) {
            event.preventDefault();
        }
        this.breadcrumbClick.emit(crumb);
    }
    render() {
        return (index.h(index.Host, { key: '9962a34f1dac82f1f3e7556d406c9d4863c014ad' }, index.h("nav", Object.assign({ key: '67435e57f382ff45f8698202aa12eb65e6d6994c', class: this.getClasses() }, this.inheritedAttributes), index.h("ol", { key: 'aeeb9c2fe11832c1b307821da45851e705b0044a' }, this.items.map((item, index$1) => {
            const isCurrentPage = index$1 === this.items.length - 1;
            return (index.h("li", Object.assign({ key: item.label }, (isCurrentPage ? { 'aria-current': 'page' } : {})), isCurrentPage ? (index.h("span", null, item.label)) : (index.h("a", { href: item.url, onClick: (event) => this.handleClick(event, item) }, item.label))));
        })))));
    }
    get el() { return index.getElement(this); }
};
ModusWcBreadcrumbs.style = modusWcBreadcrumbsCss;

exports.modus_wc_breadcrumbs = ModusWcBreadcrumbs;
