// @trimble-oss/moduswebcomponents/cjs/modus-wc-theme-provider.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var theme_store = require('./theme.store-CHszHwzP.js');

const ThemeProvider = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    componentWillLoad() {
        theme_store.initializeThemeStore(this.initialTheme);
        this.cleanup = theme_store.watchSystemTheme();
        // Subscribe to specific property changes
        this.modeUnsubscribe = theme_store.themeStore.onChange('mode', () => {
            this.updateThemeClasses();
        });
        this.themeUnsubscribe = theme_store.themeStore.onChange('theme', () => {
            this.updateThemeClasses();
        });
        this.updateThemeClasses();
    }
    disconnectedCallback() {
        var _a, _b, _c;
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        (_b = this.modeUnsubscribe) === null || _b === void 0 ? void 0 : _b.call(this);
        (_c = this.themeUnsubscribe) === null || _c === void 0 ? void 0 : _c.call(this);
    }
    updateThemeClasses() {
        const root = document.documentElement;
        const { mode, theme } = theme_store.themeStore.state;
        // Update theme and mode attributes
        root.setAttribute('data-theme', `${theme}-${mode}`);
        root.setAttribute('data-mode', mode);
        // Remove old mode classes
        root.classList.remove('light', 'dark');
        root.classList.add(mode);
        // Store preferences
        localStorage.setItem('modus-theme-config', JSON.stringify({ mode, theme }));
    }
    render() {
        return index.h("slot", { key: 'd3409eb8bdbf538d51eeacbb5b70f99d496b2d40' });
    }
};

exports.modus_wc_theme_provider = ThemeProvider;
