// @trimble-oss/moduswebcomponents/cjs/modus-wc-skeleton.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var utils = require('./utils-DiibENGv.js');

const convertPropsToClasses = ({ shape, }) => {
    let classes = '';
    if (shape === 'circle') {
        classes = `${classes} modus-wc-rounded-full`;
    }
    return classes.trim();
};

const modusWcSkeletonCss = "";

const ModusWcSkeleton = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The height of the skeleton. */
        this.height = 'var(--modus-wc-line-height-md)';
        /** The shape of the skeleton. */
        this.shape = 'rectangle';
        /** The width of the skeleton. */
        this.width = '100%';
    }
    componentWillLoad() {
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-skeleton'];
        const propClasses = convertPropsToClasses({
            shape: this.shape,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getStyles() {
        return { height: this.height, width: this.width };
    }
    render() {
        return (index.h(index.Host, { key: '9d67bdc774912492b42d055e601cc3143fe56fd8' }, index.h("div", Object.assign({ key: 'd30972478d13d8d5c604077f12c6204345fe8b0c', "aria-hidden": "true", class: this.getClasses(), style: this.getStyles(), tabindex: -1 }, this.inheritedAttributes))));
    }
    get el() { return index.getElement(this); }
};
ModusWcSkeleton.style = modusWcSkeletonCss;

exports.modus_wc_skeleton = ModusWcSkeleton;
