// @trimble-oss/moduswebcomponents/cjs/modus-wc-theme-switcher.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var theme_store = require('./theme.store-CHszHwzP.js');
var utils = require('./utils-DiibENGv.js');

const modusWcThemeSwitcherCss = "[data-theme=modus-modern-dark] .modus-wc-toggle{--tglbg:var(--modus-wc-color-gray-9);border-color:var(--modus-wc-color-black)}[data-theme=modus-modern-dark] .modus-wc-stroke-base-100{stroke:var(--modus-wc-color-white)}[data-theme=modus-modern-dark] .modus-wc-fill-base-100{fill:var(--modus-wc-color-gray-9)}[data-theme=modus-modern-dark] .modus-wc-bg-base-content{background-color:var(--modus-wc-color-black)}[data-theme=modus-modern-light] .modus-wc-toggle{--tglbg:var(--modus-wc-color-gray-0);border-color:var(--modus-wc-white)}[data-theme=modus-modern-light] .modus-wc-stroke-base-100{stroke:var(--modus-wc-color-black)}[data-theme=modus-modern-light] .modus-wc-fill-base-100{fill:var(--modus-wc-color-gray-0)}[data-theme=modus-modern-light] .modus-wc-bg-base-content{background-color:var(--modus-wc-color-white)}.modus-wc-toggle.modus-wc-stroke-base-100{stroke:var(--modus-wc-color-base-300)}.modus-wc-toggle.modus-wc-fill-base-100{fill:var(--modus-wc-color-base-300)}.modus-wc-toggle:checked.modus-wc-stroke-base-100{stroke:var(--modus-wc-color-base-100)}.modus-wc-toggle:checked.modus-wc-fill-base-100{fill:var(--modus-wc-color-base-100)}";

const ModusWcThemeSwitcher = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.themeChange = index.createEvent(this, "themeChange");
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the theme switcher element. */
        this.customClass = '';
        this.isDarkMode = theme_store.themeStore.state.mode === 'dark';
    }
    componentWillLoad() {
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Switch between light and dark theme';
        }
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    connectedCallback() {
        this.modeUnsubscribe = theme_store.themeStore.onChange('mode', (mode) => {
            this.isDarkMode = mode === 'dark';
        });
    }
    disconnectedCallback() {
        var _a;
        (_a = this.modeUnsubscribe) === null || _a === void 0 ? void 0 : _a.call(this);
    }
    getClasses() {
        const classList = [
            'modus-wc-inline-grid modus-wc-cursor-pointer modus-wc-justify-center modus-wc-place-items-center',
        ];
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    handleModeToggle(event) {
        const checkbox = event.target;
        const newMode = checkbox.checked ? 'dark' : 'light';
        theme_store.themeStore.setMode(newMode);
        this.themeChange.emit(theme_store.themeStore.state);
    }
    render() {
        return (index.h("label", Object.assign({ key: '0c000bfb23bd5bd09e00cbf0a46c08c9796760f0', class: this.getClasses() }, this.inheritedAttributes), index.h("input", { key: '133149da542266b03a2999ce6c5ed7de3fc89927', "aria-checked": this.isDarkMode, "aria-label": this.isDarkMode ? 'Dark theme' : 'Light theme', checked: this.isDarkMode, class: "modus-wc-toggle modus-wc-theme-controller modus-wc-bg-base-content modus-wc-col-span-2 modus-wc-col-start-1 modus-wc-row-start-1", onChange: (event) => this.handleModeToggle(event), type: "checkbox", value: "default" }), index.h("svg", { key: '319f212dd67affe15e7b0155d5218ea243966f80', "aria-hidden": "true", class: "modus-wc-stroke-base-100 modus-wc-fill-base-100 modus-wc-col-start-1 modus-wc-row-start-1", fill: "none", height: "14", role: "presentation", stroke: "currentColor", "stroke-linecap": "round", "stroke-linejoin": "round", "stroke-width": "2", viewBox: "0 0 24 24", width: "14", xmlns: "http://www.w3.org/2000/svg" }, index.h("circle", { key: '8bc687a4727c4b1250dfb889b63c8a91a65bada0', cx: "12", cy: "12", r: "5" }), index.h("path", { key: 'e954efc744f08401174a00acdbcf6704dfcd281b', d: "M12 1v2M12 21v2M4.2 4.2l1.4 1.4M18.4 18.4l1.4 1.4M1 12h2M21 12h2M4.2 19.8l1.4-1.4M18.4 5.6l1.4-1.4" })), index.h("svg", { key: '52a855dfb6be752cdfe172c836423f5e24673ab9', "aria-hidden": "true", class: "modus-wc-stroke-base-100 modus-wc-fill-base-100 modus-wc-col-start-2 modus-wc-row-start-1", fill: "none", height: "14", role: "presentation", stroke: "currentColor", "stroke-linecap": "round", "stroke-linejoin": "round", "stroke-width": "2", viewBox: "0 0 24 24", width: "14", xmlns: "http://www.w3.org/2000/svg" }, index.h("path", { key: 'f974f3296e776fadb8a9e50dd687bfafc8bb6824', d: "M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z" }))));
    }
    get el() { return index.getElement(this); }
};
ModusWcThemeSwitcher.style = modusWcThemeSwitcherCss;

exports.modus_wc_theme_switcher = ModusWcThemeSwitcher;
