// @trimble-oss/moduswebcomponents/cjs/modus-wc-stepper.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var utils = require('./utils-DiibENGv.js');

const convertPropsToClasses = ({ orientation, }) => {
    let classes = '';
    if (orientation) {
        switch (orientation) {
            case 'horizontal':
                break; // Default
            case 'vertical':
                classes = `${classes} modus-wc-steps-vertical`;
                break;
        }
    }
    return classes.trim();
};

const modusWcStepperCss = "modus-wc-stepper .modus-wc-steps .modus-wc-step{--fallback-b3:var(--modus-wc-color-base-200)}";

const ModusWcStepper = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the steps element. */
        this.customClass = '';
        /** The steps to display. */
        this.steps = [];
    }
    componentWillLoad() {
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-steps'];
        const propClasses = convertPropsToClasses({
            orientation: this.orientation,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getClassesForStep(step) {
        const classList = ['modus-wc-step'];
        // The order CSS classes are added matters to CSS specificity
        if (step.color)
            classList.push(`modus-wc-step-${step.color}`);
        if (step.customClass)
            classList.push(step.customClass);
        return classList.join(' ');
    }
    render() {
        return (index.h(index.Host, { key: '7d8e5badd7eedcdb7fc6a997b94f12008c5bfcac' }, index.h("div", Object.assign({ key: 'f3ce9e091ca2418bda52cae1661d2dae7d04fadd', class: this.getClasses() }, this.inheritedAttributes), this.steps.map((step, index$1) => {
            return (index.h("li", { class: this.getClassesForStep(step), key: index$1, "data-content": step.content }, step.label));
        }))));
    }
    get el() { return index.getElement(this); }
};
ModusWcStepper.style = modusWcStepperCss;

exports.modus_wc_stepper = ModusWcStepper;
