// @trimble-oss/moduswebcomponents/cjs/modus-wc-accordion.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var baseComponent = require('./base-component-ZxJWB8zi.js');
var utils = require('./utils-DiibENGv.js');

const modusWcAccordionCss = "[data-theme=modus-classic-light] modus-wc-accordion modus-wc-collapse:first-child .modus-wc-collapse,[data-theme=modus-classic-dark] modus-wc-accordion modus-wc-collapse:first-child .modus-wc-collapse{border-bottom-left-radius:0;border-bottom-right-radius:0}[data-theme=modus-classic-light] modus-wc-accordion modus-wc-collapse:not(:first-child):not(:last-child) .modus-wc-collapse,[data-theme=modus-classic-dark] modus-wc-accordion modus-wc-collapse:not(:first-child):not(:last-child) .modus-wc-collapse{border-radius:0}[data-theme=modus-classic-light] modus-wc-accordion modus-wc-collapse:last-child .modus-wc-collapse,[data-theme=modus-classic-dark] modus-wc-accordion modus-wc-collapse:last-child .modus-wc-collapse{border-top-left-radius:0;border-top-right-radius:0}[data-theme=modus-classic-light] modus-wc-accordion .modus-wc-collapse{border:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-gray-2)}[data-theme=modus-classic-dark] modus-wc-accordion .modus-wc-collapse{border:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-gray-8)}";

const ModusWcAccordion = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.expandedChange = index.createEvent(this, "expandedChange");
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        baseComponent.handleShadowDOMStyles(this.el);
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = [''];
        // The order CSS classes are added matters to CSS specificity
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (index.h(index.Host, { key: 'cd50dc732b2fc48835c00a151a3bd53baea5e45f' }, index.h("div", Object.assign({ key: 'e5449efc6472a41424ab3eb550f6cf6e965f826e', class: this.getClasses() }, this.inheritedAttributes), index.h("slot", { key: '1c32e9a0eb77ad1d5ee753f29b4bc78f881c0b48' }))));
    }
    get el() { return index.getElement(this); }
};
ModusWcAccordion.style = modusWcAccordionCss;

exports.modus_wc_accordion = ModusWcAccordion;
