// @trimble-oss/moduswebcomponents/cjs/modus-wc-date.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var baseComponent = require('./base-component-ZxJWB8zi.js');
var utils = require('./utils-DiibENGv.js');
var popper = require('./popper-Dmp8OTYh.js');

const convertPropsToClasses = ({ bordered, feedback, readOnly, size, }) => {
    let classes = '';
    if (bordered) {
        classes = `${classes} modus-wc-input-bordered`;
    }
    if (feedback) {
        classes = `${classes} modus-wc-input--${feedback.level}`;
    }
    if (readOnly) {
        classes = `${classes} modus-wc-date-input--readonly`;
    }
    if (size) {
        classes = `${classes} modus-wc-input-${size}`;
    }
    return classes.trim();
};

class DatePickerCalendar {
    constructor(firstDayOfWeek = 0) {
        this.currentDate = new Date();
        this.currentMonthDates = [];
        this.firstDayOfWeek = 0; // Default to Sunday
        this.firstDayOfWeek = firstDayOfWeek;
        const today = new Date();
        this.gotoDate(today.getFullYear(), today.getMonth());
    }
    get selectedYear() {
        return this.currentDate.getFullYear();
    }
    get selectedMonth() {
        return this.currentDate.getMonth();
    }
    get dates() {
        return this.currentMonthDates;
    }
    addMonthOffset(offset) {
        this.gotoDate(this.currentDate.getFullYear(), this.currentDate.getMonth() + offset);
        return this;
    }
    gotoDate(year, month) {
        this.currentDate = new Date(year, month, 1);
        this.calculateDates();
    }
    getDaysOfWeek(locale, firstDayOfWeek = 0) {
        /**
         * Nov 1st, 2020 starts on a Sunday,
         * assumes weeks start on Sunday,
         * but is configurable via `firstDayOfWeek`.
         */
        const intl = new Intl.DateTimeFormat(locale, { weekday: 'short' });
        const startDate = new Date('11/01/2020');
        const daysOfWeek = [];
        /**
         * For each day of the week,
         * get the day name.
         */
        for (let i = firstDayOfWeek; i < firstDayOfWeek + 7; i++) {
            const currentDate = new Date(startDate);
            currentDate.setDate(currentDate.getDate() + i);
            const d = intl.format(currentDate);
            daysOfWeek.push(d.toUpperCase().startsWith('SA') ? d : d.slice(0, 2));
        }
        return daysOfWeek;
    }
    /**
     * Get ISO week number for a given date
     * ISO 8601: Week 1 is the week with the year's first Thursday
     */
    getWeekNumber(date, weekStart = 1) {
        function isoWeek(d) {
            const temp = new Date(Date.UTC(d.getFullYear(), d.getMonth(), d.getDate()));
            const day = temp.getUTCDay() || 7;
            temp.setUTCDate(temp.getUTCDate() + 4 - day);
            const yearStart = new Date(Date.UTC(temp.getUTCFullYear(), 0, 1));
            return Math.ceil(((temp.getTime() - yearStart.getTime()) / 86400000 + 1) / 7);
        }
        // --- FIND START OF WEEK ---
        const day = date.getDay();
        const diff = (day - weekStart + 7) % 7;
        const weekStartDate = new Date(date);
        weekStartDate.setDate(date.getDate() - diff);
        // --- COLLECT ALL WEEK NUMBERS ---
        const weekCounts = {};
        for (let i = 0; i < 7; i++) {
            const d = new Date(weekStartDate);
            d.setDate(weekStartDate.getDate() + i);
            const w = isoWeek(d);
            weekCounts[w] = (weekCounts[w] || 0) + 1;
        }
        // --- PICK WEEK WITH MOST OCCURRENCES ---
        let majorityWeek = 0;
        let max = 0;
        for (const [weekStr, count] of Object.entries(weekCounts)) {
            const week = Number(weekStr);
            if (count > max) {
                max = count;
                majorityWeek = week;
            }
        }
        // Return double-digit week number
        return majorityWeek.toString().padStart(2, '0');
    }
    calculateDates() {
        const dates = [];
        const year = this.currentDate.getFullYear();
        const month = this.currentDate.getMonth();
        // Get first day of current month
        const firstDayOfMonth = new Date(year, month, 1);
        const dayOfWeek = firstDayOfMonth.getDay(); // 0 = Sunday, 1 = Monday, etc.
        // Simple calculation: if month starts on Wednesday (3) and we want Monday (1) as first day,
        // we need to show 2 days from previous month (3 - 1 = 2)
        // But if month starts on Sunday (0) and we want Monday (1) as first day,
        // we need to show 6 days from previous month (0 - 1 = -1, so we add 7: 6)
        let daysToSubtract = dayOfWeek - this.firstDayOfWeek;
        if (daysToSubtract < 0) {
            daysToSubtract += 7;
        }
        // Start from the previous month's dates to fill the first week
        const startDate = new Date(year, month, 1 - daysToSubtract);
        // Generate 42 dates (6 weeks * 7 days) to ensure consistent 6-row layout
        for (let i = 0; i < 42; i++) {
            const date = new Date(startDate);
            date.setDate(startDate.getDate() + i);
            dates.push(date);
        }
        this.currentMonthDates = dates;
    }
}

const modusWcDateCss = ".modus-wc-input--error{border-color:var(--modus-wc-color-error) !important}.modus-wc-input--info{border-color:var(--modus-wc-color-info) !important}.modus-wc-input--success{border-color:var(--modus-wc-color-success) !important}.modus-wc-input--warning{border-color:var(--modus-wc-color-warning) !important}.modus-wc-input-xs{height:var(--modus-wc-size-xs);min-height:var(--modus-wc-size-xs)}.modus-wc-input-sm{height:var(--modus-wc-size-sm);min-height:var(--modus-wc-size-sm)}.modus-wc-input-md{height:var(--modus-wc-size-md);min-height:var(--modus-wc-size-md)}.modus-wc-input-lg{height:var(--modus-wc-size-lg);min-height:var(--modus-wc-size-lg)}.modus-wc-input-xl{height:var(--modus-wc-size-xl);min-height:var(--modus-wc-size-xl)}modus-wc-date .modus-wc-date-input{background-color:var(--modus-wc-color-base-page)}modus-wc-date .modus-wc-date-input .modus-wc-input-label{padding-bottom:var(--modus-wc-spacing-sm)}modus-wc-date .modus-wc-date-input--readonly{background-color:var(--modus-wc-color-base-200)}modus-wc-date{--calendar-grid-columns:repeat(7, 1fr);--calendar-grid-columns-with-week-numbers:auto repeat(7, 1fr);position:relative}modus-wc-date.modus-wc-date-input{background-color:var(--modus-wc-color-base-page);border-radius:var(--modus-wc-border-radius-input);color:var(--modus-wc-color-base-content)}modus-wc-date .date-input-container{align-items:center;display:inline-flex;position:relative;width:100%}modus-wc-date .date-input-container .calendar-icon-button{align-items:center;display:flex;justify-content:center;position:absolute;right:var(--modus-wc-spacing-md)}modus-wc-date .date-input-container .calendar-icon-button :hover{color:var(--modus-wc-color-primary)}modus-wc-date .date-input-container .calendar-icon-button :disabled{background-color:transparent;cursor:not-allowed}modus-wc-date .calendar-container{background:var(--modus-wc-color-base-page);border:1px solid var(--modus-wc-color-base-200);border-radius:var(--rounded-box);box-shadow:0 4px 12px rgba(0, 0, 0, 0.15);height:327px;overflow:hidden;position:fixed;width:272px;z-index:9999}modus-wc-date .calendar-container.has-week-numbers{width:320px}modus-wc-date .calendar-container .calendar-header{align-items:center;background-color:var(--modus-wc-color-primary);color:var(--modus-wc-color-base-page);display:flex;gap:var(--modus-wc-spacing-sm);justify-content:center;padding:var(--modus-wc-spacing-md) var(--modus-wc-spacing-xl)}modus-wc-date .calendar-container .calendar-header .nav-btn i{color:var(--modus-wc-color-primary-content)}modus-wc-date .calendar-container .calendar-header .nav-btn:hover i{color:var(--modus-wc-color-primary)}modus-wc-date .calendar-container .calendar-header .calendar-selects{align-items:center;display:flex;gap:var(--modus-wc-spacing-xs)}modus-wc-date .calendar-container .calendar-header .calendar-selects select{box-shadow:0 1px 2px 0 rgba(0, 0, 0, 0.05);color:var(--modus-wc-color-base-content)}modus-wc-date .calendar-container .calendar-header .calendar-selects .month-select{border-radius:var(--modus-wc-border-radius-input)}modus-wc-date .calendar-container .calendar-header .calendar-selects .year-select{border-radius:var(--modus-wc-border-radius-input)}modus-wc-date .calendar-container .calendar-header .calendar-selects select option{background-color:var(--modus-wc-color-base-page);color:var(--modus-wc-color-base-content)}modus-wc-date .calendar-container .calendar-body{background-color:var(--modus-wc-color-base-page);font-size:var(--modus-wc-font-size-md);height:272px;padding:var(--modus-wc-spacing-xl)}modus-wc-date .calendar-container .calendar-body .calendar-days-week{display:grid;gap:var(--modus-wc-spacing-xs);grid-template-columns:var(--calendar-grid-columns);margin-bottom:var(--modus-wc-spacing-sm)}modus-wc-date .calendar-container .calendar-body .calendar-days-week.has-week-numbers{grid-template-columns:var(--calendar-grid-columns-with-week-numbers)}modus-wc-date .calendar-container .calendar-body .calendar-days-week .week-number-header{margin-inline-end:var(--modus-wc-spacing-md);padding:var(--modus-wc-spacing-xs)}modus-wc-date .calendar-container .calendar-body .calendar-days-week .day-header{color:var(--modus-wc-color-base-content-high-contrast);font-weight:var(--modus-wc-font-weight-bold);padding:var(--modus-wc-spacing-xs);text-align:center}modus-wc-date .calendar-container .calendar-body .calendar-dates{display:grid;grid-template-columns:var(--calendar-grid-columns)}modus-wc-date .calendar-container .calendar-body .calendar-dates.has-week-numbers{grid-template-columns:var(--calendar-grid-columns-with-week-numbers)}modus-wc-date .calendar-container .calendar-body .calendar-dates .week-number{align-items:center;border-inline-end:1px solid color-mix(in sRGB, var(--modus-wc-color-base-inverted) 10%, transparent);display:flex;font-size:var(--modus-wc-font-size-sm);font-weight:var(--modus-wc-font-weight-bold);justify-content:center;padding-inline-end:var(--modus-wc-spacing-sm)}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day{align-items:center;background:none;border:none;border-radius:50%;color:var(--modus-wc-color-base-content-high-contrast);cursor:pointer;display:flex;height:32px;justify-content:center;width:32px}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day:hover{background-color:var(--modus-wc-color-primary);color:var(--modus-wc-color-white)}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day:focus{background-color:var(--modus-wc-color-primary);color:var(--modus-wc-color-white);outline:none}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day.current-day{border:2px solid var(--modus-wc-color-primary);outline:none}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day.selected{background-color:var(--modus-wc-color-primary);color:var(--modus-wc-color-primary-content);font-weight:var(--modus-wc-font-weight-semibold)}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day.other-month{color:var(--modus-wc-color-base-content-high-contrast);opacity:0.5}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day.other-month:hover{background-color:var(--modus-wc-color-base-200);color:var(--modus-wc-color-base-content)}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day.disabled{color:var(--modus-wc-color-gray-4)}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day.disabled:hover{background-color:transparent;color:var(--modus-wc-color-gray-4)}modus-wc-date .calendar-container .calendar-body .calendar-dates .calendar-day.disabled:focus{border:none}[data-theme=modus-classic-light] modus-wc-date .modus-wc-date-input.modus-wc-input,[data-theme=modus-classic-dark] modus-wc-date .modus-wc-date-input.modus-wc-input{border-radius:var(--modus-wc-border-radius-md);color:var(--modus-wc-color-base-content)}[data-theme=modus-classic-light] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-input-sm,[data-theme=modus-classic-dark] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-input-sm{font-size:var(--modus-wc-font-size-sm);height:var(--modus-wc-input-height-sm);padding:var(--modus-wc-spacing-sm) var(--modus-wc-spacing-xs)}[data-theme=modus-classic-light] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-input-md,[data-theme=modus-classic-dark] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-input-md{font-size:var(--modus-wc-font-size-md);height:var(--modus-wc-input-height-md);padding:var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-input-lg,[data-theme=modus-classic-dark] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-input-lg{font-size:var(--modus-wc-font-size-lg);height:var(--modus-wc-input-height-lg);padding:var(--modus-wc-spacing-md) var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-date .modus-wc-date-input.modus-wc-input:focus,[data-theme=modus-classic-dark] modus-wc-date .modus-wc-date-input.modus-wc-input:focus{border-color:var(--modus-wc-color-blue-light);border-width:var(--modus-wc-border-width-sm)}[data-theme=modus-classic-light] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-date-input--readonly,[data-theme=modus-classic-dark] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-date-input--readonly{background-color:var(--modus-wc-color-base-100)}[data-theme=modus-classic-light] modus-wc-date .modus-wc-date-input.modus-wc-input.modus-wc-input-bordered:not(:disabled):not(:focus){border-color:var(--modus-wc-color-gray-6)}[data-theme=modus-classic-dark] modus-wc-date .modus-wc-date-input{color-scheme:dark}[data-theme=modus-classic-dark] modus-wc-date .modus-wc-date-input.modus-wc-input:focus{border-color:var(--modus-wc-color-highlight-blue)}[data-theme=connect-light] modus-wc-date .modus-wc-date-input,[data-theme=connect-dark] modus-wc-date .modus-wc-date-input{border-bottom-width:var(--input-bottom-border-width);outline-width:0 !important}[data-theme=connect-light] modus-wc-date .modus-wc-date-input:not(.modus-wc-select,.modus-wc-number-input),[data-theme=connect-dark] modus-wc-date .modus-wc-date-input:not(.modus-wc-select,.modus-wc-number-input){padding:0 var(--modus-wc-spacing-sm)}[data-theme=connect-light] modus-wc-date .modus-wc-date-input:hover,[data-theme=connect-dark] modus-wc-date .modus-wc-date-input:hover{border-bottom-color:var(--modus-wc-color-primary)}[data-theme=connect-light] modus-wc-date .modus-wc-date-input:active,[data-theme=connect-dark] modus-wc-date .modus-wc-date-input:active{border-bottom-color:var(--modus-wc-color-primary)}[data-theme=connect-light] modus-wc-date .modus-wc-date-input:focus,[data-theme=connect-dark] modus-wc-date .modus-wc-date-input:focus{border-bottom-color:var(--modus-wc-color-primary);outline:none}[data-theme=connect-light] modus-wc-date .modus-wc-date-input:focus-within,[data-theme=connect-dark] modus-wc-date .modus-wc-date-input:focus-within{border-bottom-color:var(--modus-wc-color-primary);outline:none}[data-theme=modus-modern-dark] modus-wc-date .calendar-container .calendar-header .nav-btn i,[data-theme=modus-classic-dark] modus-wc-date .calendar-container .calendar-header .nav-btn i{color:var(--modus-wc-color-primary-content)}[data-theme=modus-modern-dark] modus-wc-date .calendar-container .calendar-header .calendar-selects select,[data-theme=modus-classic-dark] modus-wc-date .calendar-container .calendar-header .calendar-selects select{background-color:unset;color:var(--modus-wc-color-base-content);font-weight:600}[data-theme=modus-modern-dark] modus-wc-date .calendar-container .calendar-header .calendar-selects select option,[data-theme=modus-classic-dark] modus-wc-date .calendar-container .calendar-header .calendar-selects select option{background-color:var(--modus-wc-color-black);color:var(--modus-wc-color-white)}";

const MONTH_SHORT_NAMES = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
];
const WEEK_START_DAY_MAP = {
    sunday: 0,
    monday: 1,
    tuesday: 2,
    wednesday: 3,
    thursday: 4,
    friday: 5,
    saturday: 6,
};
const ModusWcDate = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inputBlur = index.createEvent(this, "inputBlur");
        this.inputChange = index.createEvent(this, "inputChange");
        this.inputFocus = index.createEvent(this, "inputFocus");
        this.calendarMonthChange = index.createEvent(this, "calendarMonthChange");
        this.calendarYearChange = index.createEvent(this, "calendarYearChange");
        this.inheritedAttributes = {};
        this.popperInstance = null;
        this.locale = 'en-US';
        /** Show the calendar dropdown */
        this.showCalendar = false;
        /** Calendar state object */
        this.calendar = new DatePickerCalendar();
        /** Currently focused date index in calendar */
        this.focusedDateIndex = -1;
        /** Tracks whether the component currently has focus */
        this.hasFocus = false;
        /** Indicates that the input should have a border. */
        this.bordered = true;
        /** Custom CSS class to apply to the input. */
        this.customClass = '';
        /** Whether the form control is disabled. */
        this.disabled = false;
        /** Whether the value is editable. */
        this.readOnly = false;
        /** A value is required or must be checked for the form to be submittable. */
        this.required = false;
        /** The size of the input. */
        this.size = 'md';
        /** The value of the control. */
        this.value = '';
        /** The first day of the week for the calendar display */
        this.weekStartDay = 'sunday';
        /** Displays ISO 8601 week numbers in the calendar. Week numbers are calculated with Monday as the first day of the week. */
        this.showWeekNumbers = false;
        this.handleBlur = (event) => {
            // Check if focus is moving to an element within the component
            const relatedTarget = event.relatedTarget;
            // istanbul ignore next (unreachable code)
            if (relatedTarget && this.el.contains(relatedTarget)) {
                // Focus is moving within the component, don't emit blur
                return;
            }
            // Focus is leaving the component
            this.hasFocus = false;
            this.syncValueFromInput();
            this.inputBlur.emit(event);
        };
        this.handleFocus = (event) => {
            // Only emit focus if component didn't already have focus
            if (!this.hasFocus) {
                this.hasFocus = true;
                this.inputFocus.emit(event);
            }
        };
        this.handleInput = (event) => {
            var _a, _b;
            const rawValue = (_b = (_a = event.target) === null || _a === void 0 ? void 0 : _a.value) !== null && _b !== void 0 ? _b : '';
            const parsed = this.parseISODate(rawValue);
            const isoValue = parsed ? this.formatISODate(this.clampDate(parsed)) : '';
            this.inputChange.emit({
                target: { value: isoValue },
            });
        };
        this.handleInputKeyDown = (event) => {
            if (event.key === 'Enter') {
                event.preventDefault();
                this.syncValueFromInput();
            }
        };
        this.setupPopper = () => {
            if (this.popperInstance) {
                this.popperInstance.destroy();
            }
            this.popperInstance = popper.createPopper(this.inputRef, this.calendarRef, {
                placement: 'bottom-start',
                strategy: 'fixed',
                modifiers: [
                    {
                        name: 'offset',
                        options: {
                            offset: [0, 8],
                        },
                    },
                    {
                        name: 'flip',
                        options: {
                            fallbackPlacements: ['top-start', 'bottom-end', 'top-end'],
                        },
                    },
                ],
            });
        };
        this.toggleCalendar = () => {
            this.showCalendar = !this.showCalendar;
            // If opening the calendar and there's a selected date, navigate to it
            if (this.showCalendar) {
                const selectedDate = this.parseISODate(this.value);
                this.ensureCalendarWithinBounds(selectedDate);
                // Set focus to the selected date if it exists
                if (selectedDate) {
                    const selectedIndex = this.calendar.dates.findIndex((date) => date && this.compareDate(date, selectedDate) === 0);
                    if (selectedIndex !== -1) {
                        this.focusedDateIndex = selectedIndex;
                    }
                }
                // set focus to today
                else {
                    this.ensureCalendarWithinBounds(new Date());
                    this.focusedDateIndex = this.calendar.dates.findIndex((date) => date && this.compareDate(date, new Date()) === 0);
                }
            }
            else {
                // Reset focus when closing
                this.focusedDateIndex = -1;
            }
            // Always ensure input is focused when toggling calendar (opening or closing)
            if (this.inputRef) {
                this.inputRef.focus();
            }
        };
        this.handleDateSelect = (date) => {
            if (this.isDateDisabled(date)) {
                return;
            }
            // Clear hasFocus before setting value so the @Watch('value') handler
            // takes the full validation path and dispatches the input event.
            this.hasFocus = false;
            this.value = this.formatISODate(date);
            // If the selected date is from a different month, navigate to that month
            // istanbul ignore next (unreachable code)
            if (date.getMonth() !== this.calendar.selectedMonth ||
                date.getFullYear() !== this.calendar.selectedYear) {
                const firstDayOfWeek = WEEK_START_DAY_MAP[this.weekStartDay || 'sunday'];
                const newCalendar = new DatePickerCalendar(firstDayOfWeek);
                newCalendar.gotoDate(date.getFullYear(), date.getMonth());
                this.calendar = newCalendar;
            }
            this.showCalendar = false;
            this.inputBlur.emit(new FocusEvent('blur', { bubbles: true }));
        };
        this.addMonthOffset = (offset) => {
            const target = new Date(this.calendar.selectedYear, this.calendar.selectedMonth + offset, 1);
            this.updateCalendarAndEmitEvents(target.getFullYear(), target.getMonth());
        };
        this.handleMonthChange = (event) => {
            event.stopPropagation();
            // Try to get the value from the original input event
            const inputEvent = event.detail;
            const selectTarget = inputEvent === null || inputEvent === void 0 ? void 0 : inputEvent.target;
            const monthValue = selectTarget === null || selectTarget === void 0 ? void 0 : selectTarget.value;
            const newMonth = parseInt(monthValue || '0', 10);
            const currentYear = this.calendar.selectedYear;
            if (Number.isNaN(newMonth)) {
                return;
            }
            this.updateCalendarAndEmitEvents(currentYear, newMonth);
        };
        this.handleYearChange = (event) => {
            event.stopPropagation();
            // Try to get the value from the original input event
            const inputEvent = event.detail;
            const selectTarget = inputEvent === null || inputEvent === void 0 ? void 0 : inputEvent.target;
            const yearValue = selectTarget === null || selectTarget === void 0 ? void 0 : selectTarget.value;
            const newYear = parseInt(yearValue || '0', 10);
            const currentMonth = this.calendar.selectedMonth;
            if (Number.isNaN(newYear)) {
                return;
            }
            this.updateCalendarAndEmitEvents(newYear, currentMonth);
        };
        this.handleDateKeyDown = (event, date) => {
            if (this.isDateDisabled(date)) {
                return;
            }
            if (event.key === 'Enter' || event.key === ' ') {
                event.preventDefault();
                this.handleDateSelect(date);
            }
        };
    }
    /** Re-displays the stored ISO value in the new format when the `format` prop changes. */
    handleFormatChange() {
        if (this.value && this.inputRef) {
            const parsed = this.parseISODate(this.value);
            if (parsed) {
                this.inputRef.value = this.formatForDisplay(parsed);
            }
        }
    }
    handleMinChange(newValue) {
        this.minDate = this.parseISODate(newValue);
        if (this.maxDate && this.minDate && this.minDate > this.maxDate) {
            this.maxDate = this.cloneDate(this.minDate);
        }
        this.ensureValueWithinBounds();
    }
    handleMaxChange(newValue) {
        this.maxDate = this.parseISODate(newValue);
        if (this.minDate && this.maxDate && this.maxDate < this.minDate) {
            this.minDate = this.cloneDate(this.maxDate);
        }
        this.ensureValueWithinBounds();
    }
    handleValueChange(newValue) {
        if (newValue === undefined) {
            return;
        }
        if (!newValue) {
            if (this.inputRef) {
                this.inputRef.value = '';
            }
            return;
        }
        // When the input has focus, the user is actively typing.
        // Allow partial/incomplete values to pass through without validation
        // so that controlled input patterns (e.g. React) work correctly.
        // Only reformat strict ISO 8601 values set programmatically; all other
        // typed/partial input passes through unchanged.
        if (this.hasFocus) {
            const isISO = /^\d{4}-\d{2}-\d{2}$/.test(newValue);
            if (isISO) {
                const parsed = this.parseISODate(newValue);
                if (parsed) {
                    // Clamp for display only — do not write back to this.value to avoid
                    // re-triggering this watcher and risking a loop in controlled-input patterns.
                    if (this.inputRef) {
                        this.inputRef.value = this.formatForDisplay(this.clampDate(parsed));
                    }
                }
                else if (this.inputRef) {
                    this.inputRef.value = newValue;
                }
            }
            else if (this.inputRef) {
                this.inputRef.value = newValue;
            }
            return;
        }
        // Prop-driven change (no focus). Parse and update the display.
        // The parent is responsible for providing a valid value; we clamp only
        // the display here and never write back to this.value from the watcher.
        const parsed = this.parseISODate(newValue);
        if (!parsed) {
            if (this.inputRef) {
                this.inputRef.value = '';
            }
            return;
        }
        const clamped = this.clampDate(parsed);
        if (this.inputRef) {
            this.inputRef.value = this.formatForDisplay(clamped);
            const event = new Event('input', { bubbles: true });
            this.inputRef.dispatchEvent(event);
        }
        this.ensureCalendarWithinBounds(clamped);
    }
    handleWeekStartDayChange() {
        // Reinitialize calendar with new first day of week
        const firstDayOfWeek = WEEK_START_DAY_MAP[this.weekStartDay];
        this.calendar = new DatePickerCalendar(firstDayOfWeek);
        // Navigate to currently selected date if exists
        const selectedDate = this.parseISODate(this.value);
        if (selectedDate) {
            this.calendar.gotoDate(selectedDate.getFullYear(), selectedDate.getMonth());
        }
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        baseComponent.handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Date input';
        }
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
        try {
            this.locale =
                document.documentElement.lang || navigator.language || 'en-US';
        }
        catch (_a) {
            this.locale = 'en-US';
        }
        // Initialize calendar with the correct first day of week
        const firstDayOfWeek = WEEK_START_DAY_MAP[this.weekStartDay];
        this.calendar = new DatePickerCalendar(firstDayOfWeek);
        this.handleMinChange(this.min);
        this.handleMaxChange(this.max);
        this.handleValueChange(this.value);
    }
    componentDidUpdate() {
        if (this.showCalendar && this.inputRef && this.calendarRef) {
            this.setupPopper();
        }
        else if (this.popperInstance) {
            this.popperInstance.destroy();
            this.popperInstance = null;
        }
    }
    disconnectedCallback() {
        if (this.popperInstance) {
            this.popperInstance.destroy();
            this.popperInstance = null;
        }
    }
    getClasses() {
        const classList = [
            'modus-wc-date-input',
            'modus-wc-input',
            'modus-wc-w-full',
        ];
        const propClasses = convertPropsToClasses({
            bordered: this.bordered,
            feedback: this.feedback,
            readOnly: this.readOnly,
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    handleClickOutside(event) {
        const path = event.composedPath();
        const insideComponent = path.includes(this.el);
        if (!insideComponent && this.showCalendar) {
            this.showCalendar = false;
            this.hasFocus = false;
        }
    }
    handleEscapeKey(event) {
        if (event.key === 'Escape' && this.showCalendar) {
            this.showCalendar = false;
            event.preventDefault();
        }
    }
    navigateToAdjacentMonth(currentIndex, isUp) {
        var _a, _b, _c;
        const currentColumn = currentIndex % 7;
        // Navigate to previous/next month
        // Date constructor will normalize out-of-bounds months (e.g., -1 → Dec of prev year, 12 → Jan of next year)
        this.updateCalendarAndEmitEvents(this.calendar.selectedYear, this.calendar.selectedMonth + (isUp ? -1 : 1));
        // Find target date in same column
        const weekRange = isUp ? [5, 4, 3, 2, 1, 0] : [0, 1, 2, 3, 4, 5];
        for (const week of weekRange) {
            const indexInWeek = week * 7 + currentColumn;
            // istanbul ignore next (optional chaining)
            if (indexInWeek < this.calendar.dates.length &&
                ((_a = this.calendar.dates[indexInWeek]) === null || _a === void 0 ? void 0 : _a.getMonth()) ===
                    this.calendar.selectedMonth) {
                this.focusedDateIndex = indexInWeek;
                return;
            }
        }
        // Fallback to first/last current-month date
        // istanbul ignore next (fallback scenario)
        const currentMonthIndices = this.calendar.dates
            .map((date, index) => (date === null || date === void 0 ? void 0 : date.getMonth()) === this.calendar.selectedMonth ? index : -1)
            .filter((index) => index !== -1);
        // istanbul ignore next (fallback scenario)
        this.focusedDateIndex = isUp
            ? ((_b = currentMonthIndices[currentMonthIndices.length - 1]) !== null && _b !== void 0 ? _b : this.calendar.dates.length - 1)
            : ((_c = currentMonthIndices[0]) !== null && _c !== void 0 ? _c : 0);
    }
    handleArrowKeys(event) {
        var _a;
        if (!this.showCalendar) {
            return;
        }
        const key = event.key;
        if (!['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight'].includes(key)) {
            return;
        }
        event.preventDefault();
        const totalDates = this.calendar.dates.length;
        let newIndex = this.focusedDateIndex;
        // If no date is focused, start with the first date or selected date
        if (newIndex === -1) {
            if (this.value) {
                const selectedDate = this.parseISODate(this.value);
                if (selectedDate) {
                    newIndex = this.calendar.dates.findIndex((date) => this.compareDate(date, selectedDate) === 0);
                }
            }
            // istanbul ignore next (unreachable code)
            if (newIndex === -1) {
                newIndex = 0;
            }
        }
        // Calculate target position for each arrow key
        let targetIndex = newIndex;
        let shouldChangeMonth = false;
        let targetDate = null;
        // Navigate based on arrow key
        // istanbul ignore next (unreachable code)
        switch (key) {
            case 'ArrowLeft':
                targetIndex = newIndex - 1;
                break;
            case 'ArrowRight':
                targetIndex = newIndex + 1;
                break;
            case 'ArrowUp':
                targetIndex = newIndex - 7;
                break;
            case 'ArrowDown':
                targetIndex = newIndex + 7;
                break;
        }
        // Check if target index is valid and get the target date
        if (targetIndex >= 0 && targetIndex < totalDates) {
            targetDate = this.calendar.dates[targetIndex];
            if (targetDate) {
                // Skip disabled dates - keep moving in the same direction until we find a valid date
                let searchIndex = targetIndex;
                const direction = key === 'ArrowLeft'
                    ? -1
                    : key === 'ArrowRight'
                        ? 1
                        : key === 'ArrowUp'
                            ? -7
                            : 7;
                while (searchIndex >= 0 &&
                    searchIndex < totalDates &&
                    this.isDateDisabled(this.calendar.dates[searchIndex])) {
                    searchIndex += direction;
                }
                // If we found a valid date within bounds
                if (searchIndex >= 0 &&
                    searchIndex < totalDates &&
                    this.calendar.dates[searchIndex]) {
                    targetDate = this.calendar.dates[searchIndex];
                    targetIndex = searchIndex;
                }
                else {
                    // No valid date found in this direction, don't move
                    return;
                }
                // If target date is from a different month, navigate to that month
                // istanbul ignore next (optional chaining)
                if (targetDate.getMonth() !== this.calendar.selectedMonth) {
                    shouldChangeMonth = true;
                }
                this.focusedDateIndex = targetIndex;
            }
        }
        else {
            // Target is outside current calendar, navigate to appropriate month
            shouldChangeMonth = true;
            if (key === 'ArrowUp') {
                // Check if we can navigate to previous month
                const prevMonthDate = new Date(this.calendar.selectedYear, this.calendar.selectedMonth - 1, 1);
                if (!this.isDateDisabled(prevMonthDate)) {
                    this.navigateToAdjacentMonth(newIndex, true);
                    shouldChangeMonth = false; // Already handled in helper
                }
            }
            else if (key === 'ArrowDown') {
                // Check if we can navigate to next month
                const nextMonthDate = new Date(this.calendar.selectedYear, this.calendar.selectedMonth + 1, 1);
                if (!this.isDateDisabled(nextMonthDate)) {
                    this.navigateToAdjacentMonth(newIndex, false);
                    shouldChangeMonth = false; // Already handled in helper
                }
            }
            else if (key === 'ArrowLeft') {
                // Go to previous month's last day
                const prevMonthDate = new Date(this.calendar.selectedYear, this.calendar.selectedMonth - 1, 1);
                targetDate = new Date(prevMonthDate.getFullYear(), prevMonthDate.getMonth() + 1, 0); // Last day of previous month
                // Only navigate if not disabled
                if (this.isDateDisabled(targetDate)) {
                    return;
                }
            }
            else {
                // Go to next month's first day
                targetDate = new Date(this.calendar.selectedYear, this.calendar.selectedMonth + 1, 1); // First day of next month
                // Only navigate if not disabled
                if (this.isDateDisabled(targetDate)) {
                    return;
                }
            }
        }
        // Handle month change if needed
        if (shouldChangeMonth && targetDate) {
            this.updateCalendarAndEmitEvents(targetDate.getFullYear(), targetDate.getMonth());
            // Find the target date in the new calendar
            const newTargetIndex = this.calendar.dates.findIndex(
            // istanbul ignore next (optional chaining)
            (date) => date && this.compareDate(date, targetDate) === 0);
            // istanbul ignore next (inequality check)
            if (newTargetIndex !== -1) {
                this.focusedDateIndex = newTargetIndex;
            }
            else {
                // Fallback positioning
                // istanbul ignore next (fallback scenario)
                if (key === 'ArrowLeft' || key === 'ArrowUp') {
                    // Focus on last current-month date
                    // istanbul ignore next (fallback scenario)
                    const lastCurrentMonthIndex = this.calendar.dates
                        .map((date, index) => date && date.getMonth() === targetDate.getMonth() ? index : -1)
                        .filter((index) => index !== -1)
                        .pop();
                    // istanbul ignore next (fallback scenario)
                    this.focusedDateIndex =
                        lastCurrentMonthIndex !== undefined
                            ? lastCurrentMonthIndex
                            : this.calendar.dates.length - 1;
                }
                else {
                    // Focus on first current-month date
                    // istanbul ignore next (fallback scenario)
                    const firstCurrentMonthIndex = this.calendar.dates.findIndex((date) => date && date.getMonth() === targetDate.getMonth());
                    // istanbul ignore next (fallback scenario)
                    this.focusedDateIndex =
                        firstCurrentMonthIndex !== -1 ? firstCurrentMonthIndex : 0;
                }
            }
        }
        // Focus the corresponding button
        // istanbul ignore next (unreachable code)
        const dateButtons = (_a = this.calendarRef) === null || _a === void 0 ? void 0 : _a.querySelectorAll('.calendar-day');
        if (dateButtons && dateButtons[this.focusedDateIndex]) {
            // istanbul ignore next (unreachable code)
            dateButtons[this.focusedDateIndex].focus();
        }
    }
    renderCalendarHeader() {
        const currentYear = this.calendar.selectedYear;
        const currentMonth = this.calendar.selectedMonth;
        // Generate year options (current year ± 100 years)
        const yearOptions = [];
        for (let i = currentYear - 100; i <= currentYear + 100; i++) {
            yearOptions.push({ value: i.toString(), label: i.toString() });
        }
        // Generate month options
        const monthOptions = MONTH_SHORT_NAMES.map((month, index) => ({
            value: index.toString(),
            label: month,
        }));
        return (index.h("div", { class: "calendar-header" }, index.h("modus-wc-button", { type: "button", "aria-label": "Previous", variant: "borderless", shape: "circle", size: "xs", onButtonClick: 
            // istanbul ignore next (unreachable code)
            () => this.addMonthOffset(-1), class: "nav-btn" }, index.h("modus-wc-icon", { name: "chevron_left", size: "sm" })), index.h("div", { class: "calendar-selects" }, index.h("modus-wc-select", { key: `month-${currentYear}-${currentMonth}`, class: "month-select", value: currentMonth.toString(), options: monthOptions, onInputChange: 
            // istanbul ignore next (unreachable code)
            (e) => this.handleMonthChange(e), onInputBlur: 
            // istanbul ignore next (unreachable code)
            (e) => e.stopPropagation(), bordered: false, size: "sm" }), index.h("modus-wc-select", { key: `year-${currentYear}`, class: "year-select", value: currentYear.toString(), options: yearOptions, onInputChange: 
            // istanbul ignore next (unreachable code)
            (e) => this.handleYearChange(e), onInputBlur: 
            // istanbul ignore next (unreachable code)
            (e) => e.stopPropagation(), bordered: false, size: "sm" })), index.h("modus-wc-button", { type: "button", "aria-label": "Next", variant: "borderless", shape: "circle", size: "xs", onButtonClick: 
            // istanbul ignore next (unreachable code)
            () => this.addMonthOffset(1), class: "nav-btn" }, index.h("modus-wc-icon", { name: "chevron_right", size: "sm" }))));
    }
    renderCalendarBody() {
        const today = new Date();
        const selectedDate = this.parseISODate(this.value);
        const currentMonth = this.calendar.selectedMonth;
        return (index.h("div", { class: "calendar-body" }, index.h("div", { class: `calendar-days-week${this.showWeekNumbers ? ' has-week-numbers' : ''}` }, this.showWeekNumbers && index.h("div", { class: "week-number-header" }), this.calendar
            .getDaysOfWeek('default', WEEK_START_DAY_MAP[this.weekStartDay])
            .map((d) => {
            return index.h("div", { class: "day-header" }, d);
        })), index.h("div", { class: `calendar-dates${this.showWeekNumbers ? ' has-week-numbers' : ''}` }, this.calendar.dates.map((date, index$1) => {
            // Add week number at the start of each row (every 7 days)
            const weekNumberElement = this.showWeekNumbers && index$1 % 7 === 0 ? (index.h("div", { class: "week-number", "aria-label": `Week ${this.calendar.getWeekNumber(date, WEEK_START_DAY_MAP[this.weekStartDay])}` }, this.calendar.getWeekNumber(date, WEEK_START_DAY_MAP[this.weekStartDay]))) : null;
            if (!date) {
                return weekNumberElement;
            }
            const isToday = this.compareDate(date, today) === 0;
            const isSelected = (selectedDate && this.compareDate(date, selectedDate) === 0) ||
                false;
            const isCurrentMonth = date.getMonth() === currentMonth;
            const isDisabled = this.isDateDisabled(date);
            const button = (index.h("button", { type: "button", class: {
                    'calendar-day': true,
                    'current-day': isToday,
                    selected: isSelected,
                    'current-month': isCurrentMonth,
                    'other-month': !isCurrentMonth,
                    disabled: isDisabled,
                }, disabled: isDisabled, onClick: () => this.handleDateSelect(date), onKeyDown: (e) => this.handleDateKeyDown(e, date), tabIndex: isDisabled ? -1 : 0 }, date.getDate()));
            // Only create array when week number exists
            return weekNumberElement ? [weekNumberElement, button] : button;
        }))));
    }
    compareDate(date1, date2) {
        if (!date1 && !date2) {
            return 0;
        }
        else if (!date1 && date2) {
            return -1;
        }
        else if (date1 && !date2) {
            return 1;
        }
        let delta;
        delta = date1.getFullYear() - date2.getFullYear();
        if (delta !== 0) {
            return delta;
        }
        delta = date1.getMonth() - date2.getMonth();
        if (delta !== 0) {
            return delta;
        }
        return date1.getDate() - date2.getDate();
    }
    get effectiveFormat() {
        return this.format || this.getLocaleFormatGuide();
    }
    /** Generates a localized guide for the placeholder (e.g., "mm/dd/yyyy") */
    getLocaleFormatGuide() {
        const options = {
            year: 'numeric',
            month: '2-digit',
            day: '2-digit',
        };
        const parts = new Intl.DateTimeFormat(this.locale, options).formatToParts(new Date(2026, 11, 31));
        return parts
            .map((part) => {
            switch (part.type) {
                case 'day':
                    return 'dd';
                case 'month':
                    return 'mm';
                case 'year':
                    return 'yyyy';
                default:
                    return part.value;
            }
        })
            .join('');
    }
    /**
     * Parses a date string into a `Date`. Accepts pure ISO 8601 (`YYYY-MM-DD`), abbreviated month
     * name strings matching the `MMM DD, YYYY` token pattern (e.g. `Oct 15, 2025`), and any
     * numeric format whose day/month/year order is resolved from `this.format` or the locale guide.
     */
    parseISODate(value) {
        if (!value)
            return undefined;
        const isoMatch = value.match(/^(\d{4})-(\d{2})-(\d{2})$/);
        if (isoMatch) {
            const date = new Date(Number(isoMatch[1]), Number(isoMatch[2]) - 1, Number(isoMatch[3]));
            if (date.getFullYear() === Number(isoMatch[1]) &&
                date.getMonth() === Number(isoMatch[2]) - 1 &&
                date.getDate() === Number(isoMatch[3])) {
                return this.cloneDate(date);
            }
            return undefined;
        }
        const guide = this.effectiveFormat;
        // Handle abbreviated month name format (e.g. "MMM DD, YYYY" → "Oct 15, 2025")
        if (guide.includes('MMM')) {
            const mmmMatch = value.match(/^([A-Za-z]{3})\s+(\d{1,2}),?\s+(\d{4})$/);
            if (!mmmMatch)
                return undefined;
            const monthIdx = MONTH_SHORT_NAMES.findIndex((m) => m.toLowerCase() === mmmMatch[1].toLowerCase());
            if (monthIdx === -1)
                return undefined;
            const dayNum = Number(mmmMatch[2]);
            const yearNum = Number(mmmMatch[3]);
            const date = new Date(yearNum, monthIdx, dayNum);
            if (date.getFullYear() === yearNum &&
                date.getMonth() === monthIdx &&
                date.getDate() === dayNum) {
                return this.cloneDate(date);
            }
            return undefined;
        }
        // Extract numbers separated by /, -, or . only (exactly 3 groups required)
        const numbers = value.match(/^(\d+)[/\-.](\d+)[/\-.](\d+)$/);
        if (!numbers)
            return undefined;
        const [n1, n2, n3] = [
            Number(numbers[1]),
            Number(numbers[2]),
            Number(numbers[3]),
        ];
        let day, month, year;
        const guideLower = guide.toLowerCase();
        if (guideLower.startsWith('m')) {
            [month, day, year] = [n1, n2, n3];
            month -= 1;
        }
        else if (guideLower.startsWith('y')) {
            [year, month, day] = [n1, n2, n3];
            month -= 1;
        }
        else {
            [day, month, year] = [n1, n2, n3];
            month -= 1;
        }
        const date = new Date(year, month, day);
        if (date.getFullYear() === year &&
            date.getMonth() === month &&
            date.getDate() === day) {
            return this.cloneDate(date);
        }
        return undefined;
    }
    /** Formats date as ISO 8601 (YYYY-MM-DD) for the value prop */
    formatISODate(date) {
        const year = date.getFullYear();
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const day = String(date.getDate()).padStart(2, '0');
        return `${year}-${month}-${day}`;
    }
    /** Returns the current value formatted for display, or empty string if value is absent or unparseable. */
    get inputDisplayValue() {
        if (!this.value)
            return '';
        const parsed = this.parseISODate(this.value);
        return parsed ? this.formatForDisplay(parsed) : '';
    }
    /** Formats date for display in the input using the selected format pattern */
    formatForDisplay(date) {
        const fmt = this.effectiveFormat;
        const day = String(date.getDate()).padStart(2, '0');
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const year = String(date.getFullYear());
        const monthName = MONTH_SHORT_NAMES[date.getMonth()];
        const map = {
            yyyy: year,
            YYYY: year,
            mm: month,
            dd: day,
            DD: day,
            MMM: monthName,
        };
        return fmt.replace(/yyyy|YYYY|mm|dd|DD|MMM/g, (matched) => map[matched]);
    }
    cloneDate(date) {
        return new Date(date.getFullYear(), date.getMonth(), date.getDate());
    }
    clampDate(date) {
        let result = this.cloneDate(date);
        if (this.minDate && result < this.minDate) {
            result = this.cloneDate(this.minDate);
        }
        if (this.maxDate && result > this.maxDate) {
            result = this.cloneDate(this.maxDate);
        }
        return result;
    }
    isDateDisabled(date) {
        if (this.minDate && date < this.minDate) {
            return true;
        }
        if (this.maxDate && date > this.maxDate) {
            return true;
        }
        return false;
    }
    ensureValueWithinBounds() {
        if (!this.value) {
            return;
        }
        const parsed = this.parseISODate(this.value);
        if (!parsed) {
            this.value = '';
            return;
        }
        const clamped = this.clampDate(parsed);
        const formatted = this.formatISODate(clamped);
        if (formatted !== this.value) {
            this.value = formatted;
        }
    }
    ensureCalendarWithinBounds(referenceDate) {
        // Allow viewing any month, just disable dates outside min/max
        if (referenceDate) {
            const firstDayOfWeek = WEEK_START_DAY_MAP[this.weekStartDay];
            const newCalendar = new DatePickerCalendar(firstDayOfWeek);
            newCalendar.gotoDate(referenceDate.getFullYear(), referenceDate.getMonth());
            this.calendar = newCalendar;
        }
    }
    setCalendarMonth(year, month) {
        const firstDayOfWeek = WEEK_START_DAY_MAP[this.weekStartDay];
        const newCalendar = new DatePickerCalendar(firstDayOfWeek);
        newCalendar.gotoDate(year, month);
        this.calendar = newCalendar;
    }
    updateCalendarAndEmitEvents(year, month) {
        const oldYear = this.calendar.selectedYear;
        const oldMonth = this.calendar.selectedMonth;
        this.setCalendarMonth(year, month);
        // Emit events only if the values actually changed
        if (month !== oldMonth) {
            this.calendarMonthChange.emit(month);
        }
        if (year !== oldYear) {
            this.calendarYearChange.emit(year);
        }
    }
    syncValueFromInput() {
        if (!this.inputRef) {
            return;
        }
        const value = this.inputRef.value.trim();
        if (!value) {
            if (this.value) {
                this.value = '';
            }
            return;
        }
        const parsed = this.parseISODate(value);
        if (!parsed) {
            this.inputRef.value = this.inputDisplayValue;
            return;
        }
        const clamped = this.clampDate(parsed);
        this.value = this.formatISODate(clamped);
        this.inputRef.value = this.formatForDisplay(clamped);
    }
    render() {
        var _a, _b;
        return (index.h(index.Host, { key: 'ad188591ea7a87365f804252db6790e752d08391' }, this.label && (index.h("modus-wc-input-label", { key: '113a0f41188e7f0cf8eecea7916cf4903d3eacd0', forId: this.inputId, labelText: this.label, required: this.required, size: this.size })), index.h("div", { key: 'db03a6561049215aa4b29e89e0b091f348cfc283', class: "date-input-container" }, index.h("input", Object.assign({ key: '9b4f577fbf354fb3bb831d414c02282d65ffda69', ref: (el) => (this.inputRef = el), "aria-disabled": this.disabled, class: this.getClasses(), disabled: this.disabled, id: this.inputId, name: this.name, onBlur: this.handleBlur, onFocus: this.handleFocus, onInput: this.handleInput, onKeyDown: this.handleInputKeyDown, placeholder: this.effectiveFormat, readonly: this.readOnly, required: this.required, tabIndex: this.inputTabIndex, type: "text", value: this.hasFocus
                ? ((_b = (_a = this.inputRef) === null || _a === void 0 ? void 0 : _a.value) !== null && _b !== void 0 ? _b : '')
                : this.inputDisplayValue }, this.inheritedAttributes)), index.h("modus-wc-button", { key: 'beb3e7fff09a2bcd546f9c3bf8f1a495c8a51ea4', "aria-label": "Open calendar", disabled: this.disabled || this.readOnly, variant: "borderless", shape: "circle", size: "xs", color: "tertiary", class: "calendar-icon-button", onButtonClick: 
            // istanbul ignore next (unreachable code)
            () => this.toggleCalendar() }, index.h("modus-wc-icon", { key: '581777095a613a448fc2270b05af839468b1e205', name: "calendar_blank", size: "sm" }))), this.showCalendar && (index.h("div", { key: 'ad7daaa70009a975c071942f684038d9e9fa8283', ref: (el) => (this.calendarRef = el), class: `calendar-container${this.showWeekNumbers ? ' has-week-numbers' : ''}` }, this.renderCalendarHeader(), this.renderCalendarBody())), this.feedback && (index.h("modus-wc-input-feedback", { key: 'e41d0c5f861abf8d9519c0e70671a416390897d5', level: this.feedback.level, message: this.feedback.message, size: this.size }))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "format": ["handleFormatChange"],
        "min": ["handleMinChange"],
        "max": ["handleMaxChange"],
        "value": ["handleValueChange"],
        "weekStartDay": ["handleWeekStartDayChange"]
    }; }
};
ModusWcDate.style = modusWcDateCss;

exports.modus_wc_date = ModusWcDate;
