// @trimble-oss/moduswebcomponents/cjs/modus-wc-radio.cjs.entry.js
'use strict';

var index = require('./index-B47t6HmQ.js');
var baseComponent = require('./base-component-ZxJWB8zi.js');
var constants = require('./constants-DmPatsAz.js');
var utils = require('./utils-DiibENGv.js');

const convertPropsToClasses = ({ size, }) => {
    let classes = '';
    if (size) {
        classes = `${classes} modus-wc-radio-${size}`;
    }
    return classes.trim();
};

const modusWcRadioCss = "modus-wc-radio.modus-wc-radio-host{align-items:center;display:flex}modus-wc-radio.modus-wc-radio-host .modus-wc-input-label{font-weight:var(--modus-wc-font-weight-normal);padding-inline-start:var(--modus-wc-spacing-md)}modus-wc-radio.modus-wc-radio-host [type=radio].modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-primary)}modus-wc-radio.modus-wc-radio-host [type=radio].modus-wc-radio:focus{outline-color:var(--modus-wc-color-primary)}modus-wc-radio.modus-wc-radio-host [type=radio].modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-primary)}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio{--fallback-bc:var(--modus-wc-color-gray-4);border-width:var(--modus-wc-border-width-sm)}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-xs,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-xs{height:0.875rem;width:0.875rem}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-sm,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-sm{height:1rem;width:1rem}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-md,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-md{height:1.125rem;width:1.125rem}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-lg,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-lg{height:1.25rem;width:1.25rem}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio:focus{outline-color:var(--modus-wc-color-trimble-blue)}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-highlight-blue)}[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio:focus{outline-color:var(--modus-wc-color-highlight-blue)}[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=connect-light] modus-wc-radio .modus-wc-radio,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio{--fallback-bc:var(--modus-wc-color-gray-4);border-width:var(--modus-wc-border-width-sm)}[data-theme=connect-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-xs,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-xs{height:0.875rem;width:0.875rem}[data-theme=connect-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-sm,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-sm{height:1rem;width:1rem}[data-theme=connect-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-md,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-md{height:1.25rem;width:1.25rem}[data-theme=connect-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-lg,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-lg{height:1.75rem;width:1.75rem}[data-theme=connect-light] modus-wc-radio .modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=connect-light] modus-wc-radio .modus-wc-radio:focus{outline-color:var(--modus-wc-color-trimble-blue)}[data-theme=connect-light] modus-wc-radio .modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=connect-dark] modus-wc-radio .modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-highlight-blue)}[data-theme=connect-dark] modus-wc-radio .modus-wc-radio:focus{outline-color:var(--modus-wc-color-highlight-blue)}[data-theme=connect-dark] modus-wc-radio .modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-trimble-blue)}";

const ModusWcRadio = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.inputBlur = index.createEvent(this, "inputBlur");
        this.inputChange = index.createEvent(this, "inputChange");
        this.inputFocus = index.createEvent(this, "inputFocus");
        this.inheritedAttributes = {};
        this.generatedId = utils.generateElementId();
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The disabled state of the radio. */
        this.disabled = false;
        /** Name of the form control. Submitted with the form as part of a name/value pair. */
        this.name = '';
        /** A value is required for the form to be submittable. */
        this.required = false;
        /** The size of the input. */
        this.size = 'md';
        /** The value of the radio. */
        this.value = false;
        this.handleBlur = (event) => {
            this.inputBlur.emit(event);
        };
        this.handleFocus = (event) => {
            this.inputFocus.emit(event);
        };
        this.handleInput = (event) => {
            this.inputChange.emit(event);
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        baseComponent.handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Radio button';
        }
        this.inheritedAttributes = utils.inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-radio'];
        const propClasses = convertPropsToClasses({ size: this.size });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        const labelSize = this.size && constants.DAISY_TO_MODUS_LABEL_SIZE[this.size];
        const effectiveId = this.inputId || this.generatedId;
        return (index.h(index.Host, { key: '73fd40e4f2b99da5d0f16457662e727c1fe357d0', class: "modus-wc-radio-host" }, index.h("input", Object.assign({ key: '3997d51c68ec38bcb882512e589ecdc1be1a020d', "aria-checked": this.value, "aria-disabled": this.disabled, checked: this.value, class: this.getClasses(), disabled: this.disabled, id: effectiveId, name: this.name, onBlur: this.handleBlur, onFocus: this.handleFocus, onInput: this.handleInput, required: this.required, tabIndex: this.inputTabIndex, type: "radio" }, this.inheritedAttributes)), this.label && (index.h("modus-wc-input-label", { key: 'aaf9b2c94ad856483850d76e65c7f79e09e0377e', forId: effectiveId, labelText: this.label, required: this.required, size: labelSize }))));
    }
    get el() { return index.getElement(this); }
};
ModusWcRadio.style = modusWcRadioCss;

exports.modus_wc_radio = ModusWcRadio;
