// @trimble-oss/moduswebcomponents/collection/styles/themes/connect.js
import { common } from "./common";
const commonOverrides = Object.assign({}, common);
export const connect = {
    light: Object.assign(Object.assign({}, commonOverrides), { '--rounded-box': '0', '--rounded-btn': '2px', '--btn-focus-scale': '1', '--input-radius': '2px', '--rounded-badge': '2px', '--alert-border-left-width': '0', '--alert-border-width': '0', '--alert-border-radius': '4px', '--input-bottom-border-width': '2px', '--tab-radius': '0',
        // Primary Colors (Primary Palette)
        primary: '#005F9E', 'primary-focus': '#00437B', 'primary-content': '#FFFFFF',
        // Secondary Colors (Yellow Progression)
        secondary: '#6A6976', 'secondary-focus': '#474655', 'secondary-content': '#FFFFFF',
        // Tertiary/Accent Colors (Neutral Progression)
        accent: '#6A6E79', 'accent-focus': '#464B52', 'accent-content': '#FFFFFF',
        // Neutral Colors
        neutral: '#CBCDD6', 'neutral-focus': '#A3A6B1', 'neutral-content': '#252A2E',
        // Base Colors
        'base-100': '#FFFFFF', 'base-200': '#CBCDD6', 'base-300': '#B7B9C3', 'base-content': '#252A2E',
        // State/Feedback Colors
        info: '#0063A3', 'info-content': '#FFFFFF', success: '#72A544', 'success-content': '#FFFFFF', warning: '#FFBE00', 'warning-content': '#252A2E', error: '#D52A33', 'error-content': '#FFFFFF' }),
    dark: Object.assign(Object.assign({}, commonOverrides), { '--rounded-box': '0', '--rounded-btn': '2px', '--btn-focus-scale': '1', '--input-radius': '2px', '--rounded-badge': '2px', '--alert-border-left-width': '0', '--alert-border-width': '0', '--alert-border-radius': '4px', '--input-bottom-border-width': '2px', '--tab-radius': '0',
        // Primary Colors (Primary Palette)
        primary: '#019aeb', 'primary-focus': '#004F83', 'primary-content': '#000000',
        // Secondary Colors (Gray Progression)
        secondary: '#f1f1f6', 'secondary-focus': '#E49325', 'secondary-content': '#000',
        // Tertiary/Accent Colors (Neutral Progression)
        accent: '#6A6E79', 'accent-focus': '#464B52', 'accent-content': '#FFFFFF',
        // Neutral Colors
        neutral: '#353A40', 'neutral-focus': '#171C1E', 'neutral-content': '#FFFFFF',
        // Base Colors
        'base-100': '#171C1E', 'base-200': '#353A40', 'base-300': '#464B52', 'base-content': '#FFFFFF',
        // State/Feedback Colors
        info: '#217CBB', 'info-content': '#FFFFFF', success: '#4EA646', 'success-content': '#000', warning: '#FEC157', 'warning-content': '#252A2E', error: '#E86363', 'error-content': '#000' }),
};
