// @trimble-oss/moduswebcomponents/collection/styles/themes/modus-classic.js
import { common } from "./common";
const commonOverrides = Object.assign({}, common);
export const modusClassic = {
    light: Object.assign(Object.assign({}, commonOverrides), { '--rounded-box': '0.5rem', '--rounded-btn': '0.25rem',
        // Primary Colors (Primary Palette)
        primary: '#0063A3', 'primary-focus': '#004F83', 'primary-content': '#FFFFFF',
        // Secondary Colors (Yellow Progression)
        secondary: '#6A6E79', 'secondary-focus': '#464B52', 'secondary-content': '#FFFFFF',
        // Tertiary/Accent Colors (Neutral Progression)
        accent: '#6A6E79', 'accent-focus': '#464B52', 'accent-content': '#FFFFFF',
        // Neutral Colors
        neutral: '#CBCDD6', 'neutral-focus': '#A3A6B1', 'neutral-content': '#252A2E',
        // Base Colors
        'base-100': '#FFFFFF', 'base-200': '#CBCDD6', 'base-300': '#B7B9C3', 'base-content': '#252A2E',
        // State/Feedback Colors
        info: '#0063A3', 'info-content': '#FFFFFF', success: '#1E8A44', 'success-content': '#FFFFFF', warning: '#FBAD26', 'warning-content': '#252A2E', error: '#DA212C', 'error-content': '#FFFFFF' }),
    dark: Object.assign(Object.assign({}, commonOverrides), { '--rounded-box': '0.5rem', '--rounded-btn': '0.25rem',
        // Primary Colors (Primary Palette)
        primary: '#0063A3', 'primary-focus': '#004F83', 'primary-content': '#FFFFFF',
        // Secondary Colors (Yellow Progression)
        secondary: '#FBAD26', 'secondary-focus': '#E49325', 'secondary-content': '#252A2E',
        // Tertiary/Accent Colors (Neutral Progression)
        accent: '#6A6E79', 'accent-focus': '#464B52', 'accent-content': '#FFFFFF',
        // Neutral Colors
        neutral: '#353A40', 'neutral-focus': '#171C1E', 'neutral-content': '#FFFFFF',
        // Base Colors
        'base-100': '#171C1E', 'base-200': '#353A40', 'base-300': '#464B52', 'base-content': '#FFFFFF',
        // State/Feedback Colors
        info: '#0063A3', 'info-content': '#FFFFFF', success: '#1E8A44', 'success-content': '#FFFFFF', warning: '#FBAD26', 'warning-content': '#252A2E', error: '#DA212C', 'error-content': '#FFFFFF' }),
};
