// @trimble-oss/moduswebcomponents/collection/components/modus-wc-accordion/modus-wc-accordion.js
import { h, Host, } from "@stencil/core";
import { handleShadowDOMStyles } from "../base-component";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable accordion component used for showing and hiding related groups of content.
 *
 * The component supports a `<slot>` called 'content' for injecting `<modus-wc-collapse>` elements. See [Collapse](/docs/components-collapse--docs) docs for additional info.
 */
export class ModusWcAccordion {
    constructor() {
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = [''];
        // The order CSS classes are added matters to CSS specificity
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: 'cd50dc732b2fc48835c00a151a3bd53baea5e45f' }, h("div", Object.assign({ key: 'e5449efc6472a41424ab3eb550f6cf6e965f826e', class: this.getClasses() }, this.inheritedAttributes), h("slot", { key: '1c32e9a0eb77ad1d5ee753f29b4bc78f881c0b48' }))));
    }
    static get is() { return "modus-wc-accordion"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-accordion.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-accordion.css"]
        };
    }
    static get properties() {
        return {
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the inner div."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            }
        };
    }
    static get events() {
        return [{
                "method": "expandedChange",
                "name": "expandedChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "When a collapse expanded state is changed, this event outputs the relevant index and state"
                },
                "complexType": {
                    "original": "{\n    expanded: boolean;\n    index: number;\n  }",
                    "resolved": "{ expanded: boolean; index: number; }",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
}
