// @trimble-oss/moduswebcomponents/collection/components/modus-wc-panel/modus-wc-panel.js
import { h, Host } from "@stencil/core";
import { inheritAriaAttributes } from "../utils";
import { convertPropsToClasses } from "./modus-wc-panel.tailwind";
/**
 * A customizable panel component used to organize content in a structured layout.
 *
 * This component provides 'header', 'body', and 'footer' `<slot>` elements for inserting custom HTML.
 */
export class ModusWcPanel {
    constructor() {
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the outer div. */
        this.customClass = '';
        /** Width of the panel in pixels. */
        this.width = '350px';
        /** Height of the panel in pixels. */
        this.height = '700px';
        /** Enable floating mode with elevated shadow. */
        this.floating = false;
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-panel'];
        const propClasses = convertPropsToClasses({
            floating: this.floating,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: '8eca46effdba5d4c7c027a78d98e8914247d0187' }, h("div", Object.assign({ key: '106cf95c53ce95e4afa6a58845b979d575b3b318', class: this.getClasses(), style: { width: this.width, height: this.height } }, this.inheritedAttributes), h("div", { key: '200d9a028744c3027fea736461d50c27eca1ad31', class: "modus-wc-panel-header" }, h("slot", { key: '6ef2682a8d4eeb0f253ac6602da9a980b026be84', name: "header" })), h("div", { key: '61b44ce89e87e1364a18fb366f464ee25881a5c5', class: "modus-wc-panel-body" }, h("slot", { key: '4c6c82a0ad09e81b74dbe9c478f5793eaac702ee', name: "body" })), h("div", { key: '2806ffb3225c9c361b55b12e9088cb4326cbf463', class: "modus-wc-panel-footer" }, h("slot", { key: '577fadfaeb76f7c2f475685756b7da6dbb5c05fb', name: "footer" })))));
    }
    static get is() { return "modus-wc-panel"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-panel.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-panel.css"]
        };
    }
    static get properties() {
        return {
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the outer div."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "width": {
                "type": "string",
                "attribute": "width",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Width of the panel in pixels."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'350px'"
            },
            "height": {
                "type": "string",
                "attribute": "height",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Height of the panel in pixels."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'700px'"
            },
            "floating": {
                "type": "boolean",
                "attribute": "floating",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Enable floating mode with elevated shadow."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "el"; }
}
