// @trimble-oss/moduswebcomponents/components/modus-wc-accordion.js
import { p as proxyCustomElement, H, d as createEvent, h, c as Host } from './p-X1tirp06.js';
import { h as handleShadowDOMStyles } from './p-CKNBKeF3.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';

const modusWcAccordionCss = "[data-theme=modus-classic-light] modus-wc-accordion modus-wc-collapse:first-child .modus-wc-collapse,[data-theme=modus-classic-dark] modus-wc-accordion modus-wc-collapse:first-child .modus-wc-collapse{border-bottom-left-radius:0;border-bottom-right-radius:0}[data-theme=modus-classic-light] modus-wc-accordion modus-wc-collapse:not(:first-child):not(:last-child) .modus-wc-collapse,[data-theme=modus-classic-dark] modus-wc-accordion modus-wc-collapse:not(:first-child):not(:last-child) .modus-wc-collapse{border-radius:0}[data-theme=modus-classic-light] modus-wc-accordion modus-wc-collapse:last-child .modus-wc-collapse,[data-theme=modus-classic-dark] modus-wc-accordion modus-wc-collapse:last-child .modus-wc-collapse{border-top-left-radius:0;border-top-right-radius:0}[data-theme=modus-classic-light] modus-wc-accordion .modus-wc-collapse{border:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-gray-2)}[data-theme=modus-classic-dark] modus-wc-accordion .modus-wc-collapse{border:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-gray-8)}";

const ModusWcAccordion$1 = /*@__PURE__*/ proxyCustomElement(class ModusWcAccordion extends H {
    constructor() {
        super();
        this.__registerHost();
        this.expandedChange = createEvent(this, "expandedChange");
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = [''];
        // The order CSS classes are added matters to CSS specificity
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: 'cd50dc732b2fc48835c00a151a3bd53baea5e45f' }, h("div", Object.assign({ key: 'e5449efc6472a41424ab3eb550f6cf6e965f826e', class: this.getClasses() }, this.inheritedAttributes), h("slot", { key: '1c32e9a0eb77ad1d5ee753f29b4bc78f881c0b48' }))));
    }
    get el() { return this; }
    static get style() { return modusWcAccordionCss; }
}, [4, "modus-wc-accordion", {
        "customClass": [1, "custom-class"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-accordion"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-accordion":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcAccordion$1);
            }
            break;
    } });
}

const ModusWcAccordion = ModusWcAccordion$1;
const defineCustomElement = defineCustomElement$1;

export { ModusWcAccordion, defineCustomElement };
