// @trimble-oss/moduswebcomponents/components/modus-wc-stepper.js
import { p as proxyCustomElement, H, h, c as Host } from './p-X1tirp06.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';

const convertPropsToClasses = ({ orientation, }) => {
    let classes = '';
    if (orientation) {
        switch (orientation) {
            case 'horizontal':
                break; // Default
            case 'vertical':
                classes = `${classes} modus-wc-steps-vertical`;
                break;
        }
    }
    return classes.trim();
};

const modusWcStepperCss = "modus-wc-stepper .modus-wc-steps .modus-wc-step{--fallback-b3:var(--modus-wc-color-base-200)}";

const ModusWcStepper$1 = /*@__PURE__*/ proxyCustomElement(class ModusWcStepper extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the steps element. */
        this.customClass = '';
        /** The steps to display. */
        this.steps = [];
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-steps'];
        const propClasses = convertPropsToClasses({
            orientation: this.orientation,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getClassesForStep(step) {
        const classList = ['modus-wc-step'];
        // The order CSS classes are added matters to CSS specificity
        if (step.color)
            classList.push(`modus-wc-step-${step.color}`);
        if (step.customClass)
            classList.push(step.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: '7d8e5badd7eedcdb7fc6a997b94f12008c5bfcac' }, h("div", Object.assign({ key: 'f3ce9e091ca2418bda52cae1661d2dae7d04fadd', class: this.getClasses() }, this.inheritedAttributes), this.steps.map((step, index) => {
            return (h("li", { class: this.getClassesForStep(step), key: index, "data-content": step.content }, step.label));
        }))));
    }
    get el() { return this; }
    static get style() { return modusWcStepperCss; }
}, [0, "modus-wc-stepper", {
        "customClass": [1, "custom-class"],
        "orientation": [1],
        "steps": [16]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-stepper"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-stepper":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcStepper$1);
            }
            break;
    } });
}

const ModusWcStepper = ModusWcStepper$1;
const defineCustomElement = defineCustomElement$1;

export { ModusWcStepper, defineCustomElement };
