// @trimble-oss/moduswebcomponents/components/modus-wc-skeleton.js
import { p as proxyCustomElement, H, h, c as Host } from './p-X1tirp06.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';

const convertPropsToClasses = ({ shape, }) => {
    let classes = '';
    if (shape === 'circle') {
        classes = `${classes} modus-wc-rounded-full`;
    }
    return classes.trim();
};

const modusWcSkeletonCss = "";

const ModusWcSkeleton$1 = /*@__PURE__*/ proxyCustomElement(class ModusWcSkeleton extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The height of the skeleton. */
        this.height = 'var(--modus-wc-line-height-md)';
        /** The shape of the skeleton. */
        this.shape = 'rectangle';
        /** The width of the skeleton. */
        this.width = '100%';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-skeleton'];
        const propClasses = convertPropsToClasses({
            shape: this.shape,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getStyles() {
        return { height: this.height, width: this.width };
    }
    render() {
        return (h(Host, { key: '9d67bdc774912492b42d055e601cc3143fe56fd8' }, h("div", Object.assign({ key: 'd30972478d13d8d5c604077f12c6204345fe8b0c', "aria-hidden": "true", class: this.getClasses(), style: this.getStyles(), tabindex: -1 }, this.inheritedAttributes))));
    }
    get el() { return this; }
    static get style() { return modusWcSkeletonCss; }
}, [0, "modus-wc-skeleton", {
        "customClass": [1, "custom-class"],
        "height": [1],
        "shape": [1],
        "width": [1]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-skeleton"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-skeleton":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcSkeleton$1);
            }
            break;
    } });
}

const ModusWcSkeleton = ModusWcSkeleton$1;
const defineCustomElement = defineCustomElement$1;

export { ModusWcSkeleton, defineCustomElement };
