// @trimble-oss/moduswebcomponents/components/modus-wc-navbar.js
import { h, p as proxyCustomElement, H, d as createEvent, F as Fragment, c as Host } from './p-X1tirp06.js';
import { S as SearchSolidIcon, d as defineCustomElement$4 } from './p-CtRj0TZD.js';
import { h as handleShadowDOMStyles } from './p-CKNBKeF3.js';
import { i as inheritAriaAttributes, c as isLightMode } from './p-CPG9FZRU.js';
import { d as defineCustomElement$e } from './p-AfrNg3-r.js';
import { d as defineCustomElement$d } from './p-DRd0Bis4.js';
import { d as defineCustomElement$c } from './p-Ctlg6TEH.js';
import { d as defineCustomElement$b } from './p-CeZm3cg9.js';
import { d as defineCustomElement$a } from './p-vnbqBzVH.js';
import { d as defineCustomElement$9 } from './p-Bfy1MXxF.js';
import { d as defineCustomElement$8 } from './p-CqPlhPZ4.js';
import { d as defineCustomElement$7 } from './p-CRX4neeC.js';
import { d as defineCustomElement$6 } from './p-DsNHJf57.js';
import { d as defineCustomElement$5 } from './p-BvSTk6wy.js';
import { d as defineCustomElement$3 } from './p-BUr8lm9p.js';
import { d as defineCustomElement$2 } from './p-C5jBPzA7.js';

const AiDarkIcon = ({ className }) => {
    return (h("svg", { fill: "none", viewBox: "0 0 887 982", class: className || '', xmlns: "http://www.w3.org/2000/svg" },
        h("defs", null,
            h("radialGradient", { id: "gradient", cx: "18%", cy: "18%", r: "70%" },
                h("stop", { offset: "0%", "stop-color": "#FF00FF" }),
                h("stop", { offset: "50%", "stop-color": "#9933FF" }),
                h("stop", { offset: "100%", "stop-color": "#0066CC" }))),
        h("path", { d: "m36.76 749.83v231.56l201.3-116.22c-77.25-16.64-147.52-56.92-201.3-115.34zm199.83-634.65-199.83-115.18v230.14c56.05-60.9 128.22-99.28 199.83-114.97m403.73 374.35c0-176.82-143.34-320.16-320.16-320.16s-320.17 143.33-320.17 320.16 143.34 320.16 320.16 320.16 320.16-143.34 320.16-320.16m45.08-114.58c23.68 75.15 23.76 156.75-.59 232.74l201.86-116.54c-9.54-5.51-189.55-109.44-201.26-116.2", fill: "#fff" }),
        h("path", { d: "m320.13 489.53c0 142.28 115.34 257.62 257.62 257.62s257.62-115.34 257.62-257.62-115.34-257.62-257.62-257.62-257.62 115.34-257.62 257.62", fill: "url(#gradient)", transform: "translate(-256, 0)" })));
};

const AiLightIcon = ({ className }) => {
    return (h("svg", { fill: "none", viewBox: "0 0 887 982", class: className || '', xmlns: "http://www.w3.org/2000/svg" },
        h("defs", null,
            h("linearGradient", { id: "gradient", x1: "0%", y1: "0%", x2: "100%", y2: "100%" },
                h("stop", { offset: "20%", "stop-color": "#FF00FF" }),
                h("stop", { offset: "60%", "stop-color": "#0066CC" }),
                h("stop", { offset: "100%", "stop-color": "#0066CC" }))),
        h("path", { d: "m36.76 749.83v231.56l201.3-116.22c-77.25-16.64-147.52-56.92-201.3-115.34z", fill: "#0066CC" }),
        h("path", { d: "m236.59 115.18-199.83-115.18v230.14c56.05-60.9 128.22-99.28 199.83-114.97z", fill: "#FF00FF" }),
        h("path", { d: "m685.40 374.91c23.68 75.15 23.76 156.75-.59 232.74l201.86-116.54c-9.54-5.51-189.55-109.44-201.26-116.2z", fill: "#0066CC" }),
        h("path", { d: "m577.75 489.53c0 142.28-115.34 257.62-257.62 257.62s-257.62-115.34-257.62-257.62 115.34-257.62 257.63-257.62 257.62 115.34 257.62 257.62m62.57-.44c0-176.82-143.34-320.16-320.16-320.16s-320.17 143.33-320.17 320.16 143.34 320.16 320.16 320.16 320.16-143.34 320.16-320.16", fill: "url(#gradient)" })));
};

const AppsSolidIcon = ({ className }) => {
    return (h("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "currentColor", class: `mi-solid mi-apps ${className || ''}`, viewBox: "0 0 24 24" },
        h("path", { d: "M5 8h2c.6 0 1-.4 1-1V5c0-.6-.4-1-1-1H5c-.6 0-1 .4-1 1v2c0 .6.4 1 1 1m6 12h2c.6 0 1-.4 1-1v-2c0-.6-.4-1-1-1h-2c-.6 0-1 .4-1 1v2c0 .6.4 1 1 1m-6 0h2c.6 0 1-.4 1-1v-2c0-.6-.4-1-1-1H5c-.6 0-1 .4-1 1v2c0 .6.4 1 1 1m0-6h2c.6 0 1-.4 1-1v-2c0-.6-.4-1-1-1H5c-.6 0-1 .4-1 1v2c0 .6.4 1 1 1m6 0h2c.6 0 1-.4 1-1v-2c0-.6-.4-1-1-1h-2c-.6 0-1 .4-1 1v2c0 .6.4 1 1 1m5-9v2c0 .6.4 1 1 1h2c.6 0 1-.4 1-1V5c0-.6-.4-1-1-1h-2c-.6 0-1 .4-1 1m-5 3h2c.6 0 1-.4 1-1V5c0-.6-.4-1-1-1h-2c-.6 0-1 .4-1 1v2c0 .6.4 1 1 1m6 6h2c.6 0 1-.4 1-1v-2c0-.6-.4-1-1-1h-2c-.6 0-1 .4-1 1v2c0 .6.4 1 1 1m0 6h2c.6 0 1-.4 1-1v-2c0-.6-.4-1-1-1h-2c-.6 0-1 .4-1 1v2c0 .6.4 1 1 1" })));
};

const HelpSolidIcon = ({ className }) => {
    return (h("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "currentColor", class: `mi-solid mi-help ${className || ''}`, viewBox: "0 0 24 24" },
        h("path", { d: "M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 16.25a1.25 1.25 0 1 1 0-2.5 1.25 1.25 0 0 1 0 2.5m2.82-6.72-.77.79c-.42.43-.69.78-.83 1.34-.07.28-.1.58-.1.92v.25H10.9v-.68c0-.41.07-.81.2-1.21.18-.52.48-1.01.87-1.4l1.07-1.09c.35-.33.49-.81.4-1.32-.09-.52-.5-.97-1.01-1.12a1.467 1.467 0 0 0-1.81.93c-.14.44-.51.73-.94.73h-.26c-.31 0-.59-.14-.78-.39s-.25-.58-.16-.89a3.72 3.72 0 0 1 2.97-2.61c.17-.03.34-.04.51-.04 1.15 0 2.3.63 3.06 1.68 1.13 1.57.72 3.19-.19 4.1Z" })));
};

const MenuSolidIcon = ({ className }) => {
    return (h("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "currentColor", class: `mi-solid mi-menu ${className || ''}`, viewBox: "0 0 24 24" },
        h("path", { d: "M4 18h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1s.45 1 1 1m0-5h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1s.45 1 1 1M3 7c0 .55.45 1 1 1h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1" })));
};

const MoreVerticalSolidIcon = ({ className, }) => {
    return (h("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "currentColor", class: `mi-solid mi-more-vertical ${className || ''}`, viewBox: "0 0 24 24" },
        h("path", { d: "M10.59 10.59c-.78.78-.78 2.05 0 2.83s2.05.78 2.83 0 .78-2.05 0-2.83-2.05-.78-2.83 0m0-7c-.78.78-.78 2.05 0 2.83s2.05.78 2.83 0 .78-2.05 0-2.83-2.05-.78-2.83 0m0 14c-.78.78-.78 2.05 0 2.83s2.05.78 2.83 0 .78-2.05 0-2.83-2.05-.78-2.83 0" })));
};

const NotificationsSolidIcon = ({ className, }) => {
    return (h("svg", { xmlns: "http://www.w3.org/2000/svg", fill: "currentColor", class: `mi-solid mi-notifications ${className || ''}`, viewBox: "0 0 24 24" },
        h("path", { d: "M10 20h4c0 1.1-.9 2-2 2s-2-.9-2-2m9.67-2.69C18.36 15.96 17 13.66 17 10a5 5 0 0 0-4-4.9V4c0-.55-.45-1-1-1s-1 .45-1 1v1.1A5 5 0 0 0 7 10c0 3.66-1.37 5.96-2.67 7.31-.61.64-.16 1.69.72 1.69h13.91c.88 0 1.33-1.05.72-1.69Z" })));
};

const modusWcNavbarCss = "modus-wc-navbar .modus-wc-navbar{background-color:var(--modus-wc-color-base-page);border-radius:0;box-shadow:none;height:56px;min-height:0;padding:0 var(--modus-wc-spacing-md)}modus-wc-navbar .modus-wc-navbar modus-wc-button{display:inline-flex}modus-wc-navbar .modus-wc-navbar modus-wc-button svg{height:24px}modus-wc-navbar .modus-wc-navbar:not(.logo) .modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary{background:var(--modus-wc-color-base-100);border:0;box-shadow:none;color:var(--modus-wc-color-base-content)}modus-wc-navbar .modus-wc-navbar .modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary.logo{background:transparent}modus-wc-navbar .modus-wc-navbar [slot=end] .user,modus-wc-navbar .modus-wc-navbar [slot=end] .apps,modus-wc-navbar .modus-wc-navbar [slot=end] .notifications,modus-wc-navbar .modus-wc-navbar [slot=end] modus-wc-menu,modus-wc-navbar .modus-wc-navbar [slot=start] .main-menu{position:absolute;top:45px;z-index:99}modus-wc-navbar .modus-wc-navbar [slot=end] .hidden.user,modus-wc-navbar .modus-wc-navbar [slot=end] .hidden.apps,modus-wc-navbar .modus-wc-navbar [slot=end] .hidden.notifications,modus-wc-navbar .modus-wc-navbar [slot=end] modus-wc-menu.hidden,modus-wc-navbar .modus-wc-navbar [slot=start] .hidden.main-menu{display:none}modus-wc-navbar .modus-wc-navbar [slot=end] .visible.user,modus-wc-navbar .modus-wc-navbar [slot=end] .visible.apps,modus-wc-navbar .modus-wc-navbar [slot=end] .visible.notifications,modus-wc-navbar .modus-wc-navbar [slot=end] modus-wc-menu.visible,modus-wc-navbar .modus-wc-navbar [slot=start] .visible.main-menu{display:block}modus-wc-navbar .modus-wc-navbar [slot=end],modus-wc-navbar .modus-wc-navbar [slot=center],modus-wc-navbar .modus-wc-navbar [slot=start]{align-items:center;display:flex;position:relative}modus-wc-navbar .modus-wc-navbar [slot=start] modus-wc-button{padding-inline-end:var(--modus-wc-spacing-sm);padding-inline-start:var(--modus-wc-spacing-xs)}modus-wc-navbar .modus-wc-navbar [slot=start] .main-menu{height:calc(100dvh - 56px);left:calc(-1 * var(--modus-wc-spacing-md));min-width:256px}modus-wc-navbar .modus-wc-navbar [slot=start] .logo.modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary{padding:0}modus-wc-navbar .modus-wc-navbar [slot=start] .logo.modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary .modus-wc-logo svg{height:32px;max-width:260px}modus-wc-navbar .modus-wc-navbar [slot=end] modus-wc-button,modus-wc-navbar .modus-wc-navbar [slot=end] modus-wc-text-input{padding-inline-start:var(--modus-wc-spacing-sm)}modus-wc-navbar .modus-wc-navbar [slot=end] button.ai svg{padding:3px 0 3px 3px}modus-wc-navbar .modus-wc-navbar [slot=end] modus-wc-menu{right:var(--modus-wc-spacing-2xl);z-index:99}modus-wc-navbar .modus-wc-navbar [slot=end] .notifications{right:100px}modus-wc-navbar .modus-wc-navbar [slot=end] .apps{right:var(--modus-wc-spacing-2xl)}modus-wc-navbar .modus-wc-navbar [slot=end] .user{right:0}modus-wc-navbar .modus-wc-navbar [slot=end] .user modus-wc-card article.modus-wc-card{align-items:center;padding:var(--modus-wc-spacing-xl);width:max-content}modus-wc-navbar .modus-wc-navbar [slot=end] .user modus-wc-card article.modus-wc-card .initials{background-color:var(--modus-wc-color-black);border-radius:50%;color:var(--modus-wc-color-white);font-weight:var(--modus-wc-font-weight-semibold);padding:var(--modus-wc-spacing-sm)}modus-wc-navbar .modus-wc-navbar [slot=end] .user modus-wc-card article.modus-wc-card .modus-wc-card-body{align-items:center}modus-wc-navbar .modus-wc-navbar [slot=end] .user modus-wc-card article.modus-wc-card modus-wc-button{padding:0}modus-wc-navbar .modus-wc-navbar [slot=end] .user modus-wc-card article.modus-wc-card .sign-out.modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary{color:var(--modus-wc-color-primary)}modus-wc-navbar .modus-wc-navbar [slot=end] .user-button.modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary{background-color:var(--modus-wc-color-black);color:var(--modus-wc-color-white)}modus-wc-navbar .modus-wc-navbar [slot=end] .user-button.modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary .modus-wc-avatar{height:32px;width:32px}[data-theme=modus-modern-light] modus-wc-navbar .modus-wc-navbar{background-size:cover}[data-theme=modus-modern-dark] modus-wc-navbar .modus-wc-navbar{background-size:cover}[data-theme=modus-classic-light] modus-wc-navbar .modus-wc-navbar svg,[data-theme=modus-classic-dark] modus-wc-navbar .modus-wc-navbar svg{height:32px}[data-theme=modus-classic-light] modus-wc-navbar .modus-wc-navbar modus-wc-text-input .modus-wc-text-input.modus-wc-input.modus-wc-input-sm,[data-theme=modus-classic-dark] modus-wc-navbar .modus-wc-navbar modus-wc-text-input .modus-wc-text-input.modus-wc-input.modus-wc-input-sm{font-size:var(--modus-wc-font-size-md);height:var(--modus-wc-input-height-md);line-height:var(--modus-wc-line-height-md);padding:var(--modus-wc-spacing-sm)}[data-theme=modus-classic-dark] modus-wc-navbar .modus-wc-navbar{background-color:var(--modus-wc-color-gray-10)}[data-theme=modus-classic-dark] modus-wc-navbar .modus-wc-navbar:not(.logo) .modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary{background:transparent;color:var(--modus-wc-color-white)}[data-theme=modus-classic-dark] modus-wc-navbar .modus-wc-navbar .logo.modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary path{fill:var(--modus-wc-color-white)}[data-theme=modus-classic-dark] modus-wc-navbar .modus-wc-navbar .user modus-wc-card article.modus-wc-card .sign-out.modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary{color:var(--modus-wc-color-white)}[data-theme=connect-light] modus-wc-navbar .modus-wc-navbar,[data-theme=connect-dark] modus-wc-navbar .modus-wc-navbar{border-bottom:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-base-200)}[data-theme=connect-light] modus-wc-navbar .modus-wc-navbar :not(.logo) .modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary,[data-theme=connect-dark] modus-wc-navbar .modus-wc-navbar :not(.logo) .modus-wc-btn.modus-wc-btn-borderless.modus-wc-btn-primary{background:transparent}";

const ModusWcNavbar$1 = /*@__PURE__*/ proxyCustomElement(class ModusWcNavbar extends H {
    constructor() {
        super();
        this.__registerHost();
        this.aiClick = createEvent(this, "aiClick");
        this.appsClick = createEvent(this, "appsClick");
        this.appsMenuOpenChange = createEvent(this, "appsMenuOpenChange");
        this.condensedMenuOpenChange = createEvent(this, "condensedMenuOpenChange");
        this.helpClick = createEvent(this, "helpClick");
        this.mainMenuOpenChange = createEvent(this, "mainMenuOpenChange");
        this.myTrimbleClick = createEvent(this, "myTrimbleClick");
        this.notificationsClick = createEvent(this, "notificationsClick");
        this.notificationsMenuOpenChange = createEvent(this, "notificationsMenuOpenChange");
        this.searchChange = createEvent(this, "searchChange");
        this.searchClick = createEvent(this, "searchClick");
        this.searchInputOpenChange = createEvent(this, "searchInputOpenChange");
        this.signOutClick = createEvent(this, "signOutClick");
        this.trimbleLogoClick = createEvent(this, "trimbleLogoClick");
        this.userMenuOpenChange = createEvent(this, "userMenuOpenChange");
        this.inheritedAttributes = {};
        this.searchDebounceTimer = null;
        /** The name of the logo to display. Supports any valid 'logo-name' from the 'modus-wc-logo' component. Defaults to 'trimble'. */
        this.logoName = 'trimble';
        /** The open state of the apps menu. */
        this.appsMenuOpen = false;
        /** Applies condensed layout and styling. */
        this.condensed = false;
        /** The open state of the condensed menu. */
        this.condensedMenuOpen = false;
        /** Custom CSS class to apply to the host element. */
        this.customClass = '';
        /** The open state of the main menu. */
        this.mainMenuOpen = false;
        /** The open state of the notifications menu. */
        this.notificationsMenuOpen = false;
        /** Debounce time in milliseconds for search input changes. Default is 300ms. */
        this.searchDebounceMs = 300;
        /** The open state of the search input. */
        this.searchInputOpen = false;
        /** The open state of the user menu. */
        this.userMenuOpen = false;
        /** The visibility of individual navbar buttons. Default is user profile visible, others hidden. */
        this.visibility = {
            ai: false,
            apps: false,
            help: false,
            logo: true,
            mainMenu: false,
            notifications: false,
            search: false,
            searchInput: false,
            user: true,
        };
        this.isLight = true;
        this.searchValue = '';
        this.themeObserver = null;
        this.handleAiClick = (event) => {
            this.aiClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleAppsClick = (event) => {
            this.toggleApps();
            this.appsClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleHelpClick = (event) => {
            if (this.condensed) {
                this.toggleCondensedMenu();
            }
            this.helpClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleMyTrimbleClick = (event) => {
            this.myTrimbleClick.emit(event.detail);
        };
        this.handleNotificationsClick = (event) => {
            this.toggleNotifications();
            this.notificationsClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleSearchChange = (event) => {
            // Update the search value immediately for UI responsiveness
            this.searchValue = event.target.value;
            // Clear any existing timer
            if (this.searchDebounceTimer !== null) {
                window.clearTimeout(this.searchDebounceTimer);
            }
            // Set a new timer
            this.searchDebounceTimer = window.setTimeout(() => {
                this.searchChange.emit({ value: this.searchValue });
                this.searchDebounceTimer = null;
            }, this.searchDebounceMs);
        };
        this.handleSearchClick = (event) => {
            this.toggleSearch();
            this.searchClick.emit(event === null || event === void 0 ? void 0 : event.detail);
        };
        this.handleSignOutClick = (event) => {
            this.signOutClick.emit(event.detail);
        };
        this.handleTrimbleLogoClick = (event) => {
            this.trimbleLogoClick.emit(event.detail);
        };
        this.toggleApps = () => {
            if (this.condensed) {
                this.toggleCondensedMenu();
            }
            else {
                this.appsMenuOpen = !this.appsMenuOpen;
                this.appsMenuOpenChange.emit(this.appsMenuOpen);
            }
        };
        this.toggleCondensedMenu = () => {
            this.condensedMenuOpen = !this.condensedMenuOpen;
            this.condensedMenuOpenChange.emit(this.condensedMenuOpen);
        };
        this.toggleMainMenu = () => {
            this.mainMenuOpen = !this.mainMenuOpen;
            this.mainMenuOpenChange.emit(this.mainMenuOpen);
        };
        this.toggleNotifications = () => {
            if (this.condensed) {
                this.toggleCondensedMenu();
            }
            else {
                this.notificationsMenuOpen = !this.notificationsMenuOpen;
                this.notificationsMenuOpenChange.emit(this.notificationsMenuOpen);
            }
        };
        this.toggleSearch = () => {
            if (this.condensed) {
                this.toggleCondensedMenu();
            }
            else {
                this.searchInputOpen = !this.searchInputOpen;
                this.searchInputOpenChange.emit(this.searchInputOpen);
            }
        };
        this.toggleUser = () => {
            this.userMenuOpen = !this.userMenuOpen;
            this.userMenuOpenChange.emit(this.userMenuOpen);
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        this.inheritedAttributes = inheritAriaAttributes(this.el);
        this.isLight = isLightMode();
        // Watch for theme attribute changes
        this.themeObserver = new MutationObserver(() => {
            this.isLight = isLightMode();
        });
        // Observe the html element for data-theme attribute changes
        this.themeObserver.observe(document.documentElement, {
            attributes: true,
            attributeFilter: ['data-theme'],
        });
    }
    disconnectedCallback() {
        if (this.themeObserver) {
            this.themeObserver.disconnect();
        }
    }
    handleClickOutside(event) {
        const path = event.composedPath ? event.composedPath() : [event.target];
        if (this.appsMenuOpen) {
            const appsButton = this.el.querySelector('modus-wc-button:has(svg[class*="apps"])');
            if (this.appsRef &&
                !path.includes(this.appsRef) &&
                !path.includes(appsButton)) {
                this.appsMenuOpen = false;
                this.appsMenuOpenChange.emit(false);
            }
        }
        if (this.condensedMenuOpen) {
            const condenseMenuButton = this.el.querySelector('modus-wc-button:has(svg[class*="more-vertical"])');
            if (this.condensedMenuRef &&
                !path.includes(this.condensedMenuRef) &&
                !path.includes(condenseMenuButton)) {
                this.condensedMenuOpen = false;
                this.condensedMenuOpenChange.emit(false);
            }
        }
        if (this.mainMenuOpen) {
            const menuButton = this.el.querySelector('modus-wc-button:has(svg[class*="menu"])');
            if (this.menuRef &&
                !path.includes(this.menuRef) &&
                !path.includes(menuButton)) {
                this.mainMenuOpen = false;
            }
        }
        if (this.notificationsMenuOpen) {
            const notificationsButton = this.el.querySelector('modus-wc-button:has(svg[class*="notifications"])');
            if (this.notificationsRef &&
                !path.includes(this.notificationsRef) &&
                !path.includes(notificationsButton)) {
                this.notificationsMenuOpen = false;
                this.notificationsMenuOpenChange.emit(false);
            }
        }
        if (this.userMenuOpen) {
            const userButton = this.el.querySelector('modus-wc-button:has([class*="user-button"])');
            if (this.userRef &&
                !path.includes(this.userRef) &&
                !path.includes(userButton)) {
                this.userMenuOpen = false;
                this.userMenuOpenChange.emit(false);
            }
        }
    }
    getClasses() {
        const classList = [''];
        // The order CSS classes are added matters to CSS specificity
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7;
        const condensedHasItems = ((_a = this.visibility) === null || _a === void 0 ? void 0 : _a.apps) ||
            ((_b = this.visibility) === null || _b === void 0 ? void 0 : _b.help) ||
            ((_c = this.visibility) === null || _c === void 0 ? void 0 : _c.notifications) ||
            ((_d = this.visibility) === null || _d === void 0 ? void 0 : _d.search);
        const accessibleName = (this.logoName || 'trimble')
            .replace(/_/g, ' ')
            .replace(/\b\w/g, (char) => char.toUpperCase());
        return (h(Host, Object.assign({ key: 'a96c0a2319d2f2ef4a88a21526f3ae25c68a9ae7', class: this.getClasses() }, this.inheritedAttributes), h("modus-wc-toolbar", { key: 'd9c07e6f478ad4860b18df7c599602e13d780806' }, h("div", { key: 'ca4a0291a3e4bbfd039ee14309743645776b722e', slot: "start" }, ((_e = this.visibility) === null || _e === void 0 ? void 0 : _e.mainMenu) && (h(Fragment, { key: 'b6331e0655045949275e66468e67fa842ace3ea8' }, h("modus-wc-button", { key: 'b6be968a376e2b14d636302704d3a3f89541253e', "aria-label": "Main menu", onButtonClick: this.toggleMainMenu, shape: "square", size: "sm", variant: "borderless" }, h(MenuSolidIcon, { key: 'df192f45e0b87aced69ca6fe97beeca559f938cf' })), h("div", { key: 'a41e65cd3cf4db1779bbbf3996818479ff0ef251', class: `main-menu ${this.mainMenuOpen ? 'visible' : 'hidden'}`, ref: (el) => (this.menuRef = el) }, h("slot", { key: 'e24743884b7d9069d71daf361755f51aba9996d9', name: "main-menu" })))), ((_f = this.visibility) === null || _f === void 0 ? void 0 : _f.logo) !== false && (h("modus-wc-button", { key: '7069e5c3a245ade840093907959860fcfa0fd84c', "aria-label": `${accessibleName} logo`, customClass: "logo", onButtonClick: this.handleTrimbleLogoClick, size: "sm", variant: "borderless" }, h("modus-wc-logo", { key: '687aa275911464af35b570dc4a6810f9a78ba1df', name: this.logoName || 'trimble', emblem: this.condensed }))), h("slot", { key: 'e49084b82343f0d161dba0975227ff452346c541', name: "start" })), h("div", { key: '966773d1b7330f25726cff5606fa2e498b301a1f', slot: "center" }, h("slot", { key: '40e4719991c6d35fbd8ae2d6595238bd733edc61', name: "center" })), h("div", { key: 'b59fa42ff2daa381c606177e96b246eb824b99ba', slot: "end" }, h("slot", { key: 'f38a0b1fcee9a8693ef0b938ff1a5d278733d9c8', name: "end" }), ((_g = this.visibility) === null || _g === void 0 ? void 0 : _g.ai) && (h(Fragment, { key: '672c2710f8870ced86378538d69630017faee7ff' }, h("modus-wc-button", { key: 'c000cc90884e74e259d5635434096efc27dc6a41', "aria-label": "AI assistant", customClass: "ai", onButtonClick: this.handleAiClick, shape: "square", size: "sm", variant: "borderless" }, this.isLight ? h(AiLightIcon, null) : h(AiDarkIcon, null)))), this.condensed && condensedHasItems && (h(Fragment, { key: 'b23dfa323a818cbd20d3c7e98e37b724dbc00226' }, h("modus-wc-button", { key: '60df9dde5da5e66dcfaf55863b603910c859fb6e', "aria-label": "More options", onButtonClick: this.toggleCondensedMenu, shape: "square", size: "sm", variant: "borderless" }, h(MoreVerticalSolidIcon, { key: '7e4a6cf4dabf5d9ea91d18915ce6746fb2433c12' })), this.condensedMenuOpen && (h("modus-wc-menu", { key: '686ee34b1283158eaeb65dcd47664fb7db3b7d60', bordered: true, customClass: "condense-menu", ref: (el) => (this.condensedMenuRef = el) }, ((_h = this.visibility) === null || _h === void 0 ? void 0 : _h.search) && (h("modus-wc-menu-item", { key: 'aaf51a3ef070f7a2cee76e5346e745190fdf75f1', label: ((_j = this.textOverrides) === null || _j === void 0 ? void 0 : _j.search) || 'Search', onItemSelect: () => this.handleSearchClick(), value: "search" })), ((_k = this.visibility) === null || _k === void 0 ? void 0 : _k.notifications) && (h("modus-wc-menu-item", { key: '1bbfb98c1aaea55c28824afb32d3530a8fba38e8', label: ((_l = this.textOverrides) === null || _l === void 0 ? void 0 : _l.notifications) || 'Notifications', onItemSelect: () => this.handleNotificationsClick(), value: "notifications" })), ((_m = this.visibility) === null || _m === void 0 ? void 0 : _m.help) && (h("modus-wc-menu-item", { key: '3c729efc88d91bb40c46688e14c0c2ae84d49cb2', label: ((_o = this.textOverrides) === null || _o === void 0 ? void 0 : _o.help) || 'Help', onItemSelect: () => this.handleHelpClick(), value: "help" })), ((_p = this.visibility) === null || _p === void 0 ? void 0 : _p.apps) && (h("modus-wc-menu-item", { key: 'e622c9e085472d47f0a914bde52645f8d2a7f00e', label: ((_q = this.textOverrides) === null || _q === void 0 ? void 0 : _q.apps) || 'Apps', onItemSelect: () => this.handleAppsClick(), value: "apps" })))))), ((_r = this.visibility) === null || _r === void 0 ? void 0 : _r.search) && !this.condensed && (h(Fragment, { key: '89dfb9e22e604cbe60dba111b818b442cbc291e9' }, ((_s = this.visibility) === null || _s === void 0 ? void 0 : _s.searchInput) && this.searchInputOpen && (h("modus-wc-text-input", { key: '004b55ac6610f7aaaa3bb5cc9e71a1e78e67cd93', includeClear: true, includeSearch: true, onInputChange: this.handleSearchChange, placeholder: ((_t = this.textOverrides) === null || _t === void 0 ? void 0 : _t.search) || 'Search', size: "sm", value: this.searchValue })), h("modus-wc-button", { key: 'c25e08a4ca0d65437aa47eb3885db97b40a5975f', "aria-label": "Search", onButtonClick: this.handleSearchClick, shape: "square", size: "sm", variant: "borderless" }, h(SearchSolidIcon, { key: '60d6121651269f4b8d58d4cabffa694aef808684' })))), ((_u = this.visibility) === null || _u === void 0 ? void 0 : _u.notifications) && !this.condensed && (h(Fragment, { key: 'e23fbf1ecf390eb115945e9c5d81f8ccc3f08b16' }, h("modus-wc-button", { key: '81a7d2ff8e8ce919d4baaeaa9c855b8ccb29a496', "aria-label": "Notifications", onButtonClick: this.handleNotificationsClick, shape: "square", size: "sm", variant: "borderless" }, h(NotificationsSolidIcon, { key: '0bb481ee4264decd8554afa0f411e289c084456c' })), h("div", { key: '180f55b4c89a9ece7c3c9464d2fe017aa9ac2525', class: `notifications ${this.notificationsMenuOpen ? 'visible' : 'hidden'}`, ref: (el) => (this.notificationsRef = el) }, h("slot", { key: 'dc3c02f0a431740b77add477214d83edd1ead726', name: "notifications" })))), ((_v = this.visibility) === null || _v === void 0 ? void 0 : _v.help) && !this.condensed && (h("modus-wc-button", { key: 'c846db586e982dec0364b79cdf3b13e4019b1414', "aria-label": "Help", onButtonClick: this.handleHelpClick, shape: "square", size: "sm", variant: "borderless" }, h(HelpSolidIcon, { key: '8a86189c80fa17d9ad420eb0ed9ae98e368ac935' }))), ((_w = this.visibility) === null || _w === void 0 ? void 0 : _w.apps) && !this.condensed && (h(Fragment, { key: 'aa2bc840fab19c8aa2c2c77b668493e04671973f' }, h("modus-wc-button", { key: 'c680be3d30780e247c0330abe26273fa8ba25c63', "aria-label": "Apps", onButtonClick: this.handleAppsClick, shape: "square", size: "sm", variant: "borderless" }, h(AppsSolidIcon, { key: '42047b4d08693b4a16d0e3b56cadac31b2528d82' })), h("div", { key: 'c7ff74df644692b970acf2a4eec6dea83db05147', class: `apps ${this.appsMenuOpen ? 'visible' : 'hidden'}`, ref: (el) => (this.appsRef = el) }, h("slot", { key: '592395ca8508d30800561d42e60541eb409f1e2b', name: "apps" })))), ((_x = this.visibility) === null || _x === void 0 ? void 0 : _x.user) && (h(Fragment, { key: '0a94e30ed9398bf8e98badd32ffb6218637fdcad' }, h("modus-wc-button", { key: '849bf9b2e3bcb3256c800f3aba3a60f35651153a', "aria-label": "User profile", customClass: "user-button", onButtonClick: this.toggleUser, shape: "circle", size: "sm", variant: "borderless" }, h("modus-wc-avatar", { key: '4013ddeea80555a4458f852ab1c6882588fff5bf', alt: ((_y = this.userCard) === null || _y === void 0 ? void 0 : _y.avatarAlt) || '', imgSrc: (_z = this.userCard) === null || _z === void 0 ? void 0 : _z.avatarSrc, initials: (_0 = this.userCard) === null || _0 === void 0 ? void 0 : _0.name, size: "xs" })), h("div", { key: '476e159c4fdbe1c5ad737978b71b7db03d91d408', class: `user ${this.userMenuOpen ? 'visible' : 'hidden'}`, ref: (el) => (this.userRef = el) }, h("modus-wc-card", { key: 'c3d4d4303f987e1df8f351176812ebbb1c177e0d' }, h("div", { key: '2316ec90fb1caa232dd584db490d1be837940654', slot: "header" }, h("modus-wc-avatar", { key: '52811d868a45c3d5902819cabd6a06a426dbf944', alt: ((_1 = this.userCard) === null || _1 === void 0 ? void 0 : _1.avatarAlt) || '', imgSrc: (_2 = this.userCard) === null || _2 === void 0 ? void 0 : _2.avatarSrc, initials: (_3 = this.userCard) === null || _3 === void 0 ? void 0 : _3.name })), h("div", { key: 'db820f5379b1de56c8f6a7e16de8d9d80247a71a', slot: "title" }, (_4 = this.userCard) === null || _4 === void 0 ? void 0 : _4.name), h("div", { key: '79190afffffa0cc62aebf099acbe8637a15c4ffa' }, (_5 = this.userCard) === null || _5 === void 0 ? void 0 : _5.email), h("div", { key: '2b87dcebe0948a6a19ca6f63b75d640a37591c20', slot: "actions" }, h("modus-wc-button", { key: 'a2dbe18354bc72eede0eaece239df961686dcfbc', customClass: "my-trimble", onButtonClick: this.handleMyTrimbleClick }, ((_6 = this.userCard) === null || _6 === void 0 ? void 0 : _6.myTrimbleButton) || 'Access MyTrimble')), h("div", { key: '30f50417ac0d1ad4767d05172085162fdfd63f0a', slot: "footer" }, h("modus-wc-button", { key: '5019a97641df187bb58a79cdb16e0b179fe5a069', customClass: "sign-out", onButtonClick: this.handleSignOutClick, variant: "borderless" }, ((_7 = this.userCard) === null || _7 === void 0 ? void 0 : _7.signOutButton) || 'Sign out'))))))))));
    }
    get el() { return this; }
    static get style() { return modusWcNavbarCss; }
}, [4, "modus-wc-navbar", {
        "logoName": [1, "logo-name"],
        "appsMenuOpen": [1028, "apps-menu-open"],
        "condensed": [4],
        "condensedMenuOpen": [1028, "condensed-menu-open"],
        "customClass": [1, "custom-class"],
        "mainMenuOpen": [1028, "main-menu-open"],
        "notificationsMenuOpen": [1028, "notifications-menu-open"],
        "searchDebounceMs": [2, "search-debounce-ms"],
        "searchInputOpen": [1028, "search-input-open"],
        "textOverrides": [16, "text-overrides"],
        "userCard": [16, "user-card"],
        "userMenuOpen": [1028, "user-menu-open"],
        "visibility": [16],
        "isLight": [32],
        "searchValue": [32]
    }, [[4, "click", "handleClickOutside"]]]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-navbar", "modus-wc-avatar", "modus-wc-button", "modus-wc-card", "modus-wc-checkbox", "modus-wc-icon", "modus-wc-input-feedback", "modus-wc-input-label", "modus-wc-logo", "modus-wc-menu", "modus-wc-menu-item", "modus-wc-text-input", "modus-wc-toolbar", "modus-wc-tooltip"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-navbar":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcNavbar$1);
            }
            break;
        case "modus-wc-avatar":
            if (!customElements.get(tagName)) {
                defineCustomElement$e();
            }
            break;
        case "modus-wc-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$d();
            }
            break;
        case "modus-wc-card":
            if (!customElements.get(tagName)) {
                defineCustomElement$c();
            }
            break;
        case "modus-wc-checkbox":
            if (!customElements.get(tagName)) {
                defineCustomElement$b();
            }
            break;
        case "modus-wc-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$a();
            }
            break;
        case "modus-wc-input-feedback":
            if (!customElements.get(tagName)) {
                defineCustomElement$9();
            }
            break;
        case "modus-wc-input-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$8();
            }
            break;
        case "modus-wc-logo":
            if (!customElements.get(tagName)) {
                defineCustomElement$7();
            }
            break;
        case "modus-wc-menu":
            if (!customElements.get(tagName)) {
                defineCustomElement$6();
            }
            break;
        case "modus-wc-menu-item":
            if (!customElements.get(tagName)) {
                defineCustomElement$5();
            }
            break;
        case "modus-wc-text-input":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "modus-wc-toolbar":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "modus-wc-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const ModusWcNavbar = ModusWcNavbar$1;
const defineCustomElement = defineCustomElement$1;

export { ModusWcNavbar, defineCustomElement };
