// @trimble-oss/moduswebcomponents/components/modus-wc-radio.js
import { p as proxyCustomElement, H, d as createEvent, h, c as Host } from './p-X1tirp06.js';
import { h as handleShadowDOMStyles } from './p-CKNBKeF3.js';
import { D as DAISY_TO_MODUS_LABEL_SIZE } from './p-DZXd_sgn.js';
import { g as generateElementId, i as inheritAriaAttributes } from './p-CPG9FZRU.js';
import { d as defineCustomElement$2 } from './p-CqPlhPZ4.js';

const convertPropsToClasses = ({ size, }) => {
    let classes = '';
    if (size) {
        classes = `${classes} modus-wc-radio-${size}`;
    }
    return classes.trim();
};

const modusWcRadioCss = "modus-wc-radio.modus-wc-radio-host{align-items:center;display:flex}modus-wc-radio.modus-wc-radio-host .modus-wc-input-label{font-weight:var(--modus-wc-font-weight-normal);padding-inline-start:var(--modus-wc-spacing-md)}modus-wc-radio.modus-wc-radio-host [type=radio].modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-primary)}modus-wc-radio.modus-wc-radio-host [type=radio].modus-wc-radio:focus{outline-color:var(--modus-wc-color-primary)}modus-wc-radio.modus-wc-radio-host [type=radio].modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-primary)}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio{--fallback-bc:var(--modus-wc-color-gray-4);border-width:var(--modus-wc-border-width-sm)}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-xs,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-xs{height:0.875rem;width:0.875rem}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-sm,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-sm{height:1rem;width:1rem}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-md,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-md{height:1.125rem;width:1.125rem}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-lg,[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-lg{height:1.25rem;width:1.25rem}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio:focus{outline-color:var(--modus-wc-color-trimble-blue)}[data-theme=modus-classic-light] modus-wc-radio .modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-highlight-blue)}[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio:focus{outline-color:var(--modus-wc-color-highlight-blue)}[data-theme=modus-classic-dark] modus-wc-radio .modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=connect-light] modus-wc-radio .modus-wc-radio,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio{--fallback-bc:var(--modus-wc-color-gray-4);border-width:var(--modus-wc-border-width-sm)}[data-theme=connect-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-xs,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-xs{height:0.875rem;width:0.875rem}[data-theme=connect-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-sm,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-sm{height:1rem;width:1rem}[data-theme=connect-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-md,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-md{height:1.25rem;width:1.25rem}[data-theme=connect-light] modus-wc-radio .modus-wc-radio.modus-wc-radio-lg,[data-theme=connect-dark] modus-wc-radio .modus-wc-radio.modus-wc-radio-lg{height:1.75rem;width:1.75rem}[data-theme=connect-light] modus-wc-radio .modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=connect-light] modus-wc-radio .modus-wc-radio:focus{outline-color:var(--modus-wc-color-trimble-blue)}[data-theme=connect-light] modus-wc-radio .modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-trimble-blue)}[data-theme=connect-dark] modus-wc-radio .modus-wc-radio:checked{--fallback-bc:var(--modus-wc-color-highlight-blue)}[data-theme=connect-dark] modus-wc-radio .modus-wc-radio:focus{outline-color:var(--modus-wc-color-highlight-blue)}[data-theme=connect-dark] modus-wc-radio .modus-wc-radio:hover{--fallback-bc:var(--modus-wc-color-trimble-blue)}";

const ModusWcRadio$1 = /*@__PURE__*/ proxyCustomElement(class ModusWcRadio extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inputBlur = createEvent(this, "inputBlur");
        this.inputChange = createEvent(this, "inputChange");
        this.inputFocus = createEvent(this, "inputFocus");
        this.inheritedAttributes = {};
        this.generatedId = generateElementId();
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The disabled state of the radio. */
        this.disabled = false;
        /** Name of the form control. Submitted with the form as part of a name/value pair. */
        this.name = '';
        /** A value is required for the form to be submittable. */
        this.required = false;
        /** The size of the input. */
        this.size = 'md';
        /** The value of the radio. */
        this.value = false;
        this.handleBlur = (event) => {
            this.inputBlur.emit(event);
        };
        this.handleFocus = (event) => {
            this.inputFocus.emit(event);
        };
        this.handleInput = (event) => {
            this.inputChange.emit(event);
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Radio button';
        }
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-radio'];
        const propClasses = convertPropsToClasses({ size: this.size });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        const labelSize = this.size && DAISY_TO_MODUS_LABEL_SIZE[this.size];
        const effectiveId = this.inputId || this.generatedId;
        return (h(Host, { key: '73fd40e4f2b99da5d0f16457662e727c1fe357d0', class: "modus-wc-radio-host" }, h("input", Object.assign({ key: '3997d51c68ec38bcb882512e589ecdc1be1a020d', "aria-checked": this.value, "aria-disabled": this.disabled, checked: this.value, class: this.getClasses(), disabled: this.disabled, id: effectiveId, name: this.name, onBlur: this.handleBlur, onFocus: this.handleFocus, onInput: this.handleInput, required: this.required, tabIndex: this.inputTabIndex, type: "radio" }, this.inheritedAttributes)), this.label && (h("modus-wc-input-label", { key: 'aaf9b2c94ad856483850d76e65c7f79e09e0377e', forId: effectiveId, labelText: this.label, required: this.required, size: labelSize }))));
    }
    get el() { return this; }
    static get style() { return modusWcRadioCss; }
}, [0, "modus-wc-radio", {
        "customClass": [1, "custom-class"],
        "disabled": [4],
        "inputId": [1, "input-id"],
        "inputTabIndex": [2, "input-tab-index"],
        "label": [1],
        "name": [1],
        "required": [4],
        "size": [1],
        "value": [1540]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-radio", "modus-wc-input-label"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-radio":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcRadio$1);
            }
            break;
        case "modus-wc-input-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const ModusWcRadio = ModusWcRadio$1;
const defineCustomElement = defineCustomElement$1;

export { ModusWcRadio, defineCustomElement };
