// @trimble-oss/moduswebcomponents/components/modus-wc-utility-panel.js
import { p as proxyCustomElement, H, d as createEvent, h, F as Fragment } from './p-X1tirp06.js';
import { h as handleShadowDOMStyles } from './p-CKNBKeF3.js';

const modusWcUtilityPanelCss = "modus-wc-utility-panel{--modus-wc-utility-panel-padding:1rem;--modus-wc-utility-panel-width:312px;--modus-wc-utility-panel-header-height:50px;--modus-wc-utility-panel-border-color:var(--modus-wc-color-base-200);--modus-wc-utility-panel-box-shadow:-2px 0 4px rgba(0, 0, 0, 0.1);--modus-wc-utility-panel-transition-duration:0.3s}.modus-wc-utility-panel{background-color:var(--modus-wc-color-base-page);box-shadow:var(--modus-wc-utility-panel-box-shadow);color:var(--modus-wc-utility-panel-color);height:100%;position:absolute;right:calc(var(--modus-wc-utility-panel-width) * -1 - 12px);top:0;transition:right var(--modus-wc-utility-panel-transition-duration) ease-out;width:var(--modus-wc-utility-panel-width);z-index:1000}.modus-wc-utility-panel.open{background:var(--modus-wc-color-base-page);right:0;transition:right var(--modus-wc-utility-panel-transition-duration) ease-out}.modus-wc-utility-panel .modus-wc-utility-panel-content{background:var(--modus-wc-color-base-page);display:flex;flex-direction:column;height:100%}.modus-wc-utility-panel .modus-wc-utility-panel-content .modus-wc-utility-panel-header,.modus-wc-utility-panel .modus-wc-utility-panel-content .modus-wc-utility-panel-footer{align-items:center;background:var(--modus-wc-color-base-page);display:flex;height:var(--modus-wc-utility-panel-header-height);padding:0 var(--modus-wc-utility-panel-padding)}.modus-wc-utility-panel .modus-wc-utility-panel-content .modus-wc-utility-panel-body{background:var(--modus-wc-color-base-page);flex:1;overflow:auto;padding:var(--modus-wc-utility-panel-padding)}.modus-wc-utility-panel .modus-wc-utility-panel-content hr{border:none;border-top:1px solid var(--modus-wc-utility-panel-border-color);margin:0}";

const ModusWcUtilityPanel$1 = /*@__PURE__*/ proxyCustomElement(class ModusWcUtilityPanel extends H {
    constructor() {
        super();
        this.__registerHost();
        this.panelOpened = createEvent(this, "panelOpened");
        this.panelClosed = createEvent(this, "panelClosed");
        /** The panel is expanded or closed */
        this.expanded = false;
        /** Determines if the panel pushes content or displays an overlay. */
        this.pushContent = false;
        this.isInitialLoad = true;
        this.handlePanelClose = () => {
            void this.closePanel();
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
    }
    componentDidLoad() {
        // Only adjust content if panel is already expanded on load and we have a target
        if (this.pushContent && this.expanded && this.targetElement) {
            this.adjustContent();
        }
        // Mark that initial load is complete after adjusting content
        this.isInitialLoad = false;
    }
    handleExpandedChange(newValue) {
        // Skip the watcher on initial load
        if (this.isInitialLoad) {
            return;
        }
        if (newValue) {
            void this.openPanel();
        }
        else {
            void this.closePanel();
        }
    }
    handleTargetChange() {
        // Re-adjust content when target changes
        if (this.expanded && this.pushContent && this.targetElement) {
            this.adjustContent();
        }
    }
    openPanel() {
        this.panelOpened.emit();
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    closePanel() {
        this.panelClosed.emit();
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    adjustContent() {
        if (!this.pushContent || !this.targetElement)
            return;
        // Add base class for transitions
        this.targetElement.classList.add('modus-wc-utility-panel-push-target');
        // Toggle pushed class based on expanded state
        if (this.expanded) {
            this.targetElement.classList.add('modus-wc-utility-panel-pushed');
        }
        else {
            this.targetElement.classList.remove('modus-wc-utility-panel-pushed');
        }
    }
    hasSlotContent(slotName) {
        const slot = this.el.querySelector(`[slot="${slotName}"]`);
        return !!slot;
    }
    render() {
        const hasHeader = this.hasSlotContent('header');
        const hasFooter = this.hasSlotContent('footer');
        return (h("div", { key: 'c3e69d7c25cc68d477b7dff36f2429a7a90deffc', class: {
                'modus-wc-utility-panel': true,
                open: this.expanded,
            } }, h("div", { key: 'a80dc0d0155b1755114d8079b9bea8e47f59f4ce', class: "modus-wc-utility-panel-content" }, hasHeader && (h(Fragment, { key: '6037c5cd032f6b160e6a956e147445efd93061d4' }, h("div", { key: '02d222e4aac4c8ab9fd4a5c601649dcd4a870eaf', class: "modus-wc-utility-panel-header" }, h("slot", { key: '1ee7c97ef7677fd2b546dbf902a800ccf70a2d68', name: "header" })), h("hr", { key: '8f9b6e4e96bd8e9afae2f447f0c83f744c5ca175' }))), h("div", { key: 'a6dc213a2b651bfcf58e50f71f1683b45f8c3448', class: "modus-wc-utility-panel-body" }, h("slot", { key: '74a4bdc447ca4d07b9bfbe955776072570f8fd68', name: "body" })), hasFooter && (h(Fragment, { key: '00ed976af7d33d273c47199d9864fcfcafa41a65' }, h("hr", { key: 'f66ee98a4b53e033dd5303f9347547f54b44d1c0' }), h("div", { key: 'f39464a2c72dfa50fb49508d303419c7f08b6ca4', class: "modus-wc-utility-panel-footer" }, h("slot", { key: '75d25d058b3abbcc7e6153af18f1bfea8d2d7c60', name: "footer" })))))));
    }
    get el() { return this; }
    static get watchers() { return {
        "expanded": ["handleExpandedChange"],
        "targetElement": ["handleTargetChange"]
    }; }
    static get style() { return modusWcUtilityPanelCss; }
}, [4, "modus-wc-utility-panel", {
        "expanded": [4],
        "pushContent": [4, "push-content"],
        "targetElement": [16, "target-element"]
    }, undefined, {
        "expanded": ["handleExpandedChange"],
        "targetElement": ["handleTargetChange"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-utility-panel"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-utility-panel":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcUtilityPanel$1);
            }
            break;
    } });
}

const ModusWcUtilityPanel = ModusWcUtilityPanel$1;
const defineCustomElement = defineCustomElement$1;

export { ModusWcUtilityPanel, defineCustomElement };
