// @trimble-oss/moduswebcomponents/components/modus-wc-textarea.js
import { p as proxyCustomElement, H, d as createEvent, h, c as Host } from './p-X1tirp06.js';
import { h as handleShadowDOMStyles } from './p-CKNBKeF3.js';
import { i as inheritAriaAttributes, b as inheritAttributes } from './p-CPG9FZRU.js';
import { d as defineCustomElement$4 } from './p-vnbqBzVH.js';
import { d as defineCustomElement$3 } from './p-Bfy1MXxF.js';
import { d as defineCustomElement$2 } from './p-CqPlhPZ4.js';

const convertPropsToClasses = ({ bordered, feedback, size, }) => {
    let classes = '';
    if (bordered) {
        classes = `${classes} modus-wc-textarea-bordered`;
    }
    if (feedback) {
        classes = `${classes} modus-wc-input--${feedback.level}`;
    }
    if (size) {
        classes = `${classes} modus-wc-textarea-${size}`;
    }
    return classes.trim();
};

const modusWcTextareaCss = ".modus-wc-input--error{border-color:var(--modus-wc-color-error) !important}.modus-wc-input--info{border-color:var(--modus-wc-color-info) !important}.modus-wc-input--success{border-color:var(--modus-wc-color-success) !important}.modus-wc-input--warning{border-color:var(--modus-wc-color-warning) !important}modus-wc-textarea .modus-wc-input-label{padding-bottom:var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-textarea .modus-wc-textarea,[data-theme=modus-classic-dark] modus-wc-textarea .modus-wc-textarea{border-radius:var(--modus-wc-border-radius-md)}[data-theme=modus-classic-light] modus-wc-textarea .modus-wc-textarea.modus-wc-textarea-sm,[data-theme=modus-classic-dark] modus-wc-textarea .modus-wc-textarea.modus-wc-textarea-sm{font-size:var(--modus-wc-font-size-sm);padding:var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-textarea .modus-wc-textarea.modus-wc-textarea-md,[data-theme=modus-classic-dark] modus-wc-textarea .modus-wc-textarea.modus-wc-textarea-md{font-size:var(--modus-wc-font-size-md);padding:var(--modus-wc-spacing-sm) var(--modus-wc-spacing-sm) var(--modus-wc-spacing-md) var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-textarea .modus-wc-textarea.modus-wc-textarea-lg,[data-theme=modus-classic-dark] modus-wc-textarea .modus-wc-textarea.modus-wc-textarea-lg{font-size:var(--modus-wc-font-size-lg);padding:var(--modus-wc-spacing-md)}[data-theme=modus-classic-light] modus-wc-textarea .modus-wc-textarea:focus,[data-theme=modus-classic-dark] modus-wc-textarea .modus-wc-textarea:focus{border-color:var(--modus-wc-color-blue-light);border-width:var(--modus-wc-border-width-sm);box-shadow:none;outline:none}[data-theme=modus-classic-light] modus-wc-textarea .modus-wc-textarea:read-only,[data-theme=modus-classic-dark] modus-wc-textarea .modus-wc-textarea:read-only{background-color:var(--modus-wc-color-base-100)}[data-theme=modus-classic-light] modus-wc-textarea .modus-wc-textarea.modus-wc-textarea-bordered:not(:disabled):not(:focus){border-color:var(--modus-wc-color-gray-6)}[data-theme=modus-classic-dark] modus-wc-textarea .modus-wc-textarea:focus{border-color:var(--modus-wc-color-highlight-blue)}[data-theme=connect-light] modus-wc-textarea .modus-wc-textarea,[data-theme=connect-dark] modus-wc-textarea .modus-wc-textarea{border-bottom-width:var(--input-bottom-border-width);outline-width:0 !important}[data-theme=connect-light] modus-wc-textarea .modus-wc-textarea:not(.modus-wc-select,.modus-wc-number-input),[data-theme=connect-dark] modus-wc-textarea .modus-wc-textarea:not(.modus-wc-select,.modus-wc-number-input){padding:0 var(--modus-wc-spacing-sm)}[data-theme=connect-light] modus-wc-textarea .modus-wc-textarea:hover,[data-theme=connect-dark] modus-wc-textarea .modus-wc-textarea:hover{border-bottom-color:var(--modus-wc-color-primary)}[data-theme=connect-light] modus-wc-textarea .modus-wc-textarea:active,[data-theme=connect-dark] modus-wc-textarea .modus-wc-textarea:active{border-bottom-color:var(--modus-wc-color-primary)}[data-theme=connect-light] modus-wc-textarea .modus-wc-textarea:focus,[data-theme=connect-dark] modus-wc-textarea .modus-wc-textarea:focus{border-bottom-color:var(--modus-wc-color-primary);outline:none}[data-theme=connect-light] modus-wc-textarea .modus-wc-textarea:focus-within,[data-theme=connect-dark] modus-wc-textarea .modus-wc-textarea:focus-within{border-bottom-color:var(--modus-wc-color-primary);outline:none}";

const ModusWcTextarea$1 = /*@__PURE__*/ proxyCustomElement(class ModusWcTextarea extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inputBlur = createEvent(this, "inputBlur");
        this.inputChange = createEvent(this, "inputChange");
        this.inputFocus = createEvent(this, "inputFocus");
        this.inheritedAttributes = {};
        /** Indicates that the input should have a border. */
        this.bordered = true;
        /** Custom CSS class to apply to the textarea (supports DaisyUI). */
        this.customClass = '';
        /** The disabled state of the textarea. */
        this.disabled = false;
        /** The placeholder text for the textarea. */
        this.placeholder = '';
        /** The readonly state of the textarea. */
        this.readonly = false;
        /** A value is required for the form to be submittable. */
        this.required = false;
        /** The size of the input. */
        this.size = 'md';
        /** The value of the textarea. */
        this.value = '';
        this.handleBlur = (event) => {
            this.inputBlur.emit(event);
        };
        this.handleFocus = (event) => {
            this.inputFocus.emit(event);
        };
        this.handleInput = (event) => {
            this.value = event.target.value;
            this.inputChange.emit(event);
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = this.placeholder || 'Text area';
        }
        this.inheritedAttributes = Object.assign(Object.assign({}, inheritAriaAttributes(this.el)), inheritAttributes(this.el, ['spellcheck']));
    }
    getClasses() {
        const classList = ['modus-wc-textarea', 'modus-wc-w-full'];
        const propClasses = convertPropsToClasses({
            bordered: this.bordered,
            feedback: this.feedback,
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: '5872672e78191d700f9a1ce18fee8ee6a2a98f03' }, this.label && (h("modus-wc-input-label", { key: '4c81f3a774ffe514f541722cd98dadc5c50ddd53', forId: this.inputId, labelText: this.label, required: this.required, size: this.size })), h("textarea", Object.assign({ key: 'b9e71aa7b90b2dcc953595c04df21b78485a93e5', "aria-placeholder": this.placeholder, "aria-required": this.required, autocorrect: this.autoCorrect, class: this.getClasses(), disabled: this.disabled, enterkeyhint: this.enterkeyhint, id: this.inputId, maxLength: this.maxLength, minlength: this.minLength, name: this.name, onBlur: this.handleBlur, onFocus: this.handleFocus, onInput: this.handleInput, placeholder: this.placeholder, readonly: this.readonly, required: this.required, rows: this.rows, tabIndex: this.inputTabIndex, value: this.value }, this.inheritedAttributes)), this.feedback && (h("modus-wc-input-feedback", { key: 'c09873ddbca79ac613d2b542348d7e71bfa5c0f4', level: this.feedback.level, message: this.feedback.message, size: this.size }))));
    }
    get el() { return this; }
    static get style() { return modusWcTextareaCss; }
}, [0, "modus-wc-textarea", {
        "autoCorrect": [1, "auto-correct"],
        "bordered": [4],
        "customClass": [1, "custom-class"],
        "disabled": [4],
        "enterkeyhint": [1],
        "feedback": [16],
        "inputId": [1, "input-id"],
        "inputTabIndex": [2, "input-tab-index"],
        "label": [1],
        "maxLength": [2, "max-length"],
        "minLength": [2, "min-length"],
        "name": [1],
        "placeholder": [1],
        "readonly": [4],
        "required": [4],
        "rows": [2],
        "size": [1],
        "value": [1537]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-textarea", "modus-wc-icon", "modus-wc-input-feedback", "modus-wc-input-label"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-textarea":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcTextarea$1);
            }
            break;
        case "modus-wc-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "modus-wc-input-feedback":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "modus-wc-input-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const ModusWcTextarea = ModusWcTextarea$1;
const defineCustomElement = defineCustomElement$1;

export { ModusWcTextarea, defineCustomElement };
