// @trimble-oss/moduswebcomponents/components/p-AfrNg3-r.js
import { p as proxyCustomElement, H, h, c as Host } from './p-X1tirp06.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';
import { d as defineCustomElement$1 } from './p-vnbqBzVH.js';

const convertPropsToClasses = ({ shape, size, }) => {
    let classes = '';
    if (shape) {
        switch (shape) {
            case 'circle':
                classes = `${classes} modus-wc-rounded-full`;
                break;
            case 'square':
                classes = `${classes} modus-wc-rounded-lg`;
                break;
        }
    }
    if (size) {
        switch (size) {
            case 'xs':
                classes = `${classes} modus-wc-w-8`;
                break;
            case 'sm':
                classes = `${classes} modus-wc-w-12`;
                break;
            case 'md':
                classes = `${classes} modus-wc-w-16`;
                break;
            case 'lg':
                classes = `${classes} modus-wc-w-20`;
                break;
        }
    }
    return classes.trim();
};

const modusWcAvatarCss = "modus-wc-avatar{display:inline-flex}modus-wc-avatar .modus-wc-avatar .initials{align-items:center;background-color:var(--modus-wc-color-neutral);color:var(--modus-wc-color-neutral-content);display:flex;font-weight:600;height:100%;justify-content:center;text-transform:uppercase;width:100%}modus-wc-avatar .modus-wc-avatar .no-image{align-items:center;background-color:var(--modus-wc-color-base-200);color:var(--modus-wc-color-base-content);display:flex;justify-content:center}modus-wc-avatar .modus-wc-avatar .modus-wc-w-8 .initials{font-size:var(--modus-wc-font-size-xs, 0.75rem)}modus-wc-avatar .modus-wc-avatar .modus-wc-w-12 .initials{font-size:var(--modus-wc-font-size-sm, 0.875rem)}modus-wc-avatar .modus-wc-avatar .modus-wc-w-16 .initials{font-size:var(--modus-wc-font-size-md, 1rem)}modus-wc-avatar .modus-wc-avatar .modus-wc-w-20 .initials{font-size:var(--modus-wc-font-size-lg, 1.25rem)}";

const ModusWcAvatar = /*@__PURE__*/ proxyCustomElement(class ModusWcAvatar extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The location of the image. */
        this.imgSrc = '';
        /** The initials to display when no image is provided. */
        this.initials = '';
        // TODO - add placeholder support (need UX logic)
        /** The shape of the avatar. */
        this.shape = 'circle';
        /** The size of the avatar. */
        this.size = 'md';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = [];
        const propClasses = convertPropsToClasses({
            shape: this.shape,
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        if (!this.imgSrc && !this.initials)
            classList.push('no-image');
        return classList.join(' ');
    }
    getUserInitials() {
        if (!this.initials)
            return '';
        return this.initials
            .trim()
            .split(/\s+/)
            .filter(Boolean)
            .map((part) => part.charAt(0))
            .join('')
            .substring(0, 3)
            .toUpperCase();
    }
    render() {
        const altText = this.alt || 'User avatar';
        return (h(Host, { key: '8ac52db552b5093476ae19f158819a3c9d253ab3' }, h("div", Object.assign({ key: 'e3452c9c37a8adad4aaae454112f4020b7bf4f83', class: "modus-wc-avatar" }, this.inheritedAttributes), h("div", { key: 'eba4c00724a6f4dff7636d4a43fa46e93a28f75f', class: this.getClasses() }, this.imgSrc ? (h("img", { src: this.imgSrc, alt: altText })) : this.initials ? (h("span", { class: "initials", "aria-label": this.alt || this.initials }, this.getUserInitials())) : (h("modus-wc-icon", { "aria-label": altText, name: "person", size: this.size, variant: "solid" }))))));
    }
    get el() { return this; }
    static get style() { return modusWcAvatarCss; }
}, [0, "modus-wc-avatar", {
        "alt": [1],
        "customClass": [1, "custom-class"],
        "imgSrc": [1, "img-src"],
        "initials": [1],
        "shape": [1],
        "size": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-avatar", "modus-wc-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-avatar":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcAvatar);
            }
            break;
        case "modus-wc-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusWcAvatar as M, defineCustomElement as d };
