// @trimble-oss/moduswebcomponents/components/p-BUr8lm9p.js
import { p as proxyCustomElement, H, h, c as Host } from './p-X1tirp06.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';

const modusWcToolbarCss = ".modus-wc-navbar{background:var(--modus-wc-color-base-page);border-radius:var(--modus-wc-border-radius-box);box-shadow:0 1px 2px -1px color-mix(in sRGB, var(--modus-wc-color-black) 10%, transparent), 0 1px 3px 0 color-mix(in sRGB, var(--modus-wc-color-black) 10%, transparent)}";

const ModusWcToolbar = /*@__PURE__*/ proxyCustomElement(class ModusWcToolbar extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the outer div. */
        this.customClass = '';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-navbar'];
        // The order CSS classes are added matters to CSS specificity
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: '252cc26c4653ec039229cfecb7ef9049f0875b95' }, h("div", Object.assign({ key: '55664aca492ad12f8e099bc0c6a6046bd81fa545', class: this.getClasses() }, this.inheritedAttributes), h("div", { key: '2031f716b318fa5b04cb3661a2adf707644ac072', class: "modus-wc-navbar-start" }, h("slot", { key: 'a23a967e42d463f506314e8d4f80f7aad41b9de2', name: "start" })), h("div", { key: 'e796353849827c48f907d7071cd68bd8b0f7ca4c', class: "modus-wc-navbar-center" }, h("slot", { key: '3a05fd3d4cd1b45de84585f9334f9b75ef903c8f', name: "center" })), h("div", { key: '7994ec1b7f45e0b6c5c26bfa3458e24c969a2fff', class: "modus-wc-navbar-end" }, h("slot", { key: '4040c3637315c02313247be884e9719efb6c9e49', name: "end" })))));
    }
    get el() { return this; }
    static get style() { return modusWcToolbarCss; }
}, [4, "modus-wc-toolbar", {
        "customClass": [1, "custom-class"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-toolbar"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-toolbar":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcToolbar);
            }
            break;
    } });
}

export { ModusWcToolbar as M, defineCustomElement as d };
