// @trimble-oss/moduswebcomponents/components/p-BvSTk6wy.js
import { p as proxyCustomElement, H, d as createEvent, h, c as Host } from './p-X1tirp06.js';
import { h as handleShadowDOMStyles } from './p-CKNBKeF3.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';
import { d as defineCustomElement$3 } from './p-CeZm3cg9.js';
import { d as defineCustomElement$2 } from './p-CqPlhPZ4.js';
import { d as defineCustomElement$1 } from './p-C5jBPzA7.js';

const convertPropsToClasses = ({ active, bordered, disabled, focused, size, }) => {
    let classes = '';
    if (bordered) {
        classes = `${classes} modus-wc-menu-item-bordered`;
    }
    if (disabled) {
        classes = `${classes} modus-wc-menu-item-disabled`;
    }
    if (active) {
        classes = `${classes} modus-wc-menu-item-active`;
    }
    if (focused) {
        classes = `${classes} modus-wc-menu-item-focused`;
    }
    if (size) {
        classes = `${classes} modus-wc-menu-item-${size}`;
    }
    return classes.trim();
};

const modusWcMenuItemCss = "modus-wc-menu-item .modus-wc-menu-item button{align-items:center;border-radius:inherit;display:flex;width:100%}modus-wc-menu-item .modus-wc-menu-item button:hover{background-color:transparent}modus-wc-menu-item .modus-wc-menu-item button:focus{background-color:transparent}modus-wc-menu-item .modus-wc-menu-item button.modus-wc-menu-dropdown-toggle::after{color:var(--modus-wc-color-base-content)}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content{align-items:center;display:flex;width:100%}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content [slot=start-icon]{padding-inline-end:var(--modus-wc-spacing-sm)}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content modus-wc-checkbox{padding-inline-end:var(--modus-wc-spacing-sm)}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content .modus-wc-menu-item-labels{padding-inline-end:var(--modus-wc-spacing-xs);padding-inline-start:var(--modus-wc-spacing-md);white-space:nowrap}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content .modus-wc-menu-item-sublabel{font-size:var(--modus-wc-font-size-sm);white-space:nowrap}modus-wc-menu-item .modus-wc-menu-item button .modus-wc-menu-item-content .modus-wc-menu-item-selected-icon{margin-left:auto}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-sm button{font-size:var(--modus-wc-font-size-sm)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-sm button .modus-wc-menu-item-sublabel{font-size:var(--modus-wc-font-size-xs)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-md button{font-size:var(--modus-wc-font-size-md)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-md button .modus-wc-menu-item-sublabel{font-size:var(--modus-wc-font-size-sm)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-lg button{font-size:var(--modus-wc-font-size-lg)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-lg button .modus-wc-menu-item-sublabel{font-size:var(--modus-wc-font-size-md)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-bordered{border-bottom:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-base-200);border-radius:0;border-top:var(--modus-wc-border-width-xs) solid var(--modus-wc-color-base-200)}modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-selected){color:var(--modus-wc-color-base-content)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected{background-color:color-mix(in sRGB, var(--modus-wc-color-primary-pale) 50%, transparent);border-radius:inherit;color:var(--modus-wc-base-content)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected:focus{background-color:color-mix(in sRGB, var(--modus-wc-color-primary-pale) 50%, transparent)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected button{background-color:transparent}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected button:hover,modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-selected button:focus{background-color:transparent;border-radius:inherit;box-shadow:none;color:inherit}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active{background-color:color-mix(in sRGB, var(--modus-wc-color-primary-pale) 50%, transparent);border-radius:inherit;color:var(--modus-wc-color-base-content)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active:focus{background-color:color-mix(in sRGB, var(--modus-wc-color-primary-pale) 50%, transparent)}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button{background-color:transparent}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:hover,modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:focus{background-color:transparent;border-radius:inherit;box-shadow:none;color:inherit}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-disabled{opacity:0.4}modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-disabled button{pointer-events:none}modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):hover{background-color:var(--modus-wc-color-base-100);outline:none}modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):focus:not(:focus-visible),modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)).modus-wc-arrow-focused{background-color:var(--modus-wc-color-base-100);color:var(--modus-wc-color-primary);outline:none}modus-wc-menu-item .modus-wc-menu-item:focus-visible{background-color:transparent;outline:2px solid var(--modus-wc-color-primary);outline-offset:-2px}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active),[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active){color:var(--modus-wc-color-gray-light)}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item button.modus-wc-menu-dropdown-toggle::after,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item button.modus-wc-menu-dropdown-toggle::after{color:var(--modus-wc-color-gray-3)}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):hover,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):hover{background-color:var(--modus-wc-color-gray-9);color:inherit;outline:none}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)),[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)){background-color:color-mix(in sRGB, var(--modus-wc-color-primary) 30%, transparent)}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button{background-color:transparent}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:hover,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:focus,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button:hover,[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button:focus,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:hover,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item.modus-wc-menu-item-active button:focus,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button:hover,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:active:not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)) button:focus{background-color:transparent;box-shadow:none;color:inherit}[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):focus:not(:focus-visible),[data-theme=modus-classic-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)).modus-wc-arrow-focused,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)):focus:not(:focus-visible),[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item:not(.modus-wc-menu-item-active):not(.modus-wc-menu-item-disabled):not(:has(.modus-wc-menu-dropdown)).modus-wc-arrow-focused{background-color:var(--modus-wc-color-gray-9)}[data-theme=connect-light] modus-wc-menu-item .modus-wc-menu-item>button.modus-wc-menu-dropdown-toggle::after,[data-theme=connect-dark] modus-wc-menu-item .modus-wc-menu-item>button.modus-wc-menu-dropdown-toggle::after{color:var(--modus-wc-color-white)}";

const ModusWcMenuItem = /*@__PURE__*/ proxyCustomElement(class ModusWcMenuItem extends H {
    constructor() {
        super();
        this.__registerHost();
        this.itemSelect = createEvent(this, "itemSelect");
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the li element. */
        this.customClass = '';
        /** The text rendered in the menu item. */
        this.label = '';
        /** The size of the menu item. */
        this.size = 'md';
        /** The position of the tooltip relative to the menu item. */
        this.tooltipPosition = 'auto';
        /** The unique identifying value of the menu item. */
        this.value = '';
        /** Internal state to track if submenu is expanded */
        this.isExpanded = false;
        this.handleItemSelect = () => {
            if (this.hasSubmenu) {
                // Check if side nav is expanded (if this menu is inside a side nav)
                const sideNav = this.el.closest('modus-wc-side-navigation');
                if (sideNav &&
                    !sideNav.expanded) {
                    // Don't allow submenu expansion when side nav is collapsed
                    // Still emit the event for consistency
                    this.itemSelect.emit({ value: this.value });
                    return;
                }
                // The submenu should be inside this menu-item element (slotted content)
                const submenu = this.el.querySelector('.modus-wc-menu-dropdown');
                const liElement = this.el.querySelector('li');
                if (submenu && liElement) {
                    submenu.classList.toggle('modus-wc-menu-dropdown-show');
                    const buttonElement = liElement.querySelector('button');
                    // Update internal expanded state and add/remove class
                    this.isExpanded = submenu.classList.contains('modus-wc-menu-dropdown-show');
                    if (this.isExpanded) {
                        liElement.classList.add('modus-wc-menu-item-expanded');
                        if (buttonElement) {
                            buttonElement.classList.add('modus-wc-menu-dropdown-show');
                        }
                    }
                    else {
                        liElement.classList.remove('modus-wc-menu-item-expanded');
                        if (buttonElement) {
                            buttonElement.classList.remove('modus-wc-menu-dropdown-show');
                        }
                    }
                }
            }
            else if (this.resolveSelectionMode() === 'multiple' || this.checkbox) {
                this.selected = !this.selected;
            }
            else {
                if (this.resolveSelectionMode() === 'single') {
                    this.deselectSiblings();
                }
                this.selected = true;
            }
            this.itemSelect.emit({ value: this.value, selected: this.selected });
        };
    }
    componentWillLoad() {
        handleShadowDOMStyles(this.el);
        this.inheritedAttributes = inheritAriaAttributes(this.el);
        this._selectionMode = this.resolveSelectionMode();
    }
    componentDidLoad() {
        if (typeof MutationObserver === 'undefined')
            return;
        const parentMenu = this.el.closest('modus-wc-menu');
        if (parentMenu) {
            this.parentMenuObserver = new MutationObserver((mutations) => {
                for (const mutation of mutations) {
                    if (mutation.attributeName === 'selection-mode') {
                        this.handleSelectionModeChange();
                        break;
                    }
                }
            });
            this.parentMenuObserver.observe(parentMenu, { attributes: true });
        }
    }
    disconnectedCallback() {
        var _a;
        (_a = this.parentMenuObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    }
    handleSelectionModeChange() {
        this._selectionMode = this.resolveSelectionMode();
        this.selected = false;
    }
    handleKeyDown(e) {
        if (this.disabled)
            return;
        if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            this.handleItemSelect();
        }
    }
    /**
     * Public method to collapse the submenu if it's expanded
     */
    async collapseSubmenu() {
        if (this.hasSubmenu && this.isExpanded) {
            const submenu = this.el.querySelector('.modus-wc-menu-dropdown');
            const liElement = this.el.querySelector('li');
            if (submenu && liElement) {
                submenu.classList.remove('modus-wc-menu-dropdown-show');
                liElement.classList.remove('modus-wc-menu-item-expanded');
                this.isExpanded = false;
            }
        }
        return Promise.resolve();
    }
    getRootMenu() {
        var _a, _b;
        let menu = this.el.closest('modus-wc-menu');
        while (menu) {
            const parent = (_b = (_a = menu.parentElement) === null || _a === void 0 ? void 0 : _a.closest('modus-wc-menu')) !== null && _b !== void 0 ? _b : null;
            if (!parent)
                break;
            menu = parent;
        }
        return menu;
    }
    deselectSiblings() {
        const rootMenu = this.getRootMenu();
        if (!rootMenu)
            return;
        const allItems = rootMenu.querySelectorAll('modus-wc-menu-item');
        allItems.forEach((item) => {
            if (item !== this.el) {
                item.selected = false;
            }
        });
    }
    resolveSelectionMode() {
        var _a;
        return (_a = this.el.closest('modus-wc-menu')) === null || _a === void 0 ? void 0 : _a.selectionMode;
    }
    getClasses() {
        const classList = ['modus-wc-menu-item'];
        const isActive = !this.hasSubmenu && !!this.selected;
        const propClasses = convertPropsToClasses({
            active: isActive,
            bordered: this.bordered,
            disabled: this.disabled,
            focused: this.focused,
            size: this.size,
        });
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getButtonClasses() {
        return this.hasSubmenu ? 'modus-wc-menu-dropdown-toggle' : '';
    }
    getRole(mode) {
        if (mode === 'multiple')
            return 'menuitemcheckbox';
        if (mode === 'single')
            return 'menuitemradio';
        return 'menuitem';
    }
    hasCheckbox(mode) {
        return this.checkbox || mode === 'multiple';
    }
    getAriaChecked(mode) {
        if (!this.hasCheckbox(mode))
            return undefined;
        return this.selected ? 'true' : 'false';
    }
    getAriaSelected(mode) {
        if (this.hasCheckbox(mode))
            return undefined;
        if (mode === 'single')
            return this.selected ? 'true' : 'false';
        return undefined;
    }
    render() {
        const mode = this._selectionMode;
        return (h(Host, { key: 'cf1f2a9d42ca8ef8b74b3b84dbaff8dbb9e3fcbf' }, h("li", Object.assign({ key: 'c0304724c03f8508163b3680afae3660aee76ee2', "aria-checked": this.getAriaChecked(mode), "aria-selected": this.getAriaSelected(mode), "aria-disabled": this.disabled, class: this.getClasses(), role: this.getRole(mode), tabIndex: this.disabled ? -1 : 0 }, this.inheritedAttributes), h("button", { key: '1e198d92292416964ac46409ba6d088d7b1639a3', class: this.getButtonClasses(), disabled: this.disabled, onClick: this.handleItemSelect, tabIndex: -1, type: "button" }, h("div", { key: '28473bf3af4c1de33f00b35a0fcda58c74be90c0', class: "modus-wc-menu-item-content" }, (this.checkbox || mode === 'multiple') && (h("modus-wc-checkbox", { key: 'f1f2bed203505f00ed9d055e012e7c73e3bf75d1', "aria-label": "Checkbox", disabled: this.disabled, size: this.size, value: !!this.selected })), h("slot", { key: '6f1c7f217c299288f985babe0988e88dec7653c1', name: "start-icon" }), h("div", { key: '7d9f55486d9db29bc1f2a2ce5ac21ae281a5a6b8', class: "modus-wc-menu-item-labels" }, this.tooltipContent ? (h("modus-wc-tooltip", { content: this.tooltipContent, position: this.tooltipPosition, customClass: "modus-wc-menu-item-tooltip" }, h("div", null, this.label))) : (h("div", null, this.label)), this.subLabel && (h("div", { key: '694c004f591a9fd459ca15ef1a519cd8313a42b8', class: "modus-wc-menu-item-sublabel" }, this.subLabel))))), h("slot", { key: '53e15267c8984d5d30ae75d566021b6881c5ea98' }))));
    }
    get el() { return this; }
    static get style() { return modusWcMenuItemCss; }
}, [4, "modus-wc-menu-item", {
        "bordered": [4],
        "checkbox": [4],
        "customClass": [1, "custom-class"],
        "disabled": [4],
        "label": [1],
        "selected": [1028],
        "focused": [4],
        "size": [1],
        "subLabel": [1, "sub-label"],
        "tooltipContent": [1, "tooltip-content"],
        "tooltipPosition": [1, "tooltip-position"],
        "value": [1],
        "hasSubmenu": [4, "has-submenu"],
        "isExpanded": [32],
        "_selectionMode": [32],
        "collapseSubmenu": [64]
    }, [[0, "keydown", "handleKeyDown"]]]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-menu-item", "modus-wc-checkbox", "modus-wc-input-label", "modus-wc-tooltip"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-menu-item":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcMenuItem);
            }
            break;
        case "modus-wc-checkbox":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "modus-wc-input-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "modus-wc-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusWcMenuItem as M, defineCustomElement as d };
