// @trimble-oss/moduswebcomponents/components/p-DL4d_qJP.js
import { p as proxyCustomElement, H, h, c as Host } from './p-X1tirp06.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';

const convertPropsToClasses = ({ floating, }) => {
    let classes = '';
    if (floating) {
        classes = `${classes} modus-wc-panel-floating`;
    }
    return classes.trim();
};

const modusWcPanelCss = ".modus-wc-panel{background-color:var(--modus-wc-color-base-page);border:1px solid var(--modus-wc-color-base-200);border-radius:var(--rounded-box);display:flex;flex-direction:column;overflow:hidden}.modus-wc-panel-floating{box-shadow:rgba(36, 35, 45, 0.3) 1px 0 4px}.modus-wc-panel-header,.modus-wc-panel-body,.modus-wc-panel-footer{display:flex;flex-direction:column;width:100%}.modus-wc-panel-header{flex-shrink:0}.modus-wc-panel-header:empty{display:none}.modus-wc-panel-body{flex:1 1 auto;min-height:0;overflow-y:auto}.modus-wc-panel-footer{flex-shrink:0}.modus-wc-panel-footer:empty{display:none}";

const ModusWcPanel = /*@__PURE__*/ proxyCustomElement(class ModusWcPanel extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the outer div. */
        this.customClass = '';
        /** Width of the panel in pixels. */
        this.width = '350px';
        /** Height of the panel in pixels. */
        this.height = '700px';
        /** Enable floating mode with elevated shadow. */
        this.floating = false;
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-panel'];
        const propClasses = convertPropsToClasses({
            floating: this.floating,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: '8eca46effdba5d4c7c027a78d98e8914247d0187' }, h("div", Object.assign({ key: '106cf95c53ce95e4afa6a58845b979d575b3b318', class: this.getClasses(), style: { width: this.width, height: this.height } }, this.inheritedAttributes), h("div", { key: '200d9a028744c3027fea736461d50c27eca1ad31', class: "modus-wc-panel-header" }, h("slot", { key: '6ef2682a8d4eeb0f253ac6602da9a980b026be84', name: "header" })), h("div", { key: '61b44ce89e87e1364a18fb366f464ee25881a5c5', class: "modus-wc-panel-body" }, h("slot", { key: '4c6c82a0ad09e81b74dbe9c478f5793eaac702ee', name: "body" })), h("div", { key: '2806ffb3225c9c361b55b12e9088cb4326cbf463', class: "modus-wc-panel-footer" }, h("slot", { key: '577fadfaeb76f7c2f475685756b7da6dbb5c05fb', name: "footer" })))));
    }
    get el() { return this; }
    static get style() { return modusWcPanelCss; }
}, [4, "modus-wc-panel", {
        "customClass": [1, "custom-class"],
        "width": [1],
        "height": [1],
        "floating": [4]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-panel"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-panel":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcPanel);
            }
            break;
    } });
}

export { ModusWcPanel as M, defineCustomElement as d };
