// @trimble-oss/moduswebcomponents/components/p-hDsQ33Kf.js
import { p as proxyCustomElement, H, h, c as Host } from './p-X1tirp06.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';

const convertPropsToClasses = ({ size, weight, }) => {
    let classes = '';
    if (size) {
        classes = `${classes} modus-wc-text-${size}`;
    }
    if (weight) {
        classes = `${classes} modus-wc-typography-weight-${weight}`;
    }
    return classes.trim();
};

const modusWcTypographyCss = "modus-wc-typography h1.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-3xl);font-weight:var(--modus-wc-font-weight-normal);letter-spacing:0.5px;line-height:36px}modus-wc-typography h2.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-2xl);font-weight:var(--modus-wc-font-weight-normal);letter-spacing:0.15px;line-height:30px}modus-wc-typography h3.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-xl);font-weight:var(--modus-wc-font-weight-semibold);letter-spacing:0.15px;line-height:27px}modus-wc-typography h4.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-lg);font-weight:var(--modus-wc-font-weight-semibold);letter-spacing:0.15px;line-height:24px}modus-wc-typography h5.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-md);font-weight:var(--modus-wc-font-weight-bold);letter-spacing:0.45px;line-height:24px}modus-wc-typography h6.modus-wc-typography:not(.modus-wc-typography-override){font-size:var(--modus-wc-font-size-sm);font-weight:var(--modus-wc-font-weight-bold);letter-spacing:0.45px;line-height:18px}modus-wc-typography .modus-wc-typography-weight-light{font-weight:var(--modus-wc-font-weight-light)}modus-wc-typography .modus-wc-typography-weight-normal{font-weight:var(--modus-wc-font-weight-normal)}modus-wc-typography .modus-wc-typography-weight-semibold{font-weight:var(--modus-wc-font-weight-semibold)}modus-wc-typography .modus-wc-typography-weight-bold{font-weight:var(--modus-wc-font-weight-bold)}modus-wc-typography .modus-wc-text-xs{font-size:var(--modus-wc-font-size-xs)}modus-wc-typography .modus-wc-text-sm{font-size:var(--modus-wc-font-size-sm)}modus-wc-typography .modus-wc-text-md{font-size:var(--modus-wc-font-size-md)}modus-wc-typography .modus-wc-text-lg{font-size:var(--modus-wc-font-size-lg)}modus-wc-typography .modus-wc-text-xl{font-size:var(--modus-wc-font-size-xl)}modus-wc-typography .modus-wc-text-2xl{font-size:var(--modus-wc-font-size-2xl)}modus-wc-typography .modus-wc-text-3xl{font-size:var(--modus-wc-font-size-3xl)}[data-theme=connect-light] modus-wc-typography :not(a),[data-theme=connect-dark] modus-wc-typography :not(a){color:var(--modus-wc-color-base-content-low-contrast)}[data-theme=connect-light] modus-wc-typography h1.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h1.modus-wc-typography{font-size:var(--modus-wc-font-size-4xl);font-weight:var(--modus-wc-font-weight-normal);letter-spacing:0.5px;line-height:30px}[data-theme=connect-light] modus-wc-typography h2.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h2.modus-wc-typography{font-size:var(--modus-wc-font-size-3xl);font-weight:var(--modus-wc-font-weight-semibold);letter-spacing:0.15px;line-height:24px}[data-theme=connect-light] modus-wc-typography h3.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h3.modus-wc-typography{font-size:var(--modus-wc-font-size-xl);font-weight:var(--modus-wc-font-weight-semibold);letter-spacing:0.15px;line-height:18px}[data-theme=connect-light] modus-wc-typography h4.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h4.modus-wc-typography{font-size:var(--modus-wc-font-size-lg);letter-spacing:0.15px;line-height:24px}[data-theme=connect-light] modus-wc-typography h5.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h5.modus-wc-typography{font-size:var(--modus-wc-font-size-md);font-weight:var(--modus-wc-font-weight-bold);letter-spacing:0.45px;line-height:24px}[data-theme=connect-light] modus-wc-typography h6.modus-wc-typography,[data-theme=connect-dark] modus-wc-typography h6.modus-wc-typography{font-size:var(--modus-wc-font-size-xs);font-weight:var(--modus-wc-font-weight-bold);letter-spacing:0.45px;line-height:18px;text-transform:uppercase}";

const ModusWCTypography = /*@__PURE__*/ proxyCustomElement(class ModusWCTypography extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the typography element. */
        this.customClass = '';
        /** The hierarchy of the typography component. */
        this.hierarchy = 'p';
        /** The size of the font. */
        this.size = 'md';
        /** The weight of the text. */
        this.weight = 'normal';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-typography'];
        // Check if we're dealing with a heading and have size/weight overrides
        const isHeading = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].includes(this.hierarchy);
        const hasOverrides = this.size !== 'md' || this.weight !== 'normal';
        if (isHeading && hasOverrides) {
            // Add a class to indicate overrides for headings
            classList.push('modus-wc-typography-override');
        }
        const propClasses = convertPropsToClasses({
            size: this.size,
            weight: this.weight,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        const Element = this.hierarchy;
        return (h(Host, { key: 'b9f340f7348ccf18ae168020cccabe14a1a8c5de' }, h(Element, Object.assign({ key: '6bf43d72c246ac1d38b241f8596d953dbed044a0', class: this.getClasses() }, this.inheritedAttributes), h("slot", { key: '7f59a5319bbf237101acaa67065bf5bec3565c4e' }, this.label))));
    }
    get el() { return this; }
    static get style() { return modusWcTypographyCss; }
}, [4, "modus-wc-typography", {
        "customClass": [1, "custom-class"],
        "hierarchy": [1],
        "label": [1],
        "size": [1],
        "weight": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-typography"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-typography":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWCTypography);
            }
            break;
    } });
}

export { ModusWCTypography as M, defineCustomElement as d };
