// @trimble-oss/moduswebcomponents/components/p-vnbqBzVH.js
import { p as proxyCustomElement, H, h, c as Host } from './p-X1tirp06.js';
import { h as handleShadowDOMStyles } from './p-CKNBKeF3.js';
import { i as inheritAriaAttributes } from './p-CPG9FZRU.js';

const modusWcIconCss = "modus-wc-icon .modus-icons.modus-wc-icon--xs,modus-wc-icon .modus-icons-outlined.modus-wc-icon--xs,modus-wc-icon .modus-icons-solid.modus-wc-icon--xs{font-size:1rem}modus-wc-icon .modus-icons.modus-wc-icon--sm,modus-wc-icon .modus-icons-outlined.modus-wc-icon--sm,modus-wc-icon .modus-icons-solid.modus-wc-icon--sm{font-size:1.25rem}modus-wc-icon .modus-icons.modus-wc-icon--md,modus-wc-icon .modus-icons-outlined.modus-wc-icon--md,modus-wc-icon .modus-icons-solid.modus-wc-icon--md{font-size:1.5rem}modus-wc-icon .modus-icons.modus-wc-icon--lg,modus-wc-icon .modus-icons-outlined.modus-wc-icon--lg,modus-wc-icon .modus-icons-solid.modus-wc-icon--lg{font-size:2rem}modus-wc-icon .modus-icons.modus-wc-icon--xl,modus-wc-icon .modus-icons-outlined.modus-wc-icon--xl,modus-wc-icon .modus-icons-solid.modus-wc-icon--xl{font-size:4rem}modus-wc-icon i::before,modus-wc-icon i::after{font-family:inherit !important}";

const ModusWcIcon = /*@__PURE__*/ proxyCustomElement(class ModusWcIcon extends H {
    constructor() {
        super();
        this.__registerHost();
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the i element. */
        this.customClass = '';
        /** Indicates that the icon is decorative. When true, sets aria-hidden to hide the icon from screen readers. */
        this.decorative = true;
        /** The icon size, can be "sm", "md", "lg" (a custom size can be specified in CSS). This adjusts the font size for the icon. */
        this.size = 'md';
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        if (!this.decorative && !this.el.ariaLabel) {
            this.el.ariaLabel = `${this.name} icon`;
        }
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        let classList = [];
        // Add base class
        classList.push('modus-wc-icon');
        // Add icon font class based on variant
        if (this.variant === 'outlined') {
            classList.push('modus-icons-outlined');
        }
        else if (this.variant === 'solid') {
            classList.push('modus-icons-solid');
        }
        else {
            classList.push('modus-icons');
        }
        // Add size class - this is common for all variants
        classList.push(`modus-wc-icon--${this.size}`);
        // Add custom class if provided
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        const ariaHidden = this.decorative ? 'true' : null;
        const role = this.decorative ? undefined : 'img';
        return (h(Host, { key: '8736cfaf7fdb23ec43095d97a989feae1bcb33ef', class: "modus-wc-flex modus-wc-items-center" }, h("i", Object.assign({ key: '5bb45c103f9fc86924ac286258f9a230b7b1d08a', "aria-hidden": ariaHidden, "aria-label": this.decorative ? null : this.el.ariaLabel, class: this.getClasses(), role: role, tabindex: -1 }, this.inheritedAttributes), this.name)));
    }
    get el() { return this; }
    static get style() { return modusWcIconCss; }
}, [0, "modus-wc-icon", {
        "customClass": [1, "custom-class"],
        "decorative": [4],
        "name": [1],
        "size": [1],
        "variant": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-wc-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-wc-icon":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusWcIcon);
            }
            break;
    } });
}

export { ModusWcIcon as M, defineCustomElement as d };
