// @trimble-oss/moduswebcomponents/esm/modus-wc-toast.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-D5XAQERj.js';
import { i as inheritAriaAttributes } from './utils-BFec1ApF.js';

const convertPropsToClasses = ({ position, }) => {
    let classes = '';
    if (position) {
        switch (position) {
            case 'top-start':
                classes = `${classes} modus-wc-toast-top modus-wc-toast-start`;
                break;
            case 'top-center':
                classes = `${classes} modus-wc-toast-top modus-wc-toast-center`;
                break;
            case 'top-end':
                classes = `${classes} modus-wc-toast-top modus-wc-toast-end`;
                break;
            case 'middle-start':
                classes = `${classes} modus-wc-toast-start modus-wc-toast-middle`;
                break;
            case 'middle-center':
                classes = `${classes} modus-wc-toast-center modus-wc-toast-middle`;
                break;
            case 'middle-end':
                classes = `${classes} modus-wc-toast-end modus-wc-toast-middle`;
                break;
            case 'bottom-start':
                classes = `${classes} modus-wc-toast-start`;
                break;
            case 'bottom-center':
                classes = `${classes} modus-wc-toast-center`;
                break;
            case 'bottom-end':
                classes = `${classes} modus-wc-toast-end`;
                break;
        }
    }
    return classes.trim();
};

const modusWcToastCss = "";

const ModusWcToast = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.inheritedAttributes = {};
        /** Additional classes for custom styling. */
        this.customClass = '';
        /** The position of the toast in the parent container. */
        this.position = 'top-end';
        this.getClasses = () => {
            const classList = ['modus-wc-toast'];
            const propClasses = convertPropsToClasses({
                position: this.position,
            });
            // The order CSS classes are added matters to CSS specificity
            if (propClasses)
                classList.push(propClasses);
            if (this.customClass)
                classList.push(this.customClass);
            return classList.join(' ');
        };
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    delayChanged(newDelay) {
        clearTimeout(this.timerId);
        this.timerId = setTimeout(() => {
            this.dismissElement();
        }, newDelay);
    }
    dismissElement() {
        this.el.remove();
    }
    componentDidLoad() {
        if (this.delay && this.delay > 0) {
            this.timerId = setTimeout(() => {
                this.dismissElement();
            }, this.delay);
        }
    }
    disconnectedCallback() {
        clearTimeout(this.timerId);
    }
    render() {
        return (h(Host, { key: 'e94c591261fd9d32f89feb19680f54092fabeebe' }, h("div", Object.assign({ key: '0221409bd3d21a67376e70c3e1b180e76efaa7b1', class: this.getClasses() }, this.inheritedAttributes), h("slot", { key: 'fc565ef2cc0b755dd00df96bdf679518ec672857' }))));
    }
    get el() { return getElement(this); }
    static get watchers() { return {
        "delay": ["delayChanged"]
    }; }
};
ModusWcToast.style = modusWcToastCss;

export { ModusWcToast as modus_wc_toast };
