// @trimble-oss/moduswebcomponents/esm/modus-wc-input-label.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-D5XAQERj.js';
import { h as handleShadowDOMStyles } from './base-component-CKNBKeF3.js';
import { i as inheritAriaAttributes } from './utils-BFec1ApF.js';

const modusWcInputLabelCss = "modus-wc-input-label.modus-wc-input-label-host{align-items:center;display:flex}modus-wc-input-label .modus-wc-input-label{font-weight:var(--modus-wc-font-weight-bold)}modus-wc-input-label .modus-wc-input-label.modus-wc-input-label-size-sm{font-size:var(--modus-wc-font-size-xs)}modus-wc-input-label .modus-wc-input-label.modus-wc-input-label-size-sm .modus-wc-input-label-sublabel{font-size:var(--modus-wc-font-size-2xs)}modus-wc-input-label .modus-wc-input-label.modus-wc-input-label-size-md{font-size:var(--modus-wc-font-size-sm)}modus-wc-input-label .modus-wc-input-label.modus-wc-input-label-size-md .modus-wc-input-label-sublabel{font-size:var(--modus-wc-font-size-xs)}modus-wc-input-label .modus-wc-input-label.modus-wc-input-label-size-lg{font-size:var(--modus-wc-font-size-md)}modus-wc-input-label .modus-wc-input-label.modus-wc-input-label-size-lg .modus-wc-input-label-sublabel{font-size:var(--modus-wc-font-size-sm)}modus-wc-input-label .modus-wc-input-label .required-indicator{color:var(--modus-wc-color-error)}[data-theme=modus-classic-light] modus-wc-input-label .modus-wc-input-label{color:var(--modus-wc-color-gray-8)}[data-theme=modus-classic-dark] modus-wc-input-label .modus-wc-input-label{color:var(--modus-wc-color-gray-2)}[data-theme=connect-light] modus-wc-input-label .modus-wc-input-label{color:oklch(var(--s));font-weight:var(--modus-wc-font-weight-normal);margin-bottom:0;min-height:var(--modus-wc-line-height-md)}[data-theme=connect-dark] modus-wc-input-label .modus-wc-input-label{color:var(--modus-wc-color-white);font-weight:var(--modus-wc-font-weight-normal);margin-bottom:0;min-height:var(--modus-wc-line-height-md)}";

const ModusWcInputLabel = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.inheritedAttributes = {};
        /** Additional classes for custom styling. */
        this.customClass = '';
        /** Whether the label indicates a required field. */
        this.required = false;
        /** The size of the label. */
        this.size = 'md';
        this.getClasses = () => {
            const classList = ['modus-wc-input-label'];
            // The order CSS classes are added matters to CSS specificity
            if (this.size)
                classList.push(`modus-wc-input-label-size-${this.size}`);
            if (this.customClass)
                classList.push(this.customClass);
            return classList.join(' ');
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    render() {
        return (h(Host, { key: 'a1b1faf45264587fbf52653c0b3c8d1c3371f13d', class: "modus-wc-input-label-host" }, h("label", Object.assign({ key: '0ee1830f85c03ee70082e67b2c73224f84b328c3', class: this.getClasses(), htmlFor: this.forId }, this.inheritedAttributes), this.labelText, this.required && (h("span", { key: '11b6749526377e5900b5a60df1faba67e86e8337', "aria-hidden": "true", class: "required-indicator" }, '\u00A0*')), this.subLabelText && (h("div", { key: '8f403f2fc9f01733b6ac263471875146ff8fe199', class: "modus-wc-input-label-sublabel" }, this.subLabelText)), h("slot", { key: '7c0c7ef4d598a29383c21c1378522c588fcd88eb' }))));
    }
    get el() { return getElement(this); }
};
ModusWcInputLabel.style = modusWcInputLabelCss;

export { ModusWcInputLabel as modus_wc_input_label };
