// @trimble-oss/moduswebcomponents/esm/modus-wc-slider.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-D5XAQERj.js';
import { h as handleShadowDOMStyles } from './base-component-CKNBKeF3.js';
import { i as inheritAriaAttributes } from './utils-BFec1ApF.js';

const convertPropsToClasses = ({ size, }) => {
    let classes = '';
    if (size) {
        classes = `${classes} modus-wc-range-${size}`;
    }
    return classes.trim();
};

const modusWcSliderCss = "modus-wc-slider .modus-wc-range{--rounded-box:200px;--range-shadow:var(--modus-wc-color-trimble-blue)}modus-wc-slider .modus-wc-range .modus-wc-input-label{padding-bottom:var(--modus-wc-spacing-sm)}[data-theme=modus-classic-light] modus-wc-slider .modus-wc-range,[data-theme=modus-classic-dark] modus-wc-slider .modus-wc-range{--rounded-box:200px;--range-shadow:var(--modus-wc-color-trimble-blue)}";

const ModusWcSlider = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.inputBlur = createEvent(this, "inputBlur");
        this.inputChange = createEvent(this, "inputChange");
        this.inputFocus = createEvent(this, "inputFocus");
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The disabled state of the slider. */
        this.disabled = false;
        /** Name of the form control. Submitted with the form as part of a name/value pair. */
        this.name = '';
        /** A value is required for the form to be submittable. */
        this.required = false;
        /** The size of the input. */
        this.size = 'md';
        /** The value of the slider. */
        this.value = 0;
        this.handleBlur = (event) => {
            this.inputBlur.emit(event);
        };
        this.handleFocus = (event) => {
            this.inputFocus.emit(event);
        };
        this.handleInput = (event) => {
            this.inputChange.emit(event);
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Slider';
        }
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-range'];
        const propClasses = convertPropsToClasses({ size: this.size });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: 'fcf03fb6f845e045fc118b4c4b9b7fd9678d1dda' }, this.label && (h("modus-wc-input-label", { key: '4da0e14edc26d12e65b52964615e320b6059bee5', forId: this.inputId, labelText: this.label, required: this.required, size: this.size })), h("input", Object.assign({ key: '7ef8bd9690f4ed2352015c8038ff48bfd47a7673', "aria-disabled": this.disabled, class: this.getClasses(), disabled: this.disabled, id: this.inputId, max: this.max, min: this.min, name: this.name, onBlur: this.handleBlur, onFocus: this.handleFocus, onInput: this.handleInput, required: this.required, step: this.step, tabIndex: this.inputTabIndex, type: "range", value: this.value }, this.inheritedAttributes))));
    }
    get el() { return getElement(this); }
};
ModusWcSlider.style = modusWcSliderCss;

export { ModusWcSlider as modus_wc_slider };
