// @trimble-oss/moduswebcomponents/collection/components/modus-wc-stepper/modus-wc-stepper.js
import { h, Host } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-stepper.tailwind";
import { inheritAriaAttributes } from "../utils";
/**
 * Used to show a list of steps in a process.
 */
export class ModusWcStepper {
    constructor() {
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the steps element. */
        this.customClass = '';
        /** The steps to display. */
        this.steps = [];
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-steps'];
        const propClasses = convertPropsToClasses({
            orientation: this.orientation,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getClassesForStep(step) {
        const classList = ['modus-wc-step'];
        // The order CSS classes are added matters to CSS specificity
        if (step.color)
            classList.push(`modus-wc-step-${step.color}`);
        if (step.customClass)
            classList.push(step.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: '7d8e5badd7eedcdb7fc6a997b94f12008c5bfcac' }, h("div", Object.assign({ key: 'f3ce9e091ca2418bda52cae1661d2dae7d04fadd', class: this.getClasses() }, this.inheritedAttributes), this.steps.map((step, index) => {
            return (h("li", { class: this.getClassesForStep(step), key: index, "data-content": step.content }, step.label));
        }))));
    }
    static get is() { return "modus-wc-stepper"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-stepper.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-stepper.css"]
        };
    }
    static get properties() {
        return {
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the steps element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "orientation": {
                "type": "string",
                "attribute": "orientation",
                "mutable": false,
                "complexType": {
                    "original": "Orientation",
                    "resolved": "\"horizontal\" | \"vertical\" | undefined",
                    "references": {
                        "Orientation": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::Orientation"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The orientation of the steps."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "steps": {
                "type": "unknown",
                "attribute": "steps",
                "mutable": false,
                "complexType": {
                    "original": "IStepperItem[]",
                    "resolved": "IStepperItem[]",
                    "references": {
                        "IStepperItem": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-stepper/modus-wc-stepper.tsx",
                            "id": "src/components/modus-wc-stepper/modus-wc-stepper.tsx::IStepperItem"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The steps to display."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            }
        };
    }
    static get elementRef() { return "el"; }
}
