// @trimble-oss/moduswebcomponents/collection/providers/theme/theme-provider.js
import { h } from "@stencil/core";
import { initializeThemeStore, themeStore, watchSystemTheme, } from "./theme.store";
export class ThemeProvider {
    componentWillLoad() {
        initializeThemeStore(this.initialTheme);
        this.cleanup = watchSystemTheme();
        // Subscribe to specific property changes
        this.modeUnsubscribe = themeStore.onChange('mode', () => {
            this.updateThemeClasses();
        });
        this.themeUnsubscribe = themeStore.onChange('theme', () => {
            this.updateThemeClasses();
        });
        this.updateThemeClasses();
    }
    disconnectedCallback() {
        var _a, _b, _c;
        (_a = this.cleanup) === null || _a === void 0 ? void 0 : _a.call(this);
        (_b = this.modeUnsubscribe) === null || _b === void 0 ? void 0 : _b.call(this);
        (_c = this.themeUnsubscribe) === null || _c === void 0 ? void 0 : _c.call(this);
    }
    updateThemeClasses() {
        const root = document.documentElement;
        const { mode, theme } = themeStore.state;
        // Update theme and mode attributes
        root.setAttribute('data-theme', `${theme}-${mode}`);
        root.setAttribute('data-mode', mode);
        // Remove old mode classes
        root.classList.remove('light', 'dark');
        root.classList.add(mode);
        // Store preferences
        localStorage.setItem('modus-theme-config', JSON.stringify({ mode, theme }));
    }
    render() {
        return h("slot", { key: 'd3409eb8bdbf538d51eeacbb5b70f99d496b2d40' });
    }
    static get is() { return "modus-wc-theme-provider"; }
    static get properties() {
        return {
            "initialTheme": {
                "type": "unknown",
                "attribute": "initial-theme",
                "mutable": false,
                "complexType": {
                    "original": "Partial<IThemeConfig>",
                    "resolved": "undefined | { mode?: ThemeMode | undefined; theme?: ThemeName | undefined; }",
                    "references": {
                        "Partial": {
                            "location": "global",
                            "id": "global::Partial"
                        },
                        "IThemeConfig": {
                            "location": "import",
                            "path": "./theme.types",
                            "id": "src/providers/theme/theme.types.ts::IThemeConfig"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "getter": false,
                "setter": false
            }
        };
    }
}
