// @trimble-oss/moduswebcomponents-react/stencil-generated/components.js
'use client';
import { createComponent } from '@stencil/react-output-target/runtime';
import { ModusWcAccordion as ModusWcAccordionElement, defineCustomElement as defineModusWcAccordion } from "@trimble-oss/moduswebcomponents/components/modus-wc-accordion.js";
import { ModusWcAlert as ModusWcAlertElement, defineCustomElement as defineModusWcAlert } from "@trimble-oss/moduswebcomponents/components/modus-wc-alert.js";
import { ModusWcAutocomplete as ModusWcAutocompleteElement, defineCustomElement as defineModusWcAutocomplete } from "@trimble-oss/moduswebcomponents/components/modus-wc-autocomplete.js";
import { ModusWcAvatar as ModusWcAvatarElement, defineCustomElement as defineModusWcAvatar } from "@trimble-oss/moduswebcomponents/components/modus-wc-avatar.js";
import { ModusWcBadge as ModusWcBadgeElement, defineCustomElement as defineModusWcBadge } from "@trimble-oss/moduswebcomponents/components/modus-wc-badge.js";
import { ModusWcBreadcrumbs as ModusWcBreadcrumbsElement, defineCustomElement as defineModusWcBreadcrumbs } from "@trimble-oss/moduswebcomponents/components/modus-wc-breadcrumbs.js";
import { ModusWcButtonGroup as ModusWcButtonGroupElement, defineCustomElement as defineModusWcButtonGroup } from "@trimble-oss/moduswebcomponents/components/modus-wc-button-group.js";
import { ModusWcButton as ModusWcButtonElement, defineCustomElement as defineModusWcButton } from "@trimble-oss/moduswebcomponents/components/modus-wc-button.js";
import { ModusWcCard as ModusWcCardElement, defineCustomElement as defineModusWcCard } from "@trimble-oss/moduswebcomponents/components/modus-wc-card.js";
import { ModusWcCheckbox as ModusWcCheckboxElement, defineCustomElement as defineModusWcCheckbox } from "@trimble-oss/moduswebcomponents/components/modus-wc-checkbox.js";
import { ModusWcChip as ModusWcChipElement, defineCustomElement as defineModusWcChip } from "@trimble-oss/moduswebcomponents/components/modus-wc-chip.js";
import { ModusWcCollapse as ModusWcCollapseElement, defineCustomElement as defineModusWcCollapse } from "@trimble-oss/moduswebcomponents/components/modus-wc-collapse.js";
import { ModusWcDate as ModusWcDateElement, defineCustomElement as defineModusWcDate } from "@trimble-oss/moduswebcomponents/components/modus-wc-date.js";
import { ModusWcDivider as ModusWcDividerElement, defineCustomElement as defineModusWcDivider } from "@trimble-oss/moduswebcomponents/components/modus-wc-divider.js";
import { ModusWcDropdownMenu as ModusWcDropdownMenuElement, defineCustomElement as defineModusWcDropdownMenu } from "@trimble-oss/moduswebcomponents/components/modus-wc-dropdown-menu.js";
import { ModusWcFileDropzone as ModusWcFileDropzoneElement, defineCustomElement as defineModusWcFileDropzone } from "@trimble-oss/moduswebcomponents/components/modus-wc-file-dropzone.js";
import { ModusWcHandle as ModusWcHandleElement, defineCustomElement as defineModusWcHandle } from "@trimble-oss/moduswebcomponents/components/modus-wc-handle.js";
import { ModusWcIcon as ModusWcIconElement, defineCustomElement as defineModusWcIcon } from "@trimble-oss/moduswebcomponents/components/modus-wc-icon.js";
import { ModusWcInputFeedback as ModusWcInputFeedbackElement, defineCustomElement as defineModusWcInputFeedback } from "@trimble-oss/moduswebcomponents/components/modus-wc-input-feedback.js";
import { ModusWcInputLabel as ModusWcInputLabelElement, defineCustomElement as defineModusWcInputLabel } from "@trimble-oss/moduswebcomponents/components/modus-wc-input-label.js";
import { ModusWcLoader as ModusWcLoaderElement, defineCustomElement as defineModusWcLoader } from "@trimble-oss/moduswebcomponents/components/modus-wc-loader.js";
import { ModusWcLogo as ModusWcLogoElement, defineCustomElement as defineModusWcLogo } from "@trimble-oss/moduswebcomponents/components/modus-wc-logo.js";
import { ModusWcMenuItem as ModusWcMenuItemElement, defineCustomElement as defineModusWcMenuItem } from "@trimble-oss/moduswebcomponents/components/modus-wc-menu-item.js";
import { ModusWcMenu as ModusWcMenuElement, defineCustomElement as defineModusWcMenu } from "@trimble-oss/moduswebcomponents/components/modus-wc-menu.js";
import { ModusWcModal as ModusWcModalElement, defineCustomElement as defineModusWcModal } from "@trimble-oss/moduswebcomponents/components/modus-wc-modal.js";
import { ModusWcNavbar as ModusWcNavbarElement, defineCustomElement as defineModusWcNavbar } from "@trimble-oss/moduswebcomponents/components/modus-wc-navbar.js";
import { ModusWcNumberInput as ModusWcNumberInputElement, defineCustomElement as defineModusWcNumberInput } from "@trimble-oss/moduswebcomponents/components/modus-wc-number-input.js";
import { ModusWcPagination as ModusWcPaginationElement, defineCustomElement as defineModusWcPagination } from "@trimble-oss/moduswebcomponents/components/modus-wc-pagination.js";
import { ModusWcPanel as ModusWcPanelElement, defineCustomElement as defineModusWcPanel } from "@trimble-oss/moduswebcomponents/components/modus-wc-panel.js";
import { ModusWcProfileMenu as ModusWcProfileMenuElement, defineCustomElement as defineModusWcProfileMenu } from "@trimble-oss/moduswebcomponents/components/modus-wc-profile-menu.js";
import { ModusWcProgress as ModusWcProgressElement, defineCustomElement as defineModusWcProgress } from "@trimble-oss/moduswebcomponents/components/modus-wc-progress.js";
import { ModusWcRadio as ModusWcRadioElement, defineCustomElement as defineModusWcRadio } from "@trimble-oss/moduswebcomponents/components/modus-wc-radio.js";
import { ModusWcRating as ModusWcRatingElement, defineCustomElement as defineModusWcRating } from "@trimble-oss/moduswebcomponents/components/modus-wc-rating.js";
import { ModusWcSelect as ModusWcSelectElement, defineCustomElement as defineModusWcSelect } from "@trimble-oss/moduswebcomponents/components/modus-wc-select.js";
import { ModusWcSideNavigation as ModusWcSideNavigationElement, defineCustomElement as defineModusWcSideNavigation } from "@trimble-oss/moduswebcomponents/components/modus-wc-side-navigation.js";
import { ModusWcSkeleton as ModusWcSkeletonElement, defineCustomElement as defineModusWcSkeleton } from "@trimble-oss/moduswebcomponents/components/modus-wc-skeleton.js";
import { ModusWcSlider as ModusWcSliderElement, defineCustomElement as defineModusWcSlider } from "@trimble-oss/moduswebcomponents/components/modus-wc-slider.js";
import { ModusWcStepper as ModusWcStepperElement, defineCustomElement as defineModusWcStepper } from "@trimble-oss/moduswebcomponents/components/modus-wc-stepper.js";
import { ModusWcSwitch as ModusWcSwitchElement, defineCustomElement as defineModusWcSwitch } from "@trimble-oss/moduswebcomponents/components/modus-wc-switch.js";
import { ModusWcTable as ModusWcTableElement, defineCustomElement as defineModusWcTable } from "@trimble-oss/moduswebcomponents/components/modus-wc-table.js";
import { ModusWcTabs as ModusWcTabsElement, defineCustomElement as defineModusWcTabs } from "@trimble-oss/moduswebcomponents/components/modus-wc-tabs.js";
import { ModusWcTextInput as ModusWcTextInputElement, defineCustomElement as defineModusWcTextInput } from "@trimble-oss/moduswebcomponents/components/modus-wc-text-input.js";
import { ModusWcTextarea as ModusWcTextareaElement, defineCustomElement as defineModusWcTextarea } from "@trimble-oss/moduswebcomponents/components/modus-wc-textarea.js";
import { ModusWcThemeProvider as ModusWcThemeProviderElement, defineCustomElement as defineModusWcThemeProvider } from "@trimble-oss/moduswebcomponents/components/modus-wc-theme-provider.js";
import { ModusWcThemeSwitcher as ModusWcThemeSwitcherElement, defineCustomElement as defineModusWcThemeSwitcher } from "@trimble-oss/moduswebcomponents/components/modus-wc-theme-switcher.js";
import { ModusWcTimeInput as ModusWcTimeInputElement, defineCustomElement as defineModusWcTimeInput } from "@trimble-oss/moduswebcomponents/components/modus-wc-time-input.js";
import { ModusWcToast as ModusWcToastElement, defineCustomElement as defineModusWcToast } from "@trimble-oss/moduswebcomponents/components/modus-wc-toast.js";
import { ModusWcToolbar as ModusWcToolbarElement, defineCustomElement as defineModusWcToolbar } from "@trimble-oss/moduswebcomponents/components/modus-wc-toolbar.js";
import { ModusWcTooltip as ModusWcTooltipElement, defineCustomElement as defineModusWcTooltip } from "@trimble-oss/moduswebcomponents/components/modus-wc-tooltip.js";
import { ModusWcTypography as ModusWcTypographyElement, defineCustomElement as defineModusWcTypography } from "@trimble-oss/moduswebcomponents/components/modus-wc-typography.js";
import { ModusWcUtilityPanel as ModusWcUtilityPanelElement, defineCustomElement as defineModusWcUtilityPanel } from "@trimble-oss/moduswebcomponents/components/modus-wc-utility-panel.js";
import React from 'react';
export const ModusWcAccordion = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-accordion',
    elementClass: ModusWcAccordionElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onExpandedChange: 'expandedChange' },
    defineCustomElement: defineModusWcAccordion
});
export const ModusWcAlert = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-alert',
    elementClass: ModusWcAlertElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onDismissClick: 'dismissClick' },
    defineCustomElement: defineModusWcAlert
});
export const ModusWcAutocomplete = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-autocomplete',
    elementClass: ModusWcAutocompleteElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onChipRemove: 'chipRemove',
        onChipsExpansionChange: 'chipsExpansionChange',
        onClearClick: 'clearClick',
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus',
        onItemSelect: 'itemSelect'
    },
    defineCustomElement: defineModusWcAutocomplete
});
export const ModusWcAvatar = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-avatar',
    elementClass: ModusWcAvatarElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcAvatar
});
export const ModusWcBadge = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-badge',
    elementClass: ModusWcBadgeElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcBadge
});
export const ModusWcBreadcrumbs = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-breadcrumbs',
    elementClass: ModusWcBreadcrumbsElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onBreadcrumbClick: 'breadcrumbClick' },
    defineCustomElement: defineModusWcBreadcrumbs
});
export const ModusWcButton = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-button',
    elementClass: ModusWcButtonElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onButtonClick: 'buttonClick' },
    defineCustomElement: defineModusWcButton
});
export const ModusWcButtonGroup = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-button-group',
    elementClass: ModusWcButtonGroupElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onButtonGroupClick: 'buttonGroupClick',
        onButtonSelectionChange: 'buttonSelectionChange'
    },
    defineCustomElement: defineModusWcButtonGroup
});
export const ModusWcCard = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-card',
    elementClass: ModusWcCardElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcCard
});
export const ModusWcCheckbox = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-checkbox',
    elementClass: ModusWcCheckboxElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcCheckbox
});
export const ModusWcChip = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-chip',
    elementClass: ModusWcChipElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onChipClick: 'chipClick',
        onChipRemove: 'chipRemove'
    },
    defineCustomElement: defineModusWcChip
});
export const ModusWcCollapse = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-collapse',
    elementClass: ModusWcCollapseElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onExpandedChange: 'expandedChange' },
    defineCustomElement: defineModusWcCollapse
});
export const ModusWcDate = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-date',
    elementClass: ModusWcDateElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus',
        onCalendarMonthChange: 'calendarMonthChange',
        onCalendarYearChange: 'calendarYearChange'
    },
    defineCustomElement: defineModusWcDate
});
export const ModusWcDivider = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-divider',
    elementClass: ModusWcDividerElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcDivider
});
export const ModusWcDropdownMenu = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-dropdown-menu',
    elementClass: ModusWcDropdownMenuElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onMenuVisibilityChange: 'menuVisibilityChange' },
    defineCustomElement: defineModusWcDropdownMenu
});
export const ModusWcFileDropzone = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-file-dropzone',
    elementClass: ModusWcFileDropzoneElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onFileSelect: 'fileSelect' },
    defineCustomElement: defineModusWcFileDropzone
});
export const ModusWcHandle = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-handle',
    elementClass: ModusWcHandleElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcHandle
});
export const ModusWcIcon = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-icon',
    elementClass: ModusWcIconElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcIcon
});
export const ModusWcInputFeedback = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-input-feedback',
    elementClass: ModusWcInputFeedbackElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcInputFeedback
});
export const ModusWcInputLabel = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-input-label',
    elementClass: ModusWcInputLabelElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcInputLabel
});
export const ModusWcLoader = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-loader',
    elementClass: ModusWcLoaderElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcLoader
});
export const ModusWcLogo = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-logo',
    elementClass: ModusWcLogoElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcLogo
});
export const ModusWcMenu = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-menu',
    elementClass: ModusWcMenuElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onMenuFocusout: 'menuFocusout',
        onMenuSelectionChange: 'menuSelectionChange'
    },
    defineCustomElement: defineModusWcMenu
});
export const ModusWcMenuItem = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-menu-item',
    elementClass: ModusWcMenuItemElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onItemSelect: 'itemSelect' },
    defineCustomElement: defineModusWcMenuItem
});
export const ModusWcModal = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-modal',
    elementClass: ModusWcModalElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcModal
});
export const ModusWcNavbar = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-navbar',
    elementClass: ModusWcNavbarElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onAiClick: 'aiClick',
        onAppsClick: 'appsClick',
        onAppsMenuOpenChange: 'appsMenuOpenChange',
        onCondensedMenuOpenChange: 'condensedMenuOpenChange',
        onHelpClick: 'helpClick',
        onMainMenuOpenChange: 'mainMenuOpenChange',
        onMyTrimbleClick: 'myTrimbleClick',
        onNotificationsClick: 'notificationsClick',
        onNotificationsMenuOpenChange: 'notificationsMenuOpenChange',
        onSearchChange: 'searchChange',
        onSearchClick: 'searchClick',
        onSearchInputOpenChange: 'searchInputOpenChange',
        onSignOutClick: 'signOutClick',
        onTrimbleLogoClick: 'trimbleLogoClick',
        onUserMenuOpenChange: 'userMenuOpenChange'
    },
    defineCustomElement: defineModusWcNavbar
});
export const ModusWcNumberInput = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-number-input',
    elementClass: ModusWcNumberInputElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcNumberInput
});
export const ModusWcPagination = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-pagination',
    elementClass: ModusWcPaginationElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onPageChange: 'pageChange' },
    defineCustomElement: defineModusWcPagination
});
export const ModusWcPanel = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-panel',
    elementClass: ModusWcPanelElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcPanel
});
export const ModusWcProfileMenu = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-profile-menu',
    elementClass: ModusWcProfileMenuElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onSignOutClick: 'signOutClick',
        onMenuItemClick: 'menuItemClick'
    },
    defineCustomElement: defineModusWcProfileMenu
});
export const ModusWcProgress = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-progress',
    elementClass: ModusWcProgressElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcProgress
});
export const ModusWcRadio = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-radio',
    elementClass: ModusWcRadioElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcRadio
});
export const ModusWcRating = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-rating',
    elementClass: ModusWcRatingElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onRatingChange: 'ratingChange' },
    defineCustomElement: defineModusWcRating
});
export const ModusWcSelect = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-select',
    elementClass: ModusWcSelectElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcSelect
});
export const ModusWcSideNavigation = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-side-navigation',
    elementClass: ModusWcSideNavigationElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onExpandedChange: 'expandedChange' },
    defineCustomElement: defineModusWcSideNavigation
});
export const ModusWcSkeleton = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-skeleton',
    elementClass: ModusWcSkeletonElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcSkeleton
});
export const ModusWcSlider = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-slider',
    elementClass: ModusWcSliderElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcSlider
});
export const ModusWcStepper = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-stepper',
    elementClass: ModusWcStepperElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcStepper
});
export const ModusWcSwitch = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-switch',
    elementClass: ModusWcSwitchElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcSwitch
});
export const ModusWcTable = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-table',
    elementClass: ModusWcTableElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onCellEditStart: 'cellEditStart',
        onCellEditCommit: 'cellEditCommit',
        onRowClick: 'rowClick',
        onSortChange: 'sortChange',
        onPaginationChange: 'paginationChange',
        onRowSelectionChange: 'rowSelectionChange'
    },
    defineCustomElement: defineModusWcTable
});
export const ModusWcTabs = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-tabs',
    elementClass: ModusWcTabsElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onTabChange: 'tabChange' },
    defineCustomElement: defineModusWcTabs
});
export const ModusWcTextInput = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-text-input',
    elementClass: ModusWcTextInputElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onClearClick: 'clearClick',
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcTextInput
});
export const ModusWcTextarea = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-textarea',
    elementClass: ModusWcTextareaElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcTextarea
});
export const ModusWcThemeProvider = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-theme-provider',
    elementClass: ModusWcThemeProviderElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcThemeProvider
});
export const ModusWcThemeSwitcher = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-theme-switcher',
    elementClass: ModusWcThemeSwitcherElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onThemeChange: 'themeChange' },
    defineCustomElement: defineModusWcThemeSwitcher
});
export const ModusWcTimeInput = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-time-input',
    elementClass: ModusWcTimeInputElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onInputBlur: 'inputBlur',
        onInputChange: 'inputChange',
        onInputFocus: 'inputFocus'
    },
    defineCustomElement: defineModusWcTimeInput
});
export const ModusWcToast = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-toast',
    elementClass: ModusWcToastElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcToast
});
export const ModusWcToolbar = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-toolbar',
    elementClass: ModusWcToolbarElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcToolbar
});
export const ModusWcTooltip = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-tooltip',
    elementClass: ModusWcTooltipElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: { onDismissEscape: 'dismissEscape' },
    defineCustomElement: defineModusWcTooltip
});
export const ModusWcTypography = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-typography',
    elementClass: ModusWcTypographyElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {},
    defineCustomElement: defineModusWcTypography
});
export const ModusWcUtilityPanel = /*@__PURE__*/ createComponent({
    tagName: 'modus-wc-utility-panel',
    elementClass: ModusWcUtilityPanelElement,
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {
        onPanelOpened: 'panelOpened',
        onPanelClosed: 'panelClosed'
    },
    defineCustomElement: defineModusWcUtilityPanel
});
