// @trimble-oss/moduswebcomponents/collection/components/modus-wc-breadcrumbs/modus-wc-breadcrumbs.js
import { h, Host, } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-breadcrumbs.tailwind";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable breadcrumbs component used to help users navigate through a website.
 */
export class ModusWcBreadcrumbs {
    constructor() {
        this.inheritedAttributes = {};
        /** The breadcrumbs to render. */
        this.items = [];
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The size of the breadcrumbs. */
        this.size = 'md';
    }
    componentWillLoad() {
        if (!this.el.ariaLabel) {
            this.el.ariaLabel = 'Breadcrumbs';
        }
        if (!this.items || this.items.length === 0) {
            console.error('ModusWcBreadcrumbs: breadcrumb items data is required.');
        }
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-breadcrumbs'];
        const propClasses = convertPropsToClasses({
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    handleClick(event, crumb) {
        if (!crumb.url) {
            event.preventDefault();
        }
        this.breadcrumbClick.emit(crumb);
    }
    render() {
        return (h(Host, { key: '9962a34f1dac82f1f3e7556d406c9d4863c014ad' }, h("nav", Object.assign({ key: '67435e57f382ff45f8698202aa12eb65e6d6994c', class: this.getClasses() }, this.inheritedAttributes), h("ol", { key: 'aeeb9c2fe11832c1b307821da45851e705b0044a' }, this.items.map((item, index) => {
            const isCurrentPage = index === this.items.length - 1;
            return (h("li", Object.assign({ key: item.label }, (isCurrentPage ? { 'aria-current': 'page' } : {})), isCurrentPage ? (h("span", null, item.label)) : (h("a", { href: item.url, onClick: (event) => this.handleClick(event, item) }, item.label))));
        })))));
    }
    static get is() { return "modus-wc-breadcrumbs"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-breadcrumbs.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-breadcrumbs.css"]
        };
    }
    static get properties() {
        return {
            "items": {
                "type": "unknown",
                "attribute": "items",
                "mutable": false,
                "complexType": {
                    "original": "IBreadcrumb[]",
                    "resolved": "IBreadcrumb[]",
                    "references": {
                        "IBreadcrumb": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-breadcrumbs/modus-wc-breadcrumbs.tsx",
                            "id": "src/components/modus-wc-breadcrumbs/modus-wc-breadcrumbs.tsx::IBreadcrumb"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The breadcrumbs to render."
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the inner div."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "ModusSize",
                    "resolved": "\"lg\" | \"md\" | \"sm\" | undefined",
                    "references": {
                        "ModusSize": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::ModusSize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The size of the breadcrumbs."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'md'"
            }
        };
    }
    static get events() {
        return [{
                "method": "breadcrumbClick",
                "name": "breadcrumbClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when a breadcrumb is clicked."
                },
                "complexType": {
                    "original": "IBreadcrumb",
                    "resolved": "IBreadcrumb",
                    "references": {
                        "IBreadcrumb": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-breadcrumbs/modus-wc-breadcrumbs.tsx",
                            "id": "src/components/modus-wc-breadcrumbs/modus-wc-breadcrumbs.tsx::IBreadcrumb"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "el"; }
}
