// @trimble-oss/moduswebcomponents/collection/components/modus-wc-toolbar/modus-wc-toolbar.js
import { h, Host } from "@stencil/core";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable toolbar component used to organize content across the entire page.
 *
 * This component provides 'start', 'center', and 'end' `<slot>` elements for inserting custom HTML.
 */
export class ModusWcToolbar {
    constructor() {
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the outer div. */
        this.customClass = '';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-navbar'];
        // The order CSS classes are added matters to CSS specificity
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: '252cc26c4653ec039229cfecb7ef9049f0875b95' }, h("div", Object.assign({ key: '55664aca492ad12f8e099bc0c6a6046bd81fa545', class: this.getClasses() }, this.inheritedAttributes), h("div", { key: '2031f716b318fa5b04cb3661a2adf707644ac072', class: "modus-wc-navbar-start" }, h("slot", { key: 'a23a967e42d463f506314e8d4f80f7aad41b9de2', name: "start" })), h("div", { key: 'e796353849827c48f907d7071cd68bd8b0f7ca4c', class: "modus-wc-navbar-center" }, h("slot", { key: '3a05fd3d4cd1b45de84585f9334f9b75ef903c8f', name: "center" })), h("div", { key: '7994ec1b7f45e0b6c5c26bfa3458e24c969a2fff', class: "modus-wc-navbar-end" }, h("slot", { key: '4040c3637315c02313247be884e9719efb6c9e49', name: "end" })))));
    }
    static get is() { return "modus-wc-toolbar"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-toolbar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-toolbar.css"]
        };
    }
    static get properties() {
        return {
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the outer div."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            }
        };
    }
    static get elementRef() { return "el"; }
}
