// @trimble-oss/moduswebcomponents/collection/components/modus-wc-alert/modus-wc-alert.js
import { h, Host, } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-alert.tailwind";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable alert component used to inform the user about important events.
 *
 * The component supports `<slot>` elements for injecting custom content and buttons.
 */
export class ModusWcAlert {
    constructor() {
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the outer div element. */
        this.customClass = '';
        /** Whether the alert has a dismiss button */
        this.dismissible = false;
        /** The variant of the alert. */
        this.variant = 'info';
    }
    componentWillLoad() {
        // Set default role if none provided
        if (!this.el.hasAttribute('role')) {
            this.el.setAttribute('role', 'status');
        }
        // Then inherit all ARIA attributes normally
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-alert'];
        const propClasses = convertPropsToClasses({
            variant: this.variant,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getLeadingIcon() {
        if (this.icon) {
            return (h("modus-wc-icon", { "custom-class": "modus-wc-alert-icon", name: this.icon, variant: "outlined" }));
        }
        switch (this.variant) {
            case 'error':
                return (h("modus-wc-icon", { "custom-class": "modus-wc-alert-icon", name: "alert", variant: "outlined" }));
            case 'success':
                return (h("modus-wc-icon", { "custom-class": "modus-wc-alert-icon", name: "check_circle" }));
            case 'warning':
                return (h("modus-wc-icon", { "custom-class": "modus-wc-alert-icon", name: "warning", variant: "outlined" }));
            case 'info':
            default:
                return (h("modus-wc-icon", { "custom-class": "modus-wc-alert-icon", name: "info", variant: "outlined" }));
        }
    }
    delayChanged(newDelay) {
        clearTimeout(this.timerId);
        this.timerId = setTimeout(() => {
            this.dismissElement();
        }, newDelay);
    }
    dismissElement() {
        this.dismissClick.emit();
        this.el.remove();
    }
    componentDidLoad() {
        if (this.delay && this.delay > 0) {
            this.timerId = setTimeout(() => {
                this.dismissElement();
            }, this.delay);
        }
    }
    disconnectedCallback() {
        clearTimeout(this.timerId);
    }
    elementKeyupHandler(event) {
        switch (event.code) {
            case 'Escape':
                if (!this.dismissible) {
                    return;
                }
                this.dismissElement();
                break;
        }
    }
    render() {
        return (h(Host, { key: '6b8c17a8533ea7c7d9d24e5f90f0f3a1ce3e5bfa' }, h("div", Object.assign({ key: '9288fdc6818570cc5b090ac5ad8fe2e2653594cc', class: this.getClasses() }, this.inheritedAttributes), this.getLeadingIcon(), h("div", { key: 'b11bc362f74a0db454e84e3977555dc3d7c706e7' }, h("div", { key: '585b8356fa235d39304b98ad58a1ba5ae5ff5c70', class: "title" }, this.alertTitle), this.alertDescription && (h("div", { key: 'b5ce7525f6789a06b93733628831b3811fae4069', class: "description" }, this.alertDescription)), !this.alertTitle && !this.alertDescription && (h("slot", { key: 'b6600d961206555aeb0c43225a05b12331a37635', name: "content" }))), h("slot", { key: 'f30e118b87a88405c4ac84bbc40c1ccbd315c1b9', name: "button" }), this.dismissible && (h("modus-wc-button", { key: 'dcfd3b06239d15f3c4d2c035bafb2bb68f21a5bf', "aria-label": "notification button", color: "tertiary", size: "sm", slot: "button", variant: "borderless", onButtonClick: () => this.dismissElement() }, h("modus-wc-icon", { key: 'a2ce3c13ddd0531af2b80bc4224e1c5dfdd4dffd', "custom-class": "modus-wc-alert-close-icon", name: "close", variant: "outlined" }))))));
    }
    static get is() { return "modus-wc-alert"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-alert.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-alert.css"]
        };
    }
    static get properties() {
        return {
            "alertDescription": {
                "type": "string",
                "attribute": "alert-description",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The description of the alert."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "alertTitle": {
                "type": "string",
                "attribute": "alert-title",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The title of the alert."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the outer div element."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "delay": {
                "type": "number",
                "attribute": "delay",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Time taken to dismiss the alert in milliseconds"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "dismissible": {
                "type": "boolean",
                "attribute": "dismissible",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Whether the alert has a dismiss button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The Modus icon to render."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "'error' | 'info' | 'success' | 'warning'",
                    "resolved": "\"error\" | \"info\" | \"success\" | \"warning\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The variant of the alert."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'info'"
            }
        };
    }
    static get events() {
        return [{
                "method": "dismissClick",
                "name": "dismissClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires when the alert is dismissed"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "delay",
                "methodName": "delayChanged"
            }];
    }
    static get listeners() {
        return [{
                "name": "keyup",
                "method": "elementKeyupHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
