// @trimble-oss/moduswebcomponents/collection/components/modus-wc-skeleton/modus-wc-skeleton.js
import { h, Host } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-skeleton.tailwind";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable skeleton component used to create skeletons of various sizes and shapes
 */
export class ModusWcSkeleton {
    constructor() {
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The height of the skeleton. */
        this.height = 'var(--modus-wc-line-height-md)';
        /** The shape of the skeleton. */
        this.shape = 'rectangle';
        /** The width of the skeleton. */
        this.width = '100%';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = ['modus-wc-skeleton'];
        const propClasses = convertPropsToClasses({
            shape: this.shape,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    getStyles() {
        return { height: this.height, width: this.width };
    }
    render() {
        return (h(Host, { key: '9d67bdc774912492b42d055e601cc3143fe56fd8' }, h("div", Object.assign({ key: 'd30972478d13d8d5c604077f12c6204345fe8b0c', "aria-hidden": "true", class: this.getClasses(), style: this.getStyles(), tabindex: -1 }, this.inheritedAttributes))));
    }
    static get is() { return "modus-wc-skeleton"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-skeleton.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-skeleton.css"]
        };
    }
    static get properties() {
        return {
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the inner div."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "height": {
                "type": "string",
                "attribute": "height",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The height of the skeleton."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'var(--modus-wc-line-height-md)'"
            },
            "shape": {
                "type": "string",
                "attribute": "shape",
                "mutable": false,
                "complexType": {
                    "original": "'circle' | 'rectangle'",
                    "resolved": "\"circle\" | \"rectangle\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The shape of the skeleton."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'rectangle'"
            },
            "width": {
                "type": "string",
                "attribute": "width",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The width of the skeleton."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'100%'"
            }
        };
    }
    static get elementRef() { return "el"; }
}
