// @trimble-oss/moduswebcomponents/collection/components/modus-wc-input-feedback/modus-wc-input-feedback.stories.js
import { html } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
import { createShadowHostClass } from "../../providers/shadow-dom/shadow-host-helper";
const meta = {
    title: 'Components/Forms/Input Feedback',
    component: 'modus-wc-input-feedback',
    args: {
        level: 'error',
        message: 'Uh oh. You done messed up.',
        size: 'md',
    },
    argTypes: {
        level: {
            control: { type: 'select' },
            options: ['error', 'info', 'success', 'warning'],
        },
        size: {
            control: { type: 'select' },
            options: ['sm', 'md', 'lg'],
        },
    },
};
export default meta;
const Template = {
    render: (args) => {
        // prettier-ignore
        return html `
<modus-wc-input-feedback
  custom-class=${ifDefined(args['custom-class'])}
  icon=${ifDefined(args.icon)}
  level=${args.level}
  message=${ifDefined(args.message)}
  size=${ifDefined(args.size)}
>
</modus-wc-input-feedback>
    `;
    },
};
export const Default = Object.assign({}, Template);
export const WithCustomModusIcon = {
    render: (args) => {
        // prettier-ignore
        return html `
<modus-wc-input-feedback
  custom-class=${ifDefined(args['custom-class'])}
  icon='calendar_check'
  level='success'
  message='Event added to calendar!'
  size=${ifDefined(args.size)}
>
</modus-wc-input-feedback>
    `;
    },
};
export const ShadowDomParent = {
    render: (args) => {
        // Create a unique shadow host for input-feedback component
        if (!customElements.get('input-feedback-shadow-host')) {
            const InputFeedbackShadowHost = createShadowHostClass({
                componentTag: 'modus-wc-input-feedback',
                propsMapper: (v, el) => {
                    var _a, _b, _c;
                    const inputFeedbackEl = el;
                    inputFeedbackEl.customClass = v['custom-class'] || '';
                    inputFeedbackEl.icon = (_a = v.icon) !== null && _a !== void 0 ? _a : '';
                    inputFeedbackEl.level = v.level;
                    inputFeedbackEl.message = (_b = v.message) !== null && _b !== void 0 ? _b : '';
                    inputFeedbackEl.size = (_c = v.size) !== null && _c !== void 0 ? _c : 'md';
                },
            });
            customElements.define('input-feedback-shadow-host', InputFeedbackShadowHost);
        }
        return html `<input-feedback-shadow-host
      .props=${Object.assign({}, args)}
    ></input-feedback-shadow-host>`;
    },
};
