// @trimble-oss/moduswebcomponents/collection/components/modus-wc-utility-panel/modus-wc-utility-panel.js
import { Fragment, h, } from "@stencil/core";
import { handleShadowDOMStyles } from "../base-component";
export class ModusWcUtilityPanel {
    constructor() {
        /** The panel is expanded or closed */
        this.expanded = false;
        /** Determines if the panel pushes content or displays an overlay. */
        this.pushContent = false;
        this.isInitialLoad = true;
        this.handlePanelClose = () => {
            void this.closePanel();
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
    }
    componentDidLoad() {
        // Only adjust content if panel is already expanded on load and we have a target
        if (this.pushContent && this.expanded && this.targetElement) {
            this.adjustContent();
        }
        // Mark that initial load is complete after adjusting content
        this.isInitialLoad = false;
    }
    handleExpandedChange(newValue) {
        // Skip the watcher on initial load
        if (this.isInitialLoad) {
            return;
        }
        if (newValue) {
            void this.openPanel();
        }
        else {
            void this.closePanel();
        }
    }
    handleTargetChange() {
        // Re-adjust content when target changes
        if (this.expanded && this.pushContent && this.targetElement) {
            this.adjustContent();
        }
    }
    openPanel() {
        this.panelOpened.emit();
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    closePanel() {
        this.panelClosed.emit();
        if (this.pushContent) {
            this.adjustContent();
        }
    }
    adjustContent() {
        if (!this.pushContent || !this.targetElement)
            return;
        // Add base class for transitions
        this.targetElement.classList.add('modus-wc-utility-panel-push-target');
        // Toggle pushed class based on expanded state
        if (this.expanded) {
            this.targetElement.classList.add('modus-wc-utility-panel-pushed');
        }
        else {
            this.targetElement.classList.remove('modus-wc-utility-panel-pushed');
        }
    }
    hasSlotContent(slotName) {
        const slot = this.el.querySelector(`[slot="${slotName}"]`);
        return !!slot;
    }
    render() {
        const hasHeader = this.hasSlotContent('header');
        const hasFooter = this.hasSlotContent('footer');
        return (h("div", { key: 'c3e69d7c25cc68d477b7dff36f2429a7a90deffc', class: {
                'modus-wc-utility-panel': true,
                open: this.expanded,
            } }, h("div", { key: 'a80dc0d0155b1755114d8079b9bea8e47f59f4ce', class: "modus-wc-utility-panel-content" }, hasHeader && (h(Fragment, { key: '6037c5cd032f6b160e6a956e147445efd93061d4' }, h("div", { key: '02d222e4aac4c8ab9fd4a5c601649dcd4a870eaf', class: "modus-wc-utility-panel-header" }, h("slot", { key: '1ee7c97ef7677fd2b546dbf902a800ccf70a2d68', name: "header" })), h("hr", { key: '8f9b6e4e96bd8e9afae2f447f0c83f744c5ca175' }))), h("div", { key: 'a6dc213a2b651bfcf58e50f71f1683b45f8c3448', class: "modus-wc-utility-panel-body" }, h("slot", { key: '74a4bdc447ca4d07b9bfbe955776072570f8fd68', name: "body" })), hasFooter && (h(Fragment, { key: '00ed976af7d33d273c47199d9864fcfcafa41a65' }, h("hr", { key: 'f66ee98a4b53e033dd5303f9347547f54b44d1c0' }), h("div", { key: 'f39464a2c72dfa50fb49508d303419c7f08b6ca4', class: "modus-wc-utility-panel-footer" }, h("slot", { key: '75d25d058b3abbcc7e6153af18f1bfea8d2d7c60', name: "footer" })))))));
    }
    static get is() { return "modus-wc-utility-panel"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-utility-panel.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-utility-panel.css"]
        };
    }
    static get properties() {
        return {
            "expanded": {
                "type": "boolean",
                "attribute": "expanded",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The panel is expanded or closed"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "pushContent": {
                "type": "boolean",
                "attribute": "push-content",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines if the panel pushes content or displays an overlay."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "targetElement": {
                "type": "unknown",
                "attribute": "target-element",
                "mutable": false,
                "complexType": {
                    "original": "HTMLElement",
                    "resolved": "HTMLElement | undefined",
                    "references": {
                        "HTMLElement": {
                            "location": "global",
                            "id": "global::HTMLElement"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Target element reference to push content when panel opens"
                },
                "getter": false,
                "setter": false
            }
        };
    }
    static get events() {
        return [{
                "method": "panelOpened",
                "name": "panelOpened",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires when the panel is opened."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "panelClosed",
                "name": "panelClosed",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires when the panel is closed."
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get watchers() {
        return [{
                "propName": "expanded",
                "methodName": "handleExpandedChange"
            }, {
                "propName": "targetElement",
                "methodName": "handleTargetChange"
            }];
    }
}
