// @trimble-oss/moduswebcomponents/collection/components/modus-wc-input-label/modus-wc-input-label.js
import { h, Host } from "@stencil/core";
import { handleShadowDOMStyles } from "../base-component";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable input label component.
 *
 * The component supports a `<slot>` for injecting additional custom content inside the label, such as icons or formatted text.
 */
export class ModusWcInputLabel {
    constructor() {
        this.inheritedAttributes = {};
        /** Additional classes for custom styling. */
        this.customClass = '';
        /** Whether the label indicates a required field. */
        this.required = false;
        /** The size of the label. */
        this.size = 'md';
        this.getClasses = () => {
            const classList = ['modus-wc-input-label'];
            // The order CSS classes are added matters to CSS specificity
            if (this.size)
                classList.push(`modus-wc-input-label-size-${this.size}`);
            if (this.customClass)
                classList.push(this.customClass);
            return classList.join(' ');
        };
    }
    componentWillLoad() {
        // Auto-inject CSS if component is used inside user's shadow DOM
        handleShadowDOMStyles(this.el);
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    render() {
        return (h(Host, { key: 'a1b1faf45264587fbf52653c0b3c8d1c3371f13d', class: "modus-wc-input-label-host" }, h("label", Object.assign({ key: '0ee1830f85c03ee70082e67b2c73224f84b328c3', class: this.getClasses(), htmlFor: this.forId }, this.inheritedAttributes), this.labelText, this.required && (h("span", { key: '11b6749526377e5900b5a60df1faba67e86e8337', "aria-hidden": "true", class: "required-indicator" }, '\u00A0*')), this.subLabelText && (h("div", { key: '8f403f2fc9f01733b6ac263471875146ff8fe199', class: "modus-wc-input-label-sublabel" }, this.subLabelText)), h("slot", { key: '7c0c7ef4d598a29383c21c1378522c588fcd88eb' }))));
    }
    static get is() { return "modus-wc-input-label"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-input-label.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-input-label.css"]
        };
    }
    static get properties() {
        return {
            "forId": {
                "type": "string",
                "attribute": "for-id",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The `for` attribute of the label, matching the `id` of the associated input."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Additional classes for custom styling."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "labelText": {
                "type": "string",
                "attribute": "label-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The text to display within the label."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "required": {
                "type": "boolean",
                "attribute": "required",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Whether the label indicates a required field."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "ModusSize",
                    "resolved": "\"lg\" | \"md\" | \"sm\" | undefined",
                    "references": {
                        "ModusSize": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::ModusSize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The size of the label."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'md'"
            },
            "subLabelText": {
                "type": "string",
                "attribute": "sub-label-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The text rendered beneath the label."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get elementRef() { return "el"; }
}
