// @trimble-oss/moduswebcomponents/collection/components/modus-wc-profile-menu/modus-wc-profile-menu.js
import { h, Host, } from "@stencil/core";
import { inheritAriaAttributes } from "../utils";
import { renderSubMenu } from "./utils/menu_template";
export class ModusWcProfileMenu {
    constructor() {
        this.inheritedAttributes = {};
        this.handleMenuItemClick = (value) => {
            this.menuItemClick.emit(value);
        };
        this.currentYear = new Date().getFullYear();
        this.mainMenu = {
            items: [
                {
                    label: 'My Profile',
                    icon: 'info',
                    iconVariant: 'solid',
                    iconSize: 'sm',
                    value: 'my-profile',
                },
                {
                    label: 'My Products',
                    icon: 'home',
                    iconVariant: 'solid',
                    iconSize: 'sm',
                    value: 'my-products',
                },
                {
                    label: 'Support center',
                    icon: 'bar_graph',
                    iconVariant: 'solid',
                    iconSize: 'sm',
                    value: 'support-center',
                },
                {
                    label: 'Admin settings',
                    icon: 'download',
                    iconVariant: 'solid',
                    iconSize: 'sm',
                    value: 'admin-settings',
                },
            ],
        };
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    handleItemSelect(event) {
        event.stopPropagation();
    }
    render() {
        var _a;
        return (h(Host, Object.assign({ key: 'af143317f9f5ef31c523780e8328f5e7b68bcf2f' }, this.inheritedAttributes), h("modus-wc-panel", { key: '84dc810032dab6ee6bbc20a3cd6bd4c712da1ba6', height: "auto", width: "298px" }, h("div", { key: '41677a2ee358e08c0d1af4804f09dedeb3f285d6', class: "profile-menu-header", slot: "header" }, h("div", { key: 'a2106673fefb5804dd49ba8231e36230acca9ce5', class: "header-content" }, h("modus-wc-avatar", { key: 'e308ccfdcf244b0c409681ddb5055c48605928f9', "img-src": this.profileProps.profileImageUrl, alt: this.profileProps.userName || 'Profile', size: "sm" }), h("div", { key: 'af7da2713827284f8a52838dc444adef4bd85bfa', class: "text-container" }, h("modus-wc-typography", { key: 'a07a3cdb5592e228d2b9d77f8e24ba3055299115', customClass: "header-text", weight: "semibold", size: "xs", label: this.profileProps.headerName }), h("modus-wc-typography", { key: '41dd8d7b16e62a3f2bc0f616d5da2110d592de61', customClass: "user-name-text", weight: "semibold", size: "md", label: this.profileProps.userName }), h("modus-wc-typography", { key: 'e3e1251413a0b7a0834693cf96325b0bd4fced9b', customClass: "email-text", weight: "semibold", size: "xs", label: this.profileProps.userEmail }), this.profileProps.manageTrimbleId ? (h("a", { href: this.profileProps.manageTrimbleId.link, target: this.profileProps.manageTrimbleId.target, rel: (_a = this.profileProps.manageTrimbleId.rel) !== null && _a !== void 0 ? _a : (this.profileProps.manageTrimbleId.target === '_blank'
                ? 'noopener noreferrer'
                : undefined) }, h("modus-wc-typography", { customClass: "manage-link", hierarchy: "p", size: "sm", weight: "normal", label: "Manage my Trimble ID" }))) : (h("modus-wc-typography", { customClass: "manage-link", hierarchy: "p", size: "sm", weight: "normal", label: "Manage my Trimble ID" }))))), h("div", { key: '6bf80bff2b5370217cbd62a1546d69863267c8e4', slot: "body" }, renderSubMenu(this.mainMenu, this.handleMenuItemClick, true), this.menuOne &&
            renderSubMenu(this.menuOne, this.handleMenuItemClick), this.menuTwo &&
            renderSubMenu(this.menuTwo, this.handleMenuItemClick), h("modus-wc-menu", { key: 'df49d075723c4ceb7e64115b0af811b41d22702c' }, h("modus-wc-menu-item", { key: '2f9b9762f3b6af944f47aff574d7e63ef099e485', label: "Sign Out", size: "md", onItemSelect: () => this.signOutClick.emit() }, h("modus-wc-icon", { key: 'eb91a39e195961f1ef1a9b9ea0b0bd850ed4ee25', slot: "start-icon", name: "sign_out", size: "sm", variant: "solid" })))), h("div", { key: 'b1ef07cfaa14083d26d8fe3215236e9cfd5b44c7', slot: "footer", class: "profile-menu-footer" }, h("modus-wc-typography", { key: '584d92842c6e251e32164ac13d8949cd33a347ea', customClass: "footer-text", hierarchy: "p", size: "sm", weight: "normal", label: `©${this.currentYear}, Trimble Inc.` })))));
    }
    static get is() { return "modus-wc-profile-menu"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-profile-menu.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-profile-menu.css"]
        };
    }
    static get properties() {
        return {
            "profileProps": {
                "type": "unknown",
                "attribute": "profile-props",
                "mutable": false,
                "complexType": {
                    "original": "IProfileMenuProps",
                    "resolved": "IProfileMenuProps",
                    "references": {
                        "IProfileMenuProps": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-profile-menu/modus-wc-profile-menu.tsx",
                            "id": "src/components/modus-wc-profile-menu/modus-wc-profile-menu.tsx::IProfileMenuProps"
                        }
                    }
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Profile menu properties containing user information"
                },
                "getter": false,
                "setter": false
            },
            "menuOne": {
                "type": "unknown",
                "attribute": "menu-one",
                "mutable": false,
                "complexType": {
                    "original": "ISubMenu",
                    "resolved": "ISubMenu | undefined",
                    "references": {
                        "ISubMenu": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-profile-menu/modus-wc-profile-menu.tsx",
                            "id": "src/components/modus-wc-profile-menu/modus-wc-profile-menu.tsx::ISubMenu"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Configuration for the first menu including title and items"
                },
                "getter": false,
                "setter": false
            },
            "menuTwo": {
                "type": "unknown",
                "attribute": "menu-two",
                "mutable": false,
                "complexType": {
                    "original": "ISubMenu",
                    "resolved": "ISubMenu | undefined",
                    "references": {
                        "ISubMenu": {
                            "location": "local",
                            "path": "/home/runner/work/modus-wc-2.0/modus-wc-2.0/src/components/modus-wc-profile-menu/modus-wc-profile-menu.tsx",
                            "id": "src/components/modus-wc-profile-menu/modus-wc-profile-menu.tsx::ISubMenu"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Configuration for the second menu including title and items"
                },
                "getter": false,
                "setter": false
            }
        };
    }
    static get events() {
        return [{
                "method": "signOutClick",
                "name": "signOutClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the Sign Out menu item is clicked"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "menuItemClick",
                "name": "menuItemClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when any menu item is clicked, passing back the item value or label"
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
    static get listeners() {
        return [{
                "name": "itemSelect",
                "method": "handleItemSelect",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
