// @trimble-oss/moduswebcomponents/collection/components/modus-wc-avatar/modus-wc-avatar.stories.js
import { html } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
const meta = {
    title: 'Components/Avatar',
    component: 'modus-wc-avatar',
    args: {
        alt: 'Example avatar',
        'img-src': 'https://i.pinimg.com/474x/73/54/79/7354794bf3873c3ef2666f778da4bcac.jpg',
        shape: 'circle',
        initials: '',
        size: 'md',
    },
    argTypes: {
        shape: {
            control: { type: 'select' },
            options: ['circle', 'square'],
        },
        size: {
            control: { type: 'select' },
            options: ['xs', 'sm', 'md', 'lg'],
        },
    },
};
export default meta;
const Template = {
    render: (args) => {
        return html `
      <modus-wc-avatar
        alt="${args.alt}"
        custom-class="${ifDefined(args['custom-class'])}"
        img-src="${args['img-src']}"
        initials="${args.initials}"
        shape="${args.shape}"
        size="${args.size}"
      ></modus-wc-avatar>
    `;
    },
};
export const Default = Object.assign({}, Template);
