// @trimble-oss/moduswebcomponents/collection/components/modus-wc-avatar/modus-wc-avatar.js
import { h, Host } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-avatar.tailwind";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable avatar component used to create avatars with different images or user initials.
 * When no image is provided, the component can display initials (up to 3 characters) from the initials prop.
 * The component will extract the first letter of each word in the initials string.
 */
export class ModusWcAvatar {
    constructor() {
        this.inheritedAttributes = {};
        /** Custom CSS class to apply to the inner div. */
        this.customClass = '';
        /** The location of the image. */
        this.imgSrc = '';
        /** The initials to display when no image is provided. */
        this.initials = '';
        // TODO - add placeholder support (need UX logic)
        /** The shape of the avatar. */
        this.shape = 'circle';
        /** The size of the avatar. */
        this.size = 'md';
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    getClasses() {
        const classList = [];
        const propClasses = convertPropsToClasses({
            shape: this.shape,
            size: this.size,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        if (!this.imgSrc && !this.initials)
            classList.push('no-image');
        return classList.join(' ');
    }
    getUserInitials() {
        if (!this.initials)
            return '';
        return this.initials
            .trim()
            .split(/\s+/)
            .filter(Boolean)
            .map((part) => part.charAt(0))
            .join('')
            .substring(0, 3)
            .toUpperCase();
    }
    render() {
        const altText = this.alt || 'User avatar';
        return (h(Host, { key: '8ac52db552b5093476ae19f158819a3c9d253ab3' }, h("div", Object.assign({ key: 'e3452c9c37a8adad4aaae454112f4020b7bf4f83', class: "modus-wc-avatar" }, this.inheritedAttributes), h("div", { key: 'eba4c00724a6f4dff7636d4a43fa46e93a28f75f', class: this.getClasses() }, this.imgSrc ? (h("img", { src: this.imgSrc, alt: altText })) : this.initials ? (h("span", { class: "initials", "aria-label": this.alt || this.initials }, this.getUserInitials())) : (h("modus-wc-icon", { "aria-label": altText, name: "person", size: this.size, variant: "solid" }))))));
    }
    static get is() { return "modus-wc-avatar"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-avatar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-avatar.css"]
        };
    }
    static get properties() {
        return {
            "alt": {
                "type": "string",
                "attribute": "alt",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The image alt attribute for accessibility."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply to the inner div."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "imgSrc": {
                "type": "string",
                "attribute": "img-src",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The location of the image."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "initials": {
                "type": "string",
                "attribute": "initials",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The initials to display when no image is provided."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "shape": {
                "type": "string",
                "attribute": "shape",
                "mutable": false,
                "complexType": {
                    "original": "'circle' | 'square'",
                    "resolved": "\"circle\" | \"square\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The shape of the avatar."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'circle'"
            },
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "DaisySize",
                    "resolved": "\"lg\" | \"md\" | \"sm\" | \"xs\" | undefined",
                    "references": {
                        "DaisySize": {
                            "location": "import",
                            "path": "../types",
                            "id": "src/components/types.ts::DaisySize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The size of the avatar."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'md'"
            }
        };
    }
    static get elementRef() { return "el"; }
}
