// @trimble-oss/moduswebcomponents/collection/components/modus-wc-modal/modus-wc-modal.js
import { h, Host } from "@stencil/core";
import { convertPropsToClasses } from "./modus-wc-modal.tailwind";
import { CloseSolidIcon } from "../../icons/close-solid.icon";
import { CollapseSolidIcon } from "../../icons/collapse-solid.icon";
import { ExpandSolidIcon } from "../../icons/expand-solid.icon";
import { inheritAriaAttributes } from "../utils";
/**
 * A customizable modal component used to display content in a dialog.
 *
 * This component supports 'header', 'content', and 'footer' `<slot>` elements for inserting custom HTML.
 */
export class ModusWcModal {
    constructor() {
        this.inheritedAttributes = {};
        /**
         * The modal's backdrop.
         * Specify 'static' for a backdrop that doesn't close the modal when clicked outside the modal content.
         */
        this.backdrop = 'default';
        /** Custom CSS class to apply */
        this.customClass = '';
        /** Specifies whether the modal should be displayed full-screen */
        this.fullscreen = false;
        /** Specifies the position of the modal */
        this.position = 'center';
        /** Specifies whether to show the close icon button at the top right of modal */
        this.showClose = true;
        /** Specifies whether to show the fullscreen toggle icon button */
        this.showFullscreenToggle = false;
    }
    componentWillLoad() {
        if (!this.modalId) {
            console.error("The modal component requires a unique 'modalId' to be passed in as a prop.");
        }
        this.inheritedAttributes = inheritAriaAttributes(this.el);
    }
    closeDialog() {
        const dialog = this.el.querySelector('dialog');
        if (dialog)
            dialog.close();
    }
    getClasses() {
        const classList = ['modus-wc-modal'];
        const propClasses = convertPropsToClasses({
            position: this.position,
        });
        // The order CSS classes are added matters to CSS specificity
        if (propClasses)
            classList.push(propClasses);
        if (this.customClass)
            classList.push(this.customClass);
        return classList.join(' ');
    }
    render() {
        return (h(Host, { key: 'c1858299cd5b0ee89afca4f4451f4e00c10aafea' }, h("dialog", Object.assign({ key: 'd9dd15b045c723065fe4fd1e58f45aece9e185d9', class: this.getClasses(), id: this.modalId }, this.inheritedAttributes), h("div", { key: 'a0f53a90214a8abdcf2808e9436b19da7b607c4b', class: `modus-wc-modal-box ${this.fullscreen ? 'modus-wc-modal-fullscreen' : ''}` }, h("div", { key: '0f3c81b693fb86125587b54222ff993d29ffd2e0', class: "modus-wc-modal-header modus-wc-text-lg modus-wc-font-bold" }, h("slot", { key: 'dfcc4b7ac87c7e1ffb62e765a2d7c843b517e5e9', name: "header" }), h("div", { key: '4f12e4cb4e1f6633bffdd0e2b9bfea6379a66903', class: "modus-wc-modal-top-icon-buttons" }, this.showFullscreenToggle && (h("modus-wc-button", { key: '15487e73bcca465cf9b9e2c498f1020029eb5779', "aria-label": "Fullscreen toggle", onButtonClick: () => (this.fullscreen = !this.fullscreen), shape: "square", size: "sm", variant: "borderless" }, this.fullscreen ? (h(CollapseSolidIcon, null)) : (h(ExpandSolidIcon, null)))), this.showClose && (h("modus-wc-button", { key: '6768abb022949b2ded539cd6d04d637920358a94', "aria-label": "Close modal", onButtonClick: () => this.closeDialog(), shape: "square", size: "sm", variant: "borderless" }, h(CloseSolidIcon, { key: '3b9830db30966927bd4173c440d01e6a939acf57' }))))), h("div", { key: 'c595958bd6a5c63ac6429dc45be0215d1d6342d5', class: "modus-wc-modal-content modus-wc-py-4" }, h("slot", { key: '2efe016bd2e5cf3e50c14ef89c40580180aa7583', name: "content" })), h("div", { key: '81db4a915fb9250e9aaacbf7d52d046ba1a096b8', class: "modus-wc-modal-action" }, h("slot", { key: '2a1d3bb847afc3bda4b662a231013811b5fd8d2f', name: "footer" }))), this.backdrop === 'default' && (h("form", { key: 'ff7c4cbdac9246335d5cf446cf3c8576a103d9da', method: "dialog", class: "modus-wc-modal-backdrop" }, h("button", { key: '379cff51417e4a738a44d9c659b41831c3730462' }, "close"))))));
    }
    static get is() { return "modus-wc-modal"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-wc-modal.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-wc-modal.css"]
        };
    }
    static get properties() {
        return {
            "backdrop": {
                "type": "string",
                "attribute": "backdrop",
                "mutable": false,
                "complexType": {
                    "original": "'default' | 'static'",
                    "resolved": "\"default\" | \"static\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The modal's backdrop.\nSpecify 'static' for a backdrop that doesn't close the modal when clicked outside the modal content."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'default'"
            },
            "customClass": {
                "type": "string",
                "attribute": "custom-class",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Custom CSS class to apply"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "fullscreen": {
                "type": "boolean",
                "attribute": "fullscreen",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies whether the modal should be displayed full-screen"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "modalId": {
                "type": "string",
                "attribute": "modal-id",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The ID of the inner dialog element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "position": {
                "type": "string",
                "attribute": "position",
                "mutable": false,
                "complexType": {
                    "original": "'bottom' | 'center' | 'top'",
                    "resolved": "\"bottom\" | \"center\" | \"top\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies the position of the modal"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'center'"
            },
            "showClose": {
                "type": "boolean",
                "attribute": "show-close",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies whether to show the close icon button at the top right of modal"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "true"
            },
            "showFullscreenToggle": {
                "type": "boolean",
                "attribute": "show-fullscreen-toggle",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Specifies whether to show the fullscreen toggle icon button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "el"; }
}
