// @stencil/react-output-target/dist/react-output-target.js
import { Project as w, VariableDeclarationKind as D } from "ts-morph";
import C from "node:path";
const $ = (n) => n.toLowerCase().split("-").map((a) => a.charAt(0).toUpperCase() + a.slice(1)).join(""), O = (n) => n.replace(/-([_a-z])/g, (a, r) => r.toUpperCase()), U = (n) => n.replace(/\/([a-z])/g, (a, r) => r.toUpperCase()), x = (n) => {
  const a = U(n);
  return O(`on-${a}`);
}, R = ({
  components: n,
  stencilPackageName: a,
  customElementsDir: r,
  defaultExport: g = !1,
  hydrateModule: p
}) => {
  const d = new w({ useInMemoryFileSystem: !0 }), c = p ? "" : `'use client';

`, s = `/**
 * This file was automatically generated by the Stencil React Output Target.
 * Changes to this file may cause incorrect behavior and will be lost if the code is regenerated.${p ? `
 * Do __not__ import components from this file as server side rendered components
 * may not hydrate due to missing Stencil runtime. Instead, import these components through the generated 'components.ts'
 * file that re-exports all components with the 'use client' directive.` : ""}
 */

`, i = `/* eslint-disable */
`, f = p ? "createComponent, createSSRComponent" : "createComponent", e = d.createSourceFile(
    "component.ts",
    `${c}${s}${i}
import React from 'react';
import { ${f} } from '@stencil/react-output-target/runtime';
import type { EventName, StencilReactComponent } from '@stencil/react-output-target/runtime';
  `
  );
  for (const m of n) {
    const u = m.tagName, l = $(u), T = `${l}Element`, b = `${l}CustomEvent`;
    e.addImportDeclaration({
      moduleSpecifier: `${a}/${r}/${u}.js`,
      namedImports: [
        {
          name: l,
          alias: T
        },
        {
          name: "defineCustomElement",
          alias: `define${l}`
        }
      ]
    });
    const j = (m.events || []).filter((t) => t.internal === !1), y = [];
    for (const t of j)
      if (Object.keys(t.complexType.references).length > 0) {
        for (const _ of Object.keys(t.complexType.references))
          t.complexType.references[_].location === "global" || e.addImportDeclaration({
            moduleSpecifier: a,
            namedImports: [
              {
                name: _,
                isTypeOnly: !0
              }
            ]
          });
        e.addImportDeclaration({
          moduleSpecifier: a,
          namedImports: [
            {
              name: b,
              isTypeOnly: !0
            }
          ]
        }), y.push({
          originalName: t.name,
          name: x(t.name),
          type: `EventName<${b}<${t.complexType.original}>>`
        });
      } else
        y.push({
          originalName: t.name,
          name: x(t.name),
          type: `EventName<CustomEvent<${t.complexType.original}>>`
        });
    const E = `${l}Events`;
    e.addTypeAlias({
      name: E,
      type: y.length > 0 ? `{ ${y.map((t) => `${t.name}: ${t.type}`).join(`,
`)} }` : "NonNullable<unknown>"
    });
    const N = `/*@__PURE__*/ createComponent<${T}, ${E}>({
    tagName: '${u}',
    elementClass: ${T},
    // @ts-ignore - React type of Stencil Output Target may differ from the React version used in the Nuxt.js project, this can be ignored.
    react: React,
    events: {${y.map((t) => `${t.name}: '${t.originalName}'`).join(`,
`)}} as ${E},
    defineCustomElement: define${l}
  })`, F = `/*@__PURE__*/ createSSRComponent<${T}, ${E}>({
    tagName: '${u}',
    properties: {${m.properties.filter((t) => !!t.attribute).map((t) => `${t.name}: '${t.attribute}'`).join(`,
`)}},
    hydrateModule: import('${p}')
  })`, I = e.addVariableStatement({
      declarationKind: D.Const,
      // React as never is a hack to by-pass a @types/react issue.
      declarations: [
        {
          name: l,
          type: `StencilReactComponent<${T}, ${E}>`,
          initializer: p ? `typeof window !== 'undefined'
              ? ${N}
              : ${F}` : N
        }
      ]
    });
    g ? e.addExportAssignment({
      isExportEquals: !1,
      expression: l
    }) : I.setIsExported(!0);
  }
  return e.organizeImports(), e.formatText(), e.getFullText();
}, P = ({
  components: n,
  esModules: a
}) => {
  const c = new w({ useInMemoryFileSystem: !0 }).createSourceFile(
    "component.ts",
    `'use client';

` + `/**
 * This file was automatically generated by the Stencil React Output Target.
 * Changes to this file may cause incorrect behavior and will be lost if the code is regenerated.
 */

` + `/* eslint-disable */
`
  );
  if (a)
    for (const o of n) {
      const s = o.tagName, i = $(s), f = $(o.tagName);
      c.addExportDeclaration({
        moduleSpecifier: `./${f}`,
        namedExports: a ? [`default as ${i}`] : n.map((e) => $(e.tagName))
      });
    }
  else
    c.addExportDeclaration({
      moduleSpecifier: "./components.server",
      namedExports: n.map((o) => $(o.tagName))
    });
  return c.organizeImports(), c.formatText(), c.getFullText();
}, A = async ({
  stencilPackageName: n,
  components: a,
  outDir: r,
  customElementsDir: g,
  esModules: p,
  excludeComponents: d,
  project: c,
  hydrateModule: o
}) => {
  const s = [], i = a.filter((e) => !(e.internal === !0 || d != null && d.includes(e.tagName)));
  if (i.length === 0)
    return [];
  const f = {};
  if (p === !0)
    for (const e of i) {
      const m = $(e.tagName), u = C.join(r, `${m}.ts`), l = R({
        components: [e],
        stencilPackageName: n,
        customElementsDir: g,
        defaultExport: !0,
        hydrateModule: o
      });
      f[u] = l;
    }
  else {
    const e = o ? (
      /**
       * If hydrate module is provided, we bundle all components in a single file for server side rendering.
       * Further down we then create a re-export file that imports this file and exports all components with
       * the 'use client' directive to ensure that the Stencil runtime is always being loaded but components
       * are also being rendered server side
       */
      C.join(r, "components.server.ts")
    ) : C.join(r, "components.ts"), m = R({
      components: i,
      stencilPackageName: n,
      customElementsDir: g,
      defaultExport: !1,
      hydrateModule: o
    });
    f[e] = m;
  }
  if (o) {
    const e = C.join(r, "components.ts"), m = P({
      components: i,
      esModules: p
    });
    f[e] = m;
  }
  return await Promise.all(
    Object.entries(f).map(async ([e, m]) => {
      const u = c.createSourceFile(e, m, { overwrite: !0 });
      await u.save(), s.push(u);
    })
  ), s;
}, h = "react-output-target", k = "dist/components", v = "dist-custom-elements", S = "dist-hydrate-script", J = ({
  outDir: n,
  esModules: a,
  stencilPackageName: r,
  excludeComponents: g,
  customElementsDir: p,
  hydrateModule: d
}) => {
  let c = k;
  return {
    type: "custom",
    name: h,
    validate(o) {
      if (p)
        c = p;
      else {
        const s = (o.outputTargets || []).find(
          (i) => i.type === v
        );
        if (s == null)
          throw new Error(
            `The '${h}' requires '${v}' output target. Add { type: '${v}' }, to the outputTargets config.`
          );
        if (s.dir !== void 0 && (c = s.dir), s.externalRuntime !== !1)
          throw new Error(
            `The '${h}' requires the '${v}' output target to have 'externalRuntime: false' set in its configuration.`
          );
      }
      if (d && (o.outputTargets || []).find((i) => i.type === S) == null)
        throw new Error(
          `The '${h}' requires '${S}' output target when the 'hydrateModule' option is set. Add { type: '${S}' }, to the outputTargets config.`
        );
      if (!n)
        throw new Error("The 'outDir' option is required.");
      if (r === void 0) {
        if (o.sys && o.packageJsonFilePath) {
          const { name: s } = JSON.parse(o.sys.readFileSync(o.packageJsonFilePath, "utf8"));
          r = s;
        }
        if (!r)
          throw new Error(
            `Unable to find the package name in the package.json file: ${o.packageJsonFilePath}. Please provide the stencilPackageName manually to the ${h} output target.`
          );
      }
    },
    async generator(o, s, i) {
      const f = i.createTimeSpan(`generate ${h} started`, !0), e = i.components, m = new w(), u = await A({
        outDir: n,
        components: e,
        stencilPackageName: r,
        customElementsDir: c,
        esModules: a === !0,
        excludeComponents: g,
        project: m,
        hydrateModule: d
      });
      await Promise.all(
        u.map((l) => s.fs.writeFile(l.getFilePath(), l.getFullText()))
      ), f.finish(`generate ${h} finished`);
    },
    __internal_getCustomElementsDir() {
      return c;
    }
  };
};
export {
  J as reactOutputTarget
};
